{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "79666866",
   "metadata": {},
   "source": [
    "# RAG implementation with LangChain framework with local LLMs - complete class\n",
    "\n",
    "In this notebook I create Retrieval-Augmented Generation (RAG) system based on my documents (some scientific papers) for by local LLMs implemented with LangChain framework\n",
    "\n",
    "__What is RAG?__\n",
    "\n",
    "Retrieval-Augmented Generation (RAG) is an architecture that augments a foundation LLM with external, up-to-date knowledge via a retrieval step before generation. It decouples the LLM's static training data from dynamic, domain-specific information.\n",
    "\n",
    "__Core Components (Local Setup):__\n",
    "\n",
    "* Vector Database: Stores embeddings of external documents (e.g., ChromaDB, FAISS, Weaviate).\n",
    "\n",
    "* Embedding Model: Converts text → dense vectors (e.g., all-MiniLM-L6-v2, BGE-small).\n",
    "\n",
    "* Retriever: Queries the DB for top-k relevant passages via similarity search (e.g., cosine distance).\n",
    "\n",
    "* LLM: Takes the retrieved context + user query → generates a truth/fact-grounded response.\n",
    "\n",
    "\n",
    "__This notebook builds on the previous RAG-LangChain notebook by encapsulating all the separate steps, functions and my trials into a single, concise and re-usable Class.__\n",
    "\n",
    "_Note: There are still some improvements to apply, but overall it works great!_"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c08a9437",
   "metadata": {},
   "source": [
    "## Complete RAG implementation with LangChain, ChromaDB, and options for local LLMs\n",
    "The solution supports document updates, multiple file types, various retrievers, and flexible LLM integration, all wrapped in a dedicted Class."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0678186c",
   "metadata": {},
   "source": [
    "### Small example of using sentence embedding models "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "3a4f5bb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sentence_transformers import SentenceTransformer\n",
    "from sentence_transformers.util import cos_sim\n",
    "\n",
    "# Manual loading of the new Qwen3-0.6B embedding model via Transformers\n",
    "# from sentence_transformers import SentenceTransformer\n",
    "# model = SentenceTransformer(\"Qwen/Qwen3-Embedding-0.6B\")\n",
    "# # Not working at the moment:\n",
    "# #ValueError: The checkpoint you are trying to load has model type `qwen3` but Transformers does not recognize this architecture. This could be because of an issue with the checkpoint, or because your version of Transformers is out of date.\n",
    "\n",
    "# Mixedbread-ai/mxbai-embed-large-v1 embedding model\n",
    "model = SentenceTransformer(\"mixedbread-ai/mxbai-embed-large-v1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c2ba894f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The prompt used for query retrieval tasks:\n",
    "# query_prompt = 'Represent this sentence for searching relevant passages: '\n",
    "query = \"A man is eating a piece of bread\"\n",
    "docs = [\n",
    "    \"A man is eating food.\",\n",
    "    \"A man is eating pasta.\",\n",
    "    \"The girl is carrying a baby.\",\n",
    "    \"A man is riding a horse.\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "da00eb42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "similarities: tensor([[0.7920, 0.6369, 0.1651, 0.3621]])\n"
     ]
    }
   ],
   "source": [
    "query_embedding = model.encode(query, prompt_name=\"query\")\n",
    "# query_embedding = model.encode(query, prompt=query_prompt)\n",
    "docs_embeddings = model.encode(docs)\n",
    "\n",
    "similarities = cos_sim(query_embedding, docs_embeddings)\n",
    "print('similarities:', similarities)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "735f3f9e",
   "metadata": {},
   "source": [
    "### Imports & Configs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "90b57f50",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_llms = {'qwen3-coder': 'qwen3-coder-30b-a3b-instruct', \n",
    "           'qwen3-thinking': 'qwen3-30b-a3b-thinking-2507',\n",
    "           'glm-4.5-air': 'glm-4.5-air',\n",
    "           'glm-4.5-air-m': 'glm-4.5-air-mlx',\n",
    "           'glm-4.5-air-5': 'glm-4.5-air@5bit',\n",
    "           'gpt-oss': 'gpt-oss-120b',\n",
    "           'gpt-oss-us': 'unsloth/gpt-oss-120b',\n",
    "           }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3a74a52e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using device: mps\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "from typing import List, Dict, Union\n",
    "import torch\n",
    "import chromadb\n",
    "\n",
    "# LangChain/Document Processing Imports\n",
    "from langchain.document_loaders import (\n",
    "    DirectoryLoader,\n",
    "    TextLoader,\n",
    "    PyPDFLoader,\n",
    "    UnstructuredMarkdownLoader,\n",
    "    Docx2txtLoader\n",
    ")\n",
    "from langchain_community.document_loaders import Docx2txtLoader\n",
    "\n",
    "from langchain import hub\n",
    "from langchain.docstore.document import Document\n",
    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
    "from langchain_community.chat_models import ChatOpenAI\n",
    "\n",
    "# ChromaDB Vector Store\n",
    "from langchain.vectorstores import Chroma\n",
    "from langchain.embeddings import HuggingFaceEmbeddings, OpenAIEmbeddings\n",
    "\n",
    "# Retrievers\n",
    "from langchain.retrievers import BM25Retriever, EnsembleRetriever\n",
    "from langchain.memory import ConversationBufferMemory\n",
    "\n",
    "# LLM Integration\n",
    "from langchain.llms import HuggingFacePipeline, OpenAI\n",
    "from langchain.chains import RetrievalQA, ConversationalRetrievalChain\n",
    "from langchain.chains.combine_documents import create_stuff_documents_chain\n",
    "from langchain.chains.retrieval import create_retrieval_chain\n",
    "\n",
    "# Prompts\n",
    "from langchain.prompts import ChatPromptTemplate, PromptTemplate\n",
    "\n",
    "\n",
    "workdir = Path.home()/\"Downloads\"/\"RAG_personal_knowledge_base\"\n",
    "\n",
    "# Configuration of embdedding and LLM models\n",
    "PERSIST_DIR = str(workdir/\"chroma_db_mxbai-embed-large-v1\")\n",
    "#EMBEDDING_MODEL = \"BAAI/bge-small-en-v1.5\"\n",
    "EMBEDDING_MODEL = \"mixedbread-ai/mxbai-embed-large-v1\"\n",
    "BASE_URL = \"http://localhost:1234/v1\"  # LM-Studio local API\n",
    "API_KEY = \"not-needed\"\n",
    "\n",
    "LLM_CONFIG = {\n",
    "    \"provider\": \"lmstudio\",  # Options: \"openai\", \"huggingface\", \"lmstudio\"\n",
    "    \"model_name\": my_llms['gpt-oss-us'],\n",
    "    \"openai_api_key\": API_KEY,\n",
    "    \"lmstudio_url\": BASE_URL\n",
    "}\n",
    "\n",
    "# device setup\n",
    "if torch.backends.mps.is_available():\n",
    "    DEVICE = torch.device(\"mps\")\n",
    "else:\n",
    "    DEVICE = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')\n",
    "print(f\"Using device: {DEVICE}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "564c2fe9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/homebrew/lib/python3.12/site-packages/langsmith/client.py:272: LangSmithMissingAPIKeyWarning: API key must be provided when using hosted LangSmith API\n",
      "  warnings.warn(\n",
      "/opt/homebrew/lib/python3.12/site-packages/langsmith/client.py:272: LangSmithMissingAPIKeyWarning: API key must be provided when using hosted LangSmith API\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "How PromptTemplate looks:\n",
      "input_variables=['agent_scratchpad', 'input', 'tool_names', 'tools'] input_types={} partial_variables={} metadata={'lc_hub_owner': 'hwchase17', 'lc_hub_repo': 'react', 'lc_hub_commit_hash': 'd15fe3c426f1c4b3f37c9198853e4a86e20c425ca7f4752ec0c9b0e97ca7ea4d'} template='Answer the following questions as best you can. You have access to the following tools:\\n\\n{tools}\\n\\nUse the following format:\\n\\nQuestion: the input question you must answer\\nThought: you should always think about what to do\\nAction: the action to take, should be one of [{tool_names}]\\nAction Input: the input to the action\\nObservation: the result of the action\\n... (this Thought/Action/Action Input/Observation can repeat N times)\\nThought: I now know the final answer\\nFinal Answer: the final answer to the original input question\\n\\nBegin!\\n\\nQuestion: {input}\\nThought:{agent_scratchpad}'\n",
      "\n",
      "How ChatPromptTemplate looks:\n",
      "input_variables=['agent_scratchpad', 'input'] optional_variables=['chat_history'] input_types={'chat_history': list[typing.Annotated[typing.Union[typing.Annotated[langchain_core.messages.ai.AIMessage, Tag(tag='ai')], typing.Annotated[langchain_core.messages.human.HumanMessage, Tag(tag='human')], typing.Annotated[langchain_core.messages.chat.ChatMessage, Tag(tag='chat')], typing.Annotated[langchain_core.messages.system.SystemMessage, Tag(tag='system')], typing.Annotated[langchain_core.messages.function.FunctionMessage, Tag(tag='function')], typing.Annotated[langchain_core.messages.tool.ToolMessage, Tag(tag='tool')], typing.Annotated[langchain_core.messages.ai.AIMessageChunk, Tag(tag='AIMessageChunk')], typing.Annotated[langchain_core.messages.human.HumanMessageChunk, Tag(tag='HumanMessageChunk')], typing.Annotated[langchain_core.messages.chat.ChatMessageChunk, Tag(tag='ChatMessageChunk')], typing.Annotated[langchain_core.messages.system.SystemMessageChunk, Tag(tag='SystemMessageChunk')], typing.Annotated[langchain_core.messages.function.FunctionMessageChunk, Tag(tag='FunctionMessageChunk')], typing.Annotated[langchain_core.messages.tool.ToolMessageChunk, Tag(tag='ToolMessageChunk')]], FieldInfo(annotation=NoneType, required=True, discriminator=Discriminator(discriminator=<function _get_type at 0x149509940>, custom_error_type=None, custom_error_message=None, custom_error_context=None))]], 'agent_scratchpad': list[typing.Annotated[typing.Union[typing.Annotated[langchain_core.messages.ai.AIMessage, Tag(tag='ai')], typing.Annotated[langchain_core.messages.human.HumanMessage, Tag(tag='human')], typing.Annotated[langchain_core.messages.chat.ChatMessage, Tag(tag='chat')], typing.Annotated[langchain_core.messages.system.SystemMessage, Tag(tag='system')], typing.Annotated[langchain_core.messages.function.FunctionMessage, Tag(tag='function')], typing.Annotated[langchain_core.messages.tool.ToolMessage, Tag(tag='tool')], typing.Annotated[langchain_core.messages.ai.AIMessageChunk, Tag(tag='AIMessageChunk')], typing.Annotated[langchain_core.messages.human.HumanMessageChunk, Tag(tag='HumanMessageChunk')], typing.Annotated[langchain_core.messages.chat.ChatMessageChunk, Tag(tag='ChatMessageChunk')], typing.Annotated[langchain_core.messages.system.SystemMessageChunk, Tag(tag='SystemMessageChunk')], typing.Annotated[langchain_core.messages.function.FunctionMessageChunk, Tag(tag='FunctionMessageChunk')], typing.Annotated[langchain_core.messages.tool.ToolMessageChunk, Tag(tag='ToolMessageChunk')]], FieldInfo(annotation=NoneType, required=True, discriminator=Discriminator(discriminator=<function _get_type at 0x149509940>, custom_error_type=None, custom_error_message=None, custom_error_context=None))]]} partial_variables={'chat_history': []} metadata={'lc_hub_owner': 'hwchase17', 'lc_hub_repo': 'openai-tools-agent', 'lc_hub_commit_hash': 'c18672812789a3b9697656dd539edf0120285dcae36396d0b548ae42a4ed66f5'} messages=[SystemMessagePromptTemplate(prompt=PromptTemplate(input_variables=[], input_types={}, partial_variables={}, template='You are a helpful assistant'), additional_kwargs={}), MessagesPlaceholder(variable_name='chat_history', optional=True), HumanMessagePromptTemplate(prompt=PromptTemplate(input_variables=['input'], input_types={}, partial_variables={}, template='{input}'), additional_kwargs={}), MessagesPlaceholder(variable_name='agent_scratchpad')]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Example of which keys PromptTemplate and ChatPromptTemplate have\n",
    "prompttemplate = hub.pull(\"hwchase17/react\")\n",
    "chatprompttemplate = hub.pull(\"hwchase17/openai-tools-agent\")\n",
    "print(f'How PromptTemplate looks:\\n{prompttemplate}\\n')\n",
    "print(f'How ChatPromptTemplate looks:\\n{chatprompttemplate}\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8033b4c",
   "metadata": {},
   "source": [
    "### RAG Class implementation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "edeb22f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "class LocalRAGSystem:\n",
    "\n",
    "    def __init__(self, persist_dir: str = PERSIST_DIR, \n",
    "                 embedding_model_name: str = EMBEDDING_MODEL):\n",
    "        \"\"\"\n",
    "        Initialize the local RAG system with Chroma vector store, embedding model and local LLM, plus all configuration parameters.\n",
    "        \"\"\"\n",
    "\n",
    "        self.persist_dir = persist_dir\n",
    "        self.embedding_model_name = embedding_model_name\n",
    "        self.CHUNK_SIZE_THRESHOLD = 1200\n",
    "\n",
    "        # Initialize embedding model (runs locally)\n",
    "        self.embeddings = HuggingFaceEmbeddings(  # HuggingFace sentence_transformers embedding models\n",
    "            model_name=self.embedding_model_name,\n",
    "            model_kwargs={'device': DEVICE},\n",
    "            encode_kwargs={\n",
    "            'normalize_embeddings': True,  # Crucial for cosine similarity\n",
    "            'batch_size': 32,  # Batch processing speedup\n",
    "            'dtype': torch.float32  # (fp32), can also use: torch.bfloat16,\n",
    "        })\n",
    "\n",
    "        # Initialize local LLM with OpenAI-compatible API (LM-Studio)\n",
    "        self.llm = ChatOpenAI(\n",
    "            model_name=LLM_CONFIG[\"model_name\"],\n",
    "            openai_api_key=LLM_CONFIG[\"openai_api_key\"],\n",
    "            openai_api_base=LLM_CONFIG[\"lmstudio_url\"],\n",
    "            temperature=1, # 0.7 for most models, for gpt-oss 1 is recommended\n",
    "            max_tokens=32000,\n",
    "        )\n",
    "\n",
    "        # Initialize Chroma vector store\n",
    "        if os.path.exists(self.persist_dir) and bool(os.listdir(self.persist_dir)):\n",
    "            print(\"Loading existing vector store...\")\n",
    "            try:\n",
    "                self.vectordb = Chroma(\n",
    "                    persist_directory=self.persist_dir,\n",
    "                    embedding_function=self.embeddings,\n",
    "                    collection_name=\"personal_knowledge\"\n",
    "                )\n",
    "                self.split_docs = None\n",
    "                # Check available collections\n",
    "                client = chromadb.PersistentClient(path=self.persist_dir)\n",
    "                print(f'Available collections in ChromaDB: {client.list_collections()}')\n",
    "            except Exception as e:\n",
    "                print(f\"Error loading existing vector store: {e}\")\n",
    "\n",
    "        else:\n",
    "            print(\"No existing vector store found.\")\n",
    "\n",
    "\n",
    "    def load_documents(self, source_dir: Union[str, Path]) -> List[Dict]:\n",
    "        \"\"\"Load documents from directory with multiple file support\"\"\"\n",
    "\n",
    "        source_dir = Path(source_dir)\n",
    "        if not source_dir.exists():\n",
    "            print(f\"Source directory does not exist: {source_dir}\")\n",
    "            return []\n",
    "        print(f'Contents of the documents dir: {list(source_dir.glob(\"**/*\"))}')\n",
    "\n",
    "        # Define loaders for different file types\n",
    "        loaders = {\n",
    "            \".txt\": TextLoader,\n",
    "            \".md\": UnstructuredMarkdownLoader,\n",
    "            \".pdf\": PyPDFLoader,\n",
    "            \".docx\": Docx2txtLoader,\n",
    "        }\n",
    "\n",
    "        documents = []\n",
    "        for ext, loader_class in loaders.items():\n",
    "            files = list(source_dir.glob(f\"**/*{ext}\"))\n",
    "            if files:\n",
    "                loader = DirectoryLoader(\n",
    "                    str(source_dir),\n",
    "                    glob=f\"**/*{ext}\",\n",
    "                    loader_cls=loader_class,\n",
    "                    show_progress=True\n",
    "                )\n",
    "                documents.extend(loader.load())\n",
    "                print(f\"Loaded {len(loader.load())} pages\")\n",
    "\n",
    "        return documents\n",
    "    \n",
    "\n",
    "    def chunk_docs(self, docs: List[Document]) -> List[Document]:\n",
    "        \"\"\"\n",
    "        Split documents into smaller chunks for RAG, optimized for LLMs. \n",
    "        chunk_size: Max characters per text segment (e.g., 500). Too small: context loss. Too large: LLM overflows/context noise.\n",
    "        chunk_overlap: Characters reused from previous chunk (e.g., 50). Prevents splitting mid-sentence → preserves context across chunks.\n",
    "        \"\"\"\n",
    "\n",
    "        text_splitter = RecursiveCharacterTextSplitter(\n",
    "            chunk_size=self.CHUNK_SIZE_THRESHOLD, \n",
    "            chunk_overlap=self.CHUNK_SIZE_THRESHOLD/10, # 10-20% overlap\n",
    "            length_function=len, # Function to measure text length (characters not tokens!)\n",
    "            is_separator_regex=False,\n",
    "            separators=[\"\\n\\n\", \"\\n\", \" \", \"\", \".\"],\n",
    "        )\n",
    "        split_docs = text_splitter.split_documents(docs)\n",
    "        print(f\"Total chunks created: {len(split_docs)}\")\n",
    "\n",
    "        return split_docs\n",
    "\n",
    "\n",
    "    def build_vector_db_from_docs(self, file_paths: List[str]):\n",
    "        \"\"\"Build ChromaDB vector store (persist to disk) with BGE-small embeddings from document chunks.\"\"\"\n",
    "\n",
    "        self.documents = self.load_documents(file_paths)\n",
    "        self.split_docs = self.chunk_docs(self.documents)\n",
    "\n",
    "        self.vectordb = Chroma.from_documents(\n",
    "            documents=self.split_docs,\n",
    "            embedding=self.embeddings,\n",
    "            persist_directory=self.persist_dir,\n",
    "            collection_name=\"personal_knowledge\"\n",
    "        )\n",
    "        print(f\"Vector store persisted at {str(self.persist_dir)}\")\n",
    "        print(f\"Added {len(self.split_docs)} documents to vector store\")\n",
    "\n",
    "        return self.vectordb\n",
    "\n",
    "\n",
    "    def update_vector_db(self, new_docs_path: List[str]):\n",
    "        \"\"\"Update existing vector store with new documents\"\"\"\n",
    "\n",
    "        new_documents = self.load_documents(new_docs_path)\n",
    "        new_split_docs = self.chunk_docs(new_documents)\n",
    "        print(new_split_docs)\n",
    "\n",
    "        self.vectordb.add_documents(new_split_docs)\n",
    "        # texts = [doc.page_content for doc in split_docs]\n",
    "        # metadatas = [doc.metadata for doc in split_docs]\n",
    "        # self.vectordb.add_texts(texts=texts, metadatas=metadatas)\n",
    "        print(\"\\nDatabase updated with new documents\")\n",
    "        self.vectordb.persist()\n",
    "\n",
    "        return self.vectordb\n",
    "\n",
    "\n",
    "    def list_existing_documents(self) -> List[str]:\n",
    "        \"\"\"\n",
    "        List all documents (metadata) in the current vector store\n",
    "        \"\"\"\n",
    "        try:\n",
    "            docs = self.vectordb.get()\n",
    "            return docs['metadatas'] if docs else []\n",
    "        except Exception as e:\n",
    "            print(f\"Error listing documents: {e}\")\n",
    "            return []\n",
    "\n",
    "\n",
    "    def create_retrievers(self):\n",
    "            \"\"\"\n",
    "            Create different types of retrievers\n",
    "            \"\"\"\n",
    "\n",
    "            if self.split_docs is None:\n",
    "                print(\"No split documents available for BM25Retriever creation. Need to build vector DB first.\")\n",
    "                # Vector Retriever (semantic search)\n",
    "                vector_retriever = self.vectordb.as_retriever(\n",
    "                search_type=\"similarity\",\n",
    "                search_kwargs={\"k\": 30}\n",
    "            )\n",
    "                return {\"vec_semantic\": vector_retriever}\n",
    "\n",
    "            else:\n",
    "                # BM25 Retriever (for keyword-based search)\n",
    "                bm25_retriever = BM25Retriever.from_documents(\n",
    "                    self.split_docs,\n",
    "                )\n",
    "                bm25_retriever.k = 30\n",
    "\n",
    "                # Vector Retriever (semantic search)\n",
    "                vector_retriever = self.vectordb.as_retriever(\n",
    "                    search_type=\"similarity\",\n",
    "                    search_kwargs={\"k\": 30}\n",
    "                )\n",
    "\n",
    "                # Ensemble Retriever combining BM25 and Vector\n",
    "                ensemble_retriever = EnsembleRetriever(\n",
    "                    retrievers=[bm25_retriever, vector_retriever],\n",
    "                    weights=[0.4, 0.6]\n",
    "                )\n",
    "\n",
    "                return {\n",
    "                    \"bm25\": bm25_retriever,\n",
    "                    \"vec_semantic\": vector_retriever,\n",
    "                    \"ensemble\": ensemble_retriever\n",
    "                }\n",
    "\n",
    "\n",
    "    def build_qa_chain(self,  \n",
    "              retriever_type: str = \"vec_semantic\", \n",
    "              use_conversation: bool = False):\n",
    "        \"\"\"\n",
    "        Builds a RetrievalQA chain with different retrieval methods, using the given local LLM, vector store and a question template.\n",
    "        \"\"\"\n",
    "\n",
    "        retrievers = self.create_retrievers()\n",
    "\n",
    "        # template = \"\"\"\n",
    "        #             Use ONLY the following context (source documents and metadata) to answer the question. \n",
    "        #             Do not use prior knowledge.\n",
    "        #             If the context doesn't contain the answer, respond with \"INSUFFICIENT CONTEXT\".\n",
    "\n",
    "        #             Context: {context}\n",
    "\n",
    "        #             Question: {question}\n",
    "\n",
    "        #             Helpful Answer:\n",
    "        #         \"\"\"\n",
    "        # prompt = PromptTemplate.from_template(template)\n",
    "\n",
    "        # The new implementation\n",
    "        prompt = ChatPromptTemplate.from_messages(\n",
    "        [\n",
    "            (\"system\",\n",
    "            '''Use ONLY the following context (source documents and metadata) to answer the question. \n",
    "                Do not use prior knowledge.\n",
    "                If the context doesn't contain the answer, respond with \"INSUFFICIENT CONTEXT\"'''\n",
    "             \"Context: {context}\"),\n",
    "            (\"human\", \"{question}\")\n",
    "        ])\n",
    "\n",
    "        if use_conversation:\n",
    "            memory = ConversationBufferMemory(\n",
    "                memory_key=\"chat_history\",\n",
    "                output_key=\"answer\",\n",
    "                return_messages=True\n",
    "            )\n",
    "            conv_chain = ConversationalRetrievalChain.from_llm(\n",
    "                llm=self.llm,\n",
    "                retriever=retrievers[retriever_type],\n",
    "                memory=memory,\n",
    "                combine_docs_chain_kwargs={\"prompt\": prompt, \n",
    "                    \"document_separator\": \"\\n\\n\"},\n",
    "                return_source_documents=True,\n",
    "            )\n",
    "            return conv_chain\n",
    "\n",
    "        else:\n",
    "            qa_chain = RetrievalQA.from_chain_type(\n",
    "                llm=self.llm,\n",
    "                chain_type=\"stuff\",  # All context stuffed into prompt\n",
    "                retriever=retrievers[retriever_type],\n",
    "                return_source_documents=True,  # For source verification\n",
    "                chain_type_kwargs={\"prompt\": prompt,\n",
    "                    \"document_separator\": \"\\n\\n\"  # Clean context formatting, improves prompt readability for the LLM.\n",
    "                }\n",
    "            )\n",
    "            # # the new LangChain api with \"create_*\"\n",
    "            # # “stuff” strategy = concatenate all retrieved chunks\n",
    "            # combine_chain = create_stuff_documents_chain(self.llm, prompt)\n",
    "            # retrieval_chain = create_retrieval_chain(\n",
    "            #     retriever=retrievers[retriever_type],\n",
    "            #     combine_docs_chain=combine_chain,\n",
    "            # )\n",
    "            return qa_chain\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cac7facd",
   "metadata": {},
   "source": [
    "### Running the complete RAG implementation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "65fa8241",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading existing vector store...\n",
      "Available collections in ChromaDB: [Collection(name=personal_knowledge)]\n",
      "Chroma DB path: /Users/danid/Downloads/RAG_personal_knowledge_base/chroma_db_mxbai-embed-large-v1\n",
      "Embedding model: mixedbread-ai/mxbai-embed-large-v1\n",
      "LLM model: unsloth/gpt-oss-120b\n",
      "Document chunk size: 1200\n"
     ]
    }
   ],
   "source": [
    "# Initialize the local RAG system, connecting to an existing vector db (if available at PERSIST_DIR)\n",
    "local_rag_sys = LocalRAGSystem()\n",
    "\n",
    "# check config setting are correct\n",
    "print(f'Chroma DB path: {local_rag_sys.persist_dir}')\n",
    "print(f'Embedding model: {EMBEDDING_MODEL}')\n",
    "print(f'LLM model: {LLM_CONFIG[\"model_name\"]}')\n",
    "print(f'Document chunk size: {local_rag_sys.CHUNK_SIZE_THRESHOLD}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d95c3a3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List existing documents in the vector store\n",
    "docs_in_db = local_rag_sys.list_existing_documents()\n",
    "docs = []\n",
    "for doc in docs_in_db:\n",
    "    docs.append(doc['title'])\n",
    "print(set(docs))  # Display unique document titles in the vector store"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "57409613",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contents of the documents dir: [PosixPath('/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation.pdf'), PosixPath('/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/.DS_Store'), PosixPath('/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Diffusion Sequence Models for Enhanced Protein Representation and Generation.pdf'), PosixPath('/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove.pdf'), PosixPath('/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/xTrimoPGLM- unified 100-billion-parameter pretrained transformer for deciphering the language of proteins.pdf'), PosixPath('/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf'), PosixPath('/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Polyethylene Degradation by a Rhodococcous Strain Isolated from Naturally Weathered Plastic Waste Enrichment .pdf'), PosixPath('/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Using artificial intelligence to document the hidden RNA virosphere.pdf')]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 7/7 [00:10<00:00,  1.47s/it]\n",
      "100%|██████████| 7/7 [00:09<00:00,  1.42s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 131 pages\n",
      "Total chunks created: 528\n",
      "Vector store persisted at /Users/danid/Downloads/RAG_personal_knowledge_base/chroma_db_mxbai-embed-large-v1\n",
      "Added 528 documents to vector store\n"
     ]
    }
   ],
   "source": [
    "# Build a new vector database from pdf-documents\n",
    "vectordb = local_rag_sys.build_vector_db_from_docs(file_paths=workdir/\"pdfs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2d1e2d55",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contents of the documents dir: [PosixPath('/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx')]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00, 109.65it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 153.94it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 1 pages\n",
      "Total chunks created: 56\n",
      "[Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Genomic and Metabolic Insights into Microbial Biodegradation of TNT\\n\\n\\n\\nAbstract'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Abstract\\n\\n2,4,6-Trinitrotoluene (TNT) is a recalcitrant explosive pollutant whose breakdown in the environment is facilitated by diverse microbes. This document summarizes key findings on TNT biodegradation, including the spectrum of TNT-transforming bacteria, fungi, and yeasts, the major genes and enzymes driving TNT metabolism, and the primary degradation pathways. Genomic analyses have identified critical catabolic genes (e.g. nitroreductases, oxygen-insensitive reductases of the Old Yellow Enzyme family, and azoreductases) that initiate TNT denitration and ring reduction. Metabolic pathway studies reveal that TNT is usually reduced stepwise to aminodinitrotoluenes and other intermediates, with some microbes forming Meisenheimer complexes leading to denitration. Omics approaches (transcriptomics, proteomics) provide insights into global stress responses and pathway activation in TNT-exposed microbes, highlighting inducible enzymes and regulatory networks. These advances lay the groundwork for biotechnological applications, from engineered microbial consortia to transgenic plants, aimed at efficient TNT bioremediation.\\n\\n\\n\\nIntroduction'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Introduction\\n\\nTNT as a Contaminant: TNT is a synthetic nitroaromatic explosive widely used in military and industrial applications. Past munitions manufacturing and usage have led to soil and groundwater contamination with TNT, which is toxic, persistent, and classified as a priority pollutant . TNT’s chemical structure (a toluene ring with three electron-withdrawing nitro groups) makes it highly resistant to conventional biodegradation. The nitro substituents strongly deactivate the aromatic ring, hindering direct oxidative attack by typical aromatic-degrading enzymes . Consequently, TNT remains in the environment for long periods, posing health and ecological risks.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Microbial Bioremediation: The use of microorganisms to detoxify and remove TNT from the environment offers a promising alternative to costly physicochemical cleanup methods . Certain bacteria and fungi have evolved enzymatic strategies to transform TNT despite its recalcitrance . Microbial TNT biodegradation typically involves reduction of nitro groups and other transformations that render TNT less toxic or more amenable to further breakdown. Over decades of research, numerous TNT-degrading strains have been isolated from soils, compost, and extreme environments, underscoring the metabolic diversity and adaptability of microbes . Understanding the genomics and metabolism of these organisms is crucial for enhancing bioremediation processes.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Scope of Document: This document provides a comprehensive analysis of microbial TNT degradation. We first catalog the known TNT-degrading species across bacteria, fungi, and yeasts, highlighting their roles in TNT transformation. We then delve into genomic insights, summarizing key catabolic genes, enzymes, and regulatory elements identified in these organisms. The major metabolic pathways for TNT breakdown (aerobic and anaerobic) are described with emphasis on intermediates and enzymatic steps. We also review findings from transcriptomic and proteomic studies that illuminate how microbes respond at a systems level to TNT.\\n\\n\\n\\nTNT-Degrading Microbial Species\\n\\nMicroorganisms capable of transforming or degrading TNT include a wide range of bacteria (both aerobes and anaerobes), as well as certain fungi and yeasts. These organisms employ different metabolic tactics to attack the TNT molecule, from nitro-group reduction to aromatic ring modifications. Below is an overview of the major TNT-degrading species and their known transformations:'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Aerobic Bacteria: Under oxic conditions, many soil bacteria can initiate TNT degradation by partial reduction of its nitro groups or removal of nitrite. Notable examples include:\\n\\n• Pseudomonas spp.: Various Pseudomonas strains can use TNT as a nitrogen source by sequentially removing nitro groups. Pseudomonas fluorescens strains harbor flavoprotein enzymes that reduce TNT to hydroxylamino- and amino-dinitrotoluenes (ADNTs) and even catalyze ring hydrogenation to form Meisenheimer complexes . P. putida and related species are frequently cited for their versatile TNT transformations and served as models in comparative genomic studies . Notably, Pseudomonas sp. strain IIBx and 1-2wt provided the first evidence of TNT mineralization (conversion of ring carbon to CO₂) in pure culture , although only a small fraction of TNT was fully mineralized.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Enterobacteriaceae: Enterobacter cloacae PB2 is capable of denitrating TNT. It reduces TNT via formation of a Meisenheimer complex and releases nitrite . An NADPH-dependent enzyme called pentaerythritol tetranitrate reductase (PETN reductase) was identified in this strain, which attacks TNT’s aromatic ring by hydride transfer . The result is TNT conversion to mono- and dihydride Meisenheimer complexes and subsequent elimination of nitrite. Other Enterobacteriaceae like Citrobacter and Klebsiella have also been reported to transform TNT, often producing amino-dinitrotoluenes and other polar metabolites .'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Other Aerobes: Bacillus and Staphylococcus species can reduce TNT, typically to monoamino dinitrotoluene (e.g., 2-amino-4,6-dinitrotoluene) with nitrite release. Mycobacterium and Rhodococcus (Gram-positive actinomycetes) are notable for their ability to perform aromatic ring reductions – for instance, Rhodococcus erythropolis (grown on picric acid) can hydrogenate the TNT ring to form Meisenheimer complexes. Some Serratia marcescens and Stenotrophomonas isolates even use TNT as a sole nitrogen source, co-metabolizing it in the presence of additional carbon sources. However, complete mineralization by single aerobic strains is rare; most yield partially reduced end-products that accumulate or bind to soil humic material.\\n\\n\\n\\nAnaerobic Bacteria: In oxygen-depleted environments, certain bacteria can reduce TNT more completely. The absence of O₂ favors multi-electron reductions that can remove all three nitro groups:'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Clostridium spp.: These obligate anaerobes are among the most effective TNT degraders in anoxic conditions. Clostridium bifermentans strains transform TNT through successive reductions to diaminonitrotoluenes (DANTs) and ultimately to 2,4,6-triaminotoluene (TAT) . Clostridium acetobutylicum, C. pasterianum, C. sordellii, and others have been shown to convert TNT almost fully to TAT . In some Clostridia, TNT reduction proceeds via hydoxylamino-dinitrotoluene intermediates that undergo a Bamberger rearrangement, producing aminophenols. The reduced aromatic ring may then fragment into smaller compounds, although the exact downstream steps are often unclear.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Sulfate-Reducers: Desulfovibrio species can anaerobically respire TNT. For example, Desulfovibrio sp. strain B uses TNT as a nitrogen source and may produce toluene as a downstream product . Other Desulfovibrio isolates reduce TNT to TAT and incorporate a substantial portion of TNT-derived carbon and nitrogen into biomass . These reactions suggest TNT can serve as a terminal electron acceptor in anaerobic respiration, supporting growth while yielding triaminotoluene . Indeed, in Pseudomonas sp. strain JLR11 (a denitrifying pseudomonad capable of both oxic and anoxic TNT metabolism), TNT reduction is coupled to ATP synthesis during anaerobic respiration.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Other Anaerobes: Veillonella alkalescens was one of the first organisms noted to reduce TNT to TAT under strictly anaerobic conditions. Even Escherichia coli (facultative anaerobe) can perform complete reduction of TNT to TAT when oxygen is absent, though it cannot metabolize TNT further. Methanogenic archaea like Methanococcus have shown limited TNT transformation (e.g., to DANT) . Overall, anaerobic bacteria tend to remove TNT’s nitro groups more completely than aerobes, but they often stall at the TAT stage, requiring additional steps (chemical or aerobic) to break the aromatic ring.\\n\\n\\n\\nFungi: Certain fungi, especially white-rot basidiomycetes, are able to degrade TNT by leveraging their lignin-degrading enzyme systems. Key fungal contributors include:'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• White-Rot Fungi (Basidiomycetes): Phanerochaete chrysosporium, a well-studied ligninolytic fungus, can mineralize a portion of TNT under nutrient-limited (lignin-degrading) conditions . Fungal TNT degradation generally starts with reduction of nitro groups to form ADNTs, similar to bacteria . However, P. chrysosporium secretes powerful oxidative enzymes – lignin peroxidase (LiP) and manganese peroxidase (MnP) – that can further attack the partially reduced TNT intermediates. Under ligninolytic conditions, TNT’s amino-dinitrotoluene and azoxy intermediates disappear as they are oxidatively fragmented, leading to significant mineralization to CO₂ . The exact fungal pathway is not fully resolved, but evidence suggests that after initial nitro reductions, extracellular peroxidases and oxidases perform unconventional ring cleavage and oxidation steps that bacteria generally cannot . Other white-rot fungi (e.g., Trametes versicolor, Nematoloma frowardii, Phlebia radiata) have also shown TNT transformation abilities. Notably, purified MnP from N. frowardii and Stropharia rugosoannulata was able to mineralize TNT reduction products in vitro . These fungi convert TNT into less'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Stropharia rugosoannulata was able to mineralize TNT reduction products in vitro . These fungi convert TNT into less recognizable fragments, with up to ~20–40% mineralization achieved in some cases .'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Other Fungi: Soil fungi like Cladosporium resinae and Cunninghamella echinulata (a filamentous fungus sometimes classified among zygomycetes) tolerate high TNT concentrations and reduce TNT predominantly to azoxy dimers . Although these fungi do not mineralize TNT themselves, their TNT-reducing activity can be harnessed in a two-step process: first, a tolerant fungus (e.g., Cladosporium) rapidly reduces TNT to less toxic amines; second, a white-rot fungus like P. chrysosporium mineralizes those products . This cooperative strategy can overcome TNT toxicity that would otherwise inhibit the white-rot fungus . Litter-decomposing basidiomycetes (e.g., Clitocybula dusenii) have also shown notable TNT mineralization (over 30% to ^14CO₂ under lab conditions) . A caveat for fungal bioremediation is that many of these fungi are adapted to wood or compost environments and may struggle to compete with native soil microbiota in the field .\\n\\n\\n\\nYeasts: Several yeast strains (unicellular fungi) have demonstrated the ability to transform TNT, often via unique mechanisms:'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Candida and Geotrichum: Environmental yeast isolates Candida sp. AN-L15 and Geotrichum sp. AN-Z4 degrade TNT through an alternative pathway dominated by hydride attack on the aromatic ring . These yeasts form mono- and dihydride Meisenheimer complexes of TNT (H^-–TNT and 2H^-–TNT), effectively adding hydrogen to the ring and eliminating nitrite . They produce relatively high levels of nitrite in the medium, indicating efficient denitration. For instance, Candida AN-L15 and Geotrichum AN-Z4 were observed to convert TNT into 2,4-dinitrotoluene (2,4-DNT) by removing one nitro group . The formation of 2,4-DNT (a compound more amenable to complete degradation) shows that yeasts can channel TNT into known dinitrotoluene degradation pathways . These yeasts also carry out the conventional nitro-group reductions (forming hydroxylamino- and amino-DNTs), but the hydride-mediated denitration pathway is dominant . Such capabilities make them interesting for co-contaminated sites where TNT and other pollutants (e.g., hydrocarbons) coexist, as Candida AN-L15 retained TNT-transforming activity even when grown on an n-alkane as carbon source .'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Yarrowia lipolytica: This robust yeast is known for degrading hydrocarbons and was also reported to degrade TNT. Studies found that factors like pH influence TNT transformation by Y. lipolytica, and like Candida, it likely produces reduced TNT derivatives (hydride complexes and ADNTs) (Ziganshin et al. 2010, Chemosphere). While details are fewer, Yarrowia’s tolerance to toxic organics makes it a candidate for TNT bioremediation in harsh conditions.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Comparative Genomic Insights: Genome sequencing of TNT-degrading strains has shed light on their evolutionary adaptations. For example, three Antarctic Pseudomonas isolates (TNT3, TNT11, TNT19) were sequenced and compared to other pseudomonads . These TNT-transforming strains harbored an unusually high abundance of xenobiotic-degrading genes, including those for nitroaromatic degradation, distinguishing them from less specialized relatives . Phylogenomics indicated these isolates are novel Pseudomonas species, potentially evolved in isolation to tackle xenobiotic compounds . In general, many TNT-degrading bacteria carry plasmid-borne or unique gene clusters for explosives degradation, suggesting horizontal gene transfer has disseminated TNT-catabolic traits among environmental microbes. Sequenced genomes (e.g., Pseudomonas putida KT2440, Rhodococcus opacus, Enterobacter cloacae PB2) serve as references to identify genes encoding TNT-transforming enzymes and regulatory proteins. These genomic resources allow researchers to compare pathways across species and to identify the minimal gene sets required for TNT degradation.\\n\\n\\n\\n\\n\\n\\n\\nGenomic Insights: Key Genes and Enzymes'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Genomic Insights: Key Genes and Enzymes\\n\\n\\n\\nMicrobial TNT degradation relies on a suite of specialized enzymes that initiate TNT transformation despite its inert aromatic structure. Genomic and biochemical studies have identified several key genes and enzymes involved in TNT breakdown:'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Nitroreductases (Type I): A common entry point for TNT metabolism is the stepwise reduction of nitro (-NO₂) groups to nitroso (-NO), hydroxylamino (-NHOH), and amino (-NH₂) groups. These reactions are carried out by oxygen-insensitive nitroreductases, often classified as type I nitroreductases. They typically use FMN or FAD cofactors and NAD(P)H as an electron donor to perform two-electron transfers without generating radical intermediates. Genes encoding such enzymes (e.g., nfsA, nfsB in E. coli) have been found in many TNT-respiring bacteria. An important example is the N-ethylmaleimide reductase nemA, identified in a TNT-degrading Citrobacter sp. Through integrated transcriptomics/proteomics, nemA was found to be the most upregulated gene during TNT exposure, and its product converts TNT to 4-ADNT and 2-ADNT. Nitroreductases are crucial for initiating TNT degradation, but they usually only remove nitro groups; they do not cleave the aromatic ring. Thus, their activity often yields dead-end aminodinitrotoluene products if no further pathway exists.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Nitrite Reductases and Assimilatory Enzymes: Though not directly modifying TNT, enzymes like assimilatory nitrite reductase (NasB) and ammonium-assimilating enzymes (e.g., glutamine synthetase) are vital for microbes that use TNT as a nitrogen source. Once nitrite is released from TNT (either extracellularly or intracellularly), it must be quickly taken up and reduced to ammonia . P. putida JLR11’s genome encodes a high-affinity nitrite uptake system and NasB, which together convert TNT-derived nitrite into ammonium for incorporation into amino acids. Similarly, some TNT degraders upregulate nitric oxide dioxygenase and related enzymes to detoxify any nitric oxide (NO) that might form from nitrite or hydroxylamine intermediates. These “housekeeping” enzymes are not unique to TNT pathways, but their activity is crucial to link TNT degradation with growth.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Old Yellow Enzyme (OYE) Family & Flavoprotein Reductases: Enzymes of the OYE family are broad-spectrum flavoprotein oxidoreductases that can attack nitroaromatic compounds in alternative ways. Several TNT-active enzymes belong to this group, including xenobiotic reductases and pentaerythritol tetranitrate reductase (PETN reductase). These enzymes can perform hydride ion attack on the aromatic ring of TNT, forming Meisenheimer complexes. For instance, Enterobacter cloacae PB2’s PETN reductase (gene petnR, also known as pnrA) was shown to catalyze the addition of hydride from NADPH to TNT, yielding a σ-complex that spontaneously eliminates nitrite. OYE-family enzymes like PETN reductase, XenA/B from P. putida, and NEM reductases have the unique ability to both reduce nitro groups and destabilize the aromatic ring. They are thus pivotal for routes that lead to ring fission or denitration. Notably, these enzymes are often promiscuous (they metabolize various nitroaromatics and quinones); TNT may not be their primary natural substrate, but evolution or engineering can enhance their TNT activity.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Azoreductases: An unexpected finding from genomic studies is the involvement of azoreductase enzymes in TNT degradation. Azoreductases normally reduce azo bonds (–N=N–) in dyes, but some can also reduce nitroaromatics or similar structures. A recent transcriptomic analysis of Pseudomonas sp. TNT3 (an Antarctic isolate) revealed an azoreductase-encoding gene as the most highly induced gene (>1000-fold upregulated) during TNT exposure. This suggests the enzyme plays a crucial role in TNT biotransformation, possibly by attacking TNT or its hydroxylamine intermediates. While the exact mechanism is under investigation, the strong induction implies that TNT (or a derivative) serves as a substrate or an inducer for this azoreductase. The widespread distribution of azoreductase genes in soil bacteria means they could be significant contributors to TNT reduction in the environment, especially under aerobic conditions where they might help prevent toxic intermediate buildup.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Dioxygenases and Ring-Cleavage Enzymes: Direct oxidative ring cleavage of TNT is very challenging due to its electron-poor ring, and no dedicated TNT dioxygenase has been identified. However, once TNT is partially reduced (e.g., to DNT or amino-DNTs), conventional aromatic degradative enzymes can act. For example, if TNT is denitrated to 2,4-DNT (as P. savastanoi can do), then the dinitrotoluene dioxygenase pathway (found in Burkholderia and Pseudomonas strains that degrade 2,4-DNT) can convert DNT into catechols and further into Krebs cycle intermediates. Some TNT-degraders have been noted to produce ring-fission products like p-cresol and methylphloroglucinol, hinting at cleavage pathways. Pseudomonas fluorescens strain B3468, for instance, metabolized TNT into phloroglucinol and pyrogallol, compounds indicative of ring cleavage followed by rearrangement. These steps likely require multi-component oxygenases and hydrolases acting on partially reduced rings. While specific genes for TNT ring cleavage are not unique (they overlap with general aromatic degradation genes), their coordinated expression is necessary to achieve complete TNT mineralization.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Monooxygenases and Dioxygenases: Enzymes that incorporate oxygen into aromatic rings (like ring-hydroxylating dioxygenases or monooxygenases) are central to degrading simpler nitroaromatics (e.g., dinitrotoluenes, nitrobenzene). However, with TNT, their role is limited due to TNT’s resistance to direct oxidation . Still, once an amino-dinitrotoluene is formed, certain dioxygenases can act. For instance, Pseudomonas aeruginosa strain MA01 was reported to further metabolize 2-ADNT and 4-ADNT aerobically. This likely involves a gentisate or catechol pathway after initial hydroxylation. Indeed, the β-ketoadipate pathway enzymes (e.g., catechol 1,2-dioxygenase, protocatechuate 3,4-dioxygenase) have been implicated when TNT metabolism proceeds past the initial reduction steps . In Buttiauxella S19-1, a monooxygenase and an aldehyde dehydrogenase were upregulated with TNT, hinting at a possible TNT transformation that produces an aldehyde intermediate (perhaps via methyl oxidation).'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Accessory and Regulatory Genes: Alongside catabolic enzymes, microbes possess regulatory and support genes that enable TNT degradation. These include: (1) oxidative stress response genes – TNT redox cycling can produce reactive oxygen species, so enzymes like superoxide dismutases and peroxidases are often upregulated to protect the cell ; (2) nitrosative stress genes – nitroso and hydroxylamino intermediates can cause nitrosative damage, leading to induction of nitric oxide detoxifying enzymes and DNA repair proteins; (3) transporters and efflux pumps – TNT and its products can be expelled from the cell by efflux pumps, and indeed TNT-challenged cells show increased expression of multidrug efflux systems , likely to reduce intracellular toxin accumulation; (4) regulators – specialized regulators (e.g., LysR-type or MarR-type regulators) may sense TNT or its metabolites and activate expression of degradation genes. Some TNT-degrading plasmids carry such regulatory genes ensuring the rest of the pathway is expressed only when TNT is present. The coordination of these elements is critical: for example, Pseudomonas sp. TNT3 downregulates energy-intensive processes (like motility'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='these elements is critical: for example, Pseudomonas sp. TNT3 downregulates energy-intensive processes (like motility and PQQ-dependent alcohol oxidation) while upregulating TNT-detoxifying enzymes under TNT exposure , indicating a global resource re-allocation controlled at the regulatory network level.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='In summary, the genetic toolkit for TNT biodegradation comprises mostly reductive enzymes that either remove nitro groups or destabilize the aromatic ring, along with supporting systems to manage stress and assimilate released nitrogen. Ongoing genome sequencing and functional studies continue to uncover new genes (such as novel reductases) that expand our understanding of TNT metabolism in microbes.\\n\\n\\n\\nMetabolic Pathways of TNT Degradation\\n\\nMicrobial TNT degradation proceeds via a few primary pathways / major routes: (A) nitro-group reduction and (B) aromatic ring (Meisenheimer) reduction, along with subsequent steps. Below is an overview of the main pathways and their initial products:'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Aerobic Degradation Routes: In the presence of oxygen, TNT is not readily mineralized, but it undergoes a series of reductive transformations: the dominant pathway is sequential nitro-group reduction. Typically, one nitro group is reduced to a hydroxylamino derivative (e.g., 4-hydroxylamino-2,6-dinitrotoluene), which can then be further reduced to an amino group, yielding an aminodinitrotoluene (e.g., 4-amino-2,6-dinitrotoluene) . Often the 2- and 4- positions are targeted, producing 2-ADNT and 4-ADNT as major metabolites. Both 2-ADNT and 4-ADNT are typically recalcitrant – they accumulate or bind to soil components rather than being metabolized fully . Moreover, the hydroxylamino intermediates of TNT can undergo abiotic condensation reactions: two such molecules can combine (via a nitroso and a hydroxylamino group) to form azoxy dimers (e.g., 2,2’-azoxytetranitrotoluene) . These dimers (and their further reduced azo forms) are highly recalcitrant polymers that are not readily metabolized . Thus, a common outcome of aerobic TNT transformation is the formation of amino-DNTs and azoxy compounds that persist. This is why many early aerobic isolates showed no mineralization – most of'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='amino-DNTs and azoxy compounds that persist. This is why many early aerobic isolates showed no mineralization – most of the ^14C from TNT ended up in bound residue or stable metabolites.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Nitro Group Reduction (Aerobic Nitro-Reduction Pathway): Most TNT-degrading microbes reduce one or more of TNT’s nitro substituents in a stepwise, two-electron process. Type I nitroreductases (and some OYEs) catalyze the successive reduction of –NO₂ groups to nitroso (-NO), hydroxylamino (-NHOH), and finally amino (-NH₂) groups. For example, TNT is first converted to 4-nitroso-2,6-dinitrotoluene, then 4-hydroxylamino-2,6-DNT, and then 4-amino-2,6-dinitrotoluene (4-ADNT). Similarly, reduction at the 2-position yields 2-ADNT via 2-HADNT and 2-nitroso intermediates. These aminodinitrotoluene (ADNT) isomers are common end-products in aerobic cultures. Because of TNT’s symmetry, both 2-ADNT and 4-ADNT (and small amounts of 2,4-diamino-6-nitrotoluene) accumulate. Importantly, these reductions do not break the aromatic ring; instead they yield slightly less toxic derivatives that often adsorb to soil or biopolymers. A side reaction of hydroxylamino-DNTs is spontaneous coupling: two hydroxylaminodinitrotoluene molecules can condensate to form azoxy dimers (e.g., 2,2’,6,6’-tetranitro-4,4’-azoxytoluene) . These azoxy compounds are intensely colored, highly recalcitrant, and represent'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='. These azoxy compounds are intensely colored, highly recalcitrant, and represent dead-end immobilization products in many TNT-fed cultures . In summary, aerobic nitro-reduction rapidly eliminates TNT’s explosive character and toxicity by producing ADNTs and polymeric azoxy compounds, but it typically does not lead to full mineralization. However, some aerobic pathways allow further degradation: a fraction of TNT may undergo hydride attack (Meisenheimer complex formation). Enzymatic hydride addition (via OYEs or related reductases) leads to a σ-complex that spontaneously ejects a nitro group as nitrite . For example, Pseudomonas savastanoi was shown to eliminate nitrite from TNT, yielding 2,4-DNT . Once TNT is denitrated to a dinitrotoluene, the ring is less electron-poor and can be attacked by oxygenases. In favorable cases, this can connect TNT degradation to known DNT degradation pathways, involving dioxygenation of the aromatic ring and subsequent ring fission. Additionally, some organisms can remove multiple nitro groups aerobically: a Pseudomonas isolate grew on TNT as sole nitrogen source and was suggested to sequentially remove all three nitro substituents, as evidenced by'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='on TNT as sole nitrogen source and was suggested to sequentially remove all three nitro substituents, as evidenced by detection of DNTs, mononitrotoluene, and toluene in culture . The release of nitrite and its reduction to ammonium (a process called denitration and assimilation) means TNT’s nitrogen can be incorporated into biomass . Indeed, certain pseudomonads can assimilate up to ~85% of TNT’s nitrogen into organic forms in the cell . The toluene that remains after complete denitration can then be degraded via normal aerobic toluene pathways (e.g., toluene dioxygenase to catechol). There are also reports of partial ring cleavage under aerobic conditions yielding products like p-cresol and methylphloroglucinol【17†look】, likely via complex rearrangements of triaminotoluene or aminophenol intermediates. Overall, aerobic TNT metabolism tends to transform TNT into less toxic or less mobile forms (by reduction and binding), but rarely achieves full mineralization unless additional steps (chemical or a secondary organism) are employed.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Partial Oxidative Pathways: Although TNT’s nitro substituents strongly inhibit direct oxygenase attack, there are instances of cometabolic oxidation of TNT. One example is oxidation of the methyl group. Mycobacterium strain wa1 (initially grown on propane) oxidized TNT’s methyl side chain to 4,6-dinitro-2-nitrobenzyl alcohol, and further to 2,4,6-trinitrobenzoate. This likely involved a nonspecific monooxygenase or alkane hydroxylase. The benzyl alcohol and benzoate produced are still highly substituted with nitro groups and were not fully degraded in that study. Another possible oxidative route is monooxygenation of an aromatic nitro group to form a nitroarene oxide or a substituted catechol, but for TNT this has not been clearly documented (such reactions are more common with dinitrotoluenes). Generally, oxidative transformations of TNT by single organisms are minor and require prior reduction steps. Ligninolytic fungi use oxidative enzymes (peroxidases) that attack TNT’s aromatic ring, but the outcome is typically fragmentation into smaller aromatic pieces or coupling to form complex polymers . These oxidative cometabolic steps are important in bioremediation (e.g., fungal'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='or coupling to form complex polymers . These oxidative cometabolic steps are important in bioremediation (e.g., fungal enzymatic treatment of soil) but yield a mixture of partial oxidation products rather than a defined pathway to CO₂.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Anaerobic Degradation Routes: In oxygen-free environments, TNT undergoes more complete reduction. The hallmark of anaerobic TNT metabolism is the production of triaminotoluene (TAT). This results from successive 6-electron reductions (each nitro group accepting six electrons to become an amine). For example, cell extracts of Veillonella and several Clostridium spp. quantitatively convert TNT to TAT with H₂ as an electron donor . The strongly reducing conditions needed for the third nitro group reduction are usually only reached anaerobically . Once formed, TAT is relatively unstable in aerobic conditions and can undergo auto-oxidation or polymerization. Under strictly anaerobic conditions, some novel transformations of TAT have been observed: studies with Clostridium suggest TAT might undergo a Bamberger rearrangement if partial reoxidation occurs, producing aminophenols (e.g., 2,4-diamino-6-nitrophenol) . Alternatively, TAT can lose the amino substituents through deamination or ammonification reactions. In one pathway, TAT was proposed to break down into toluene (regenerating the hydrocarbon backbone) plus ammonia【17†look】. For instance, Desulfovibrio cultures yielded toluene as'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='(regenerating the hydrocarbon backbone) plus ammonia【17†look】. For instance, Desulfovibrio cultures yielded toluene as a putative intermediate, implying a reduction of TNT all the way to toluene (which would require elimination of all nitro substituents as either NH₄⁺ or molecular nitrogen) . Toluene could then be anaerobically oxidized to p-cresol or other molecules by certain anaerobes, or it might accumulate until oxic conditions allow its degradation. Another observed anaerobic product is p-cresol, which suggests a methyl group shift or ring cleavage of a fully reduced intermediate【17†look】. It’s important to note that purely anaerobic stages tend not to cleave the aromatic ring completely to CO₂; instead, they set the stage for subsequent aerobic oxidation. This is the principle behind sequential anaerobic-aerobic treatments: anaerobes fully reduce TNT to TAT (removing toxicity and releasing nitrite as benign ammonium), then upon exposure to oxygen, the reduced aromatic (TAT or its derivatives) is more amenable to oxidative ring fission by aerobic microbes or fungi.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Complete Reduction (Anaerobic Pathway): In oxygen-free environments, TNT can be fully reduced to TAT (2,4,6-triaminotoluene) by certain bacteria. This occurs via sequential addition of six electrons (two per nitro group) through nitroso and hydroxylamino intermediates. Clostridium bifermentans and similar anaerobes perform this reduction as part of fermentative metabolism. TAT is not readily utilized as a growth substrate; instead, it tends to polymerize (auto-oxidizing to colored oligomers) or adsorb to soil. There is speculation that some anaerobes might further metabolize TAT after initial formation – e.g., via deamination or ring reduction – but the exact steps remain unknown. In essence, the anaerobic pathway detoxifies TNT by eliminating all nitro groups, often immobilizing the product. A variant of anaerobic TNT metabolism is found in Desulfovibrio spp. (sulfate-reducers), which also produce TAT and may reduce TNT as an electron sink. While complete reduction does not yield energy-rich products for the microbes, it does alleviate toxicity and can be coupled with other anaerobic processes (e.g., use of TNT as a terminal electron acceptor). In remediation strategies, an'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='with other anaerobic processes (e.g., use of TNT as a terminal electron acceptor). In remediation strategies, an anaerobic phase is sometimes used to reduce TNT to TAT, followed by an aerobic phase to deal with the resulting amino compounds.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Aromatic Ring Reduction and Denitration (Meisenheimer Pathway): A fewer number of microbes initiate TNT degradation via direct attack on the aromatic ring. In this route, a hydride ion (from NAD(P)H via an OYE-family enzyme) adds to TNT’s ring, forming a Meisenheimer complex (a cyclohexadienyl anion). This hydride-Meisenheimer complex (H⁻-TNT) is a transient intermediate that can either accept a second hydride to form a dihydride-Meisenheimer (2H⁻-TNT) or react with other TNT metabolites. Crucially, Meisenheimer complexes can eliminate a nitro group as nitrite (NO₂⁻), denitrating the ring. The result is the formation of a dinitrotoluene (DNT) or nitrotoluene product. For instance, some Pseudomonas and Enterobacter isolates produce 2,4-dinitrotoluene from TNT by this mechanism. Enterobacter PB2’s PETN reductase exemplifies this pathway: it reduces TNT to H⁻-TNT and 2H⁻-TNT, which then eliminate nitrite. Releasing nitrite lowers the aromatic ring’s electron-withdrawal, making it more susceptible to oxygenase attack. Indeed, removal of even one nitro group is critical for further metabolism of TNT . Once a DNT or mononitrotoluene forms, bacteria can channel those into known catabolic'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='for further metabolism of TNT . Once a DNT or mononitrotoluene forms, bacteria can channel those into known catabolic pathways (many strains readily metabolize DNTs via dioxygenases that cleave the ring and release the remaining nitro as nitrite ). However, TNT denitration is relatively rare. Pseudomonas savastanoi was reported to convert TNT to 2,4-DNT , and Rhodococcus erythropolis and some Mycobacterium spp. have been observed to form Meisenheimer complexes in vivo . In pure enzyme studies, P. fluorescens XenB catalyzes both nitro reductions and aromatic-ring reductions; notably, XenB-mediated TNT turnover produces nitrite and novel dimers (e.g. amino-tetranitrobiphenyls) via coupling of Meisenheimer intermediates with hydroxylamines . In practice, the Meisenheimer/denitration pathway often occurs concurrently with nitro reduction. Its key benefit is that it opens the door to complete aromatic degradation by removing recalcitrant nitro groups. If the resultant DNT (or further oxidized products like 4-nitrocatechol) enters central aromatic pathways, mineralization of part of the TNT molecule can proceed . Overall, this pathway highlights how some microbes use unusual enzymatic'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='of part of the TNT molecule can proceed . Overall, this pathway highlights how some microbes use unusual enzymatic reactions to destabilize TNT’s ring, releasing nitrite as a nutrient and facilitating further breakdown.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Fungal Pathways: Fungi like P. chrysosporium combine both strategies. The fungus initially behaves like an anaerobe in that its mycelial cells (especially under low-nitrogen conditions) reduce TNT to ADNTs and related products without oxygen involvement (possibly via a plasma membrane redox system) . Then, in the ligninolytic phase (which is oxidative), extracellular enzymes attack the reduced compounds. Azoxy dimers produced by the fungus are broken down by MnP and LiP, preventing their accumulation . The fungal system thereby circumvents the dead-end that bacteria face with ADNTs/azoxytoluenes, leading to further oxidation and some mineralization. For example, P. chrysosporium converted 4-ADNT into 4-formyl-2,6-DNT (an oxidized derivative), suggesting an unusual formamide intermediate as the ring is oxidatively opened . Ultimately, the fungal pathway mineralizes a portion of TNT, with the remainder becoming incorporated into humic polymers (which are biologically benign and immobile) .'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Key Intermediates and Fate: Across these pathways, the recurring intermediates are: hydroxylamino-dinitrotoluenes, amino-dinitrotoluenes (ADNTs), diaminonitrotoluenes (DANTs), and triaminotoluene (TAT). Side products include nitrite (NO₂⁻) and ammonium (NH₄⁺) from nitro group removal, which many microbes assimilate as nitrogen source . The ADNTs can follow one of three fates: (1) condense to dimers (environmental sink); (2) be further reduced to DANT/TAT (mostly anaerobic); or (3) be oxidized/cometabolized if another organism or enzyme can act on the aminated ring (e.g., certain Rhodococcus can oxidize aminophenols). The formation of Meisenheimer complexes is a branching point that leads to denitration – producing DNT – which is significant because 2,4-DNT and 2,6-DNT have known bacterial degradation pathways (via dioxygenase to dihydrodiol, then dihydroxylated benzoates). Thus, any TNT-degrading community that can convert TNT to DNT in situ has a chance at complete degradation through cooperative metabolism. Indeed, complete mineralization of TNT in soil is most often achieved by consortia, where one organism’s partially degraded product is further metabolized by another . Many'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='often achieved by consortia, where one organism’s partially degraded product is further metabolized by another . Many TNT metabolites (especially those with –NH₂ groups) readily bind to soil organic matter, reducing bioavailability. While this binding “locks away” the toxicity and limits spread , it also means the TNT carbon skeleton persists in the environment in bound form. Therefore, an active area of research is encouraging pathways that lead to either mineralization or safe immobilization, depending on remediation goals.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='In summary, TNT degradation pathways are complex and can follow multiple routes depending on the organism and conditions. Aerobic processes excel at initial transformations and co-metabolic turnover of some intermediates, whereas anaerobic processes excel at complete nitro reduction. Combining these processes (either within a single versatile organism or in staged treatments) is often key to achieving extensive TNT breakdown. The pathway figure above encapsulates these routes, emphasizing that while complete TNT mineralization is challenging, microbes can dismantle TNT’s structure through a network of reductive and oxidative steps.\\n\\n\\n\\nMulti-Omics Approaches and Systems Biology Insights'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Transcriptomic Profiling: Genome-wide RNA expression analyses reveal which genes are turned on or off in the presence of TNT. A recent study examined the transcriptome of Pseudomonas sp. TNT3 (an Antarctic TNT-degrader) during exposure to 100 mg/L TNT. Out of the bacterium’s ~5,000 genes, 194 showed significant differential expression upon TNT challeng  The most upregulated gene (>1000-fold increase in mRNA) encoded an azoreductase enzyme, underscoring the importance of this enzyme in TNT degradation for this strain. Other upregulated genes included those involved in nitroaromatic detoxification (likely various reductases and hydrolases), as well as genes for managing oxidative stress, thiol-specific stress, and nitrosative stress. This indicates TNT exposure generates reactive oxygen and nitrogen species inside the cell, triggering defensive responses. Additionally, genes for efflux pumps were induced, suggesting the cell actively exports TNT or toxic intermediates to reduce intracellular damage. On the flip side, Pseudomonas TNT3 downregulated many genes related to its electron transport chain, certain metabolic pathways, and motility functions during TNT exposure. This likely'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='to its electron transport chain, certain metabolic pathways, and motility functions during TNT exposure. This likely reflects a shift to a slower-growing, stress-resistant state: the bacterium conserves energy and curtails activities like flagella-driven movement while it deals with TNT. Such transcriptomic data provide a global picture: TNT degradation is not just a matter of expressing a few catabolic enzymes, but involves broad adjustments to metabolism and stress physiology.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Proteomic Analyses: Complementing transcriptomics, proteomics examines changes in the protein complement of a cell in response to TNT. In one study, researchers performed an integrated proteome and transcriptome analysis of a TNT-degrading Citrobacter sp. before and after TNT treatment. They found hundreds of genes and dozens of proteins with altered expression. Notably, the NemA nitroreductase (discussed earlier) was strongly upregulated at both mRNA and protein levels, correlating with its role in producing ADNTs from TNT. Proteomics also revealed an increase in proteins for energy production and conversion (indicating the cells ramp up certain respiratory or fermentative processes, possibly to regenerate NADPH for reductases) and for amino acid transport and metabolism. The latter might suggest the cell is scavenging nitrogen from TNT (after reducing NO₂ to NH₄⁺, assimilating it into amino acids). Upregulation of chaperones and turnover proteins was also observed, which is common under toxic stress as cells refold or degrade damaged proteins. Meanwhile, proteins involved in carbohydrate metabolism and carbon transport were downregulated, implying the presence of TNT causes'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='involved in carbohydrate metabolism and carbon transport were downregulated, implying the presence of TNT causes the cell to slow down growth-related pathways (since TNT itself is not a carbon source that can feed central metabolism efficiently). Another proteomic study on Pseudomonas sp. HK-6 found that TNT stress induced production of exopolymer and biofilm-related proteins (Lee et al. 2008), hinting that bacteria may form biofilms to protect themselves from TNT toxicity. By surrounding themselves in extracellular polysaccharides, the cells can limit TNT diffusion into their cytoplasm and localize degradative enzymes in the matrix. Such a strategy would be beneficial in soils where microbes often exist in biofilms.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Metabolomics and Pathway Confirmation: While less reported than transcript/proteome data, metabolomic analyses (identifying intermediate compounds) and isotope tracing have been used to confirm TNT degradation pathways. For example, the detection of unique metabolites like 4-formamide-2,6-DNT in Phanerochaete cultures provided evidence for previously hypothesized steps in fungal TNT degradation. Stable isotope labeling of TNT (with ^15N or ^13C) has been used to track the flow of TNT-derived atoms through metabolic networks. These studies show incorporation of TNT’s nitrogen into amino acids and biomass (confirming nitrite assimilation) and reveal partial oxidation of the aromatic ring into central metabolites in some systems. Combined with transcriptomics, if genes of unknown function are upregulated alongside detection of unusual metabolites, researchers can link those genes to specific pathway steps.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Systems Biology and Modeling: With large omics datasets, scientists have begun constructing regulatory and metabolic network models for TNT-degrading organisms. For instance, integrating the transcriptomic data of Pseudomonas TNT3 with its genome annotation allows identification of regulatory motifs and potential transcription factors controlling the TNT response. One can infer that a particular regulator might activate the azoreductase and efflux pump genes in the presence of TNT. These models help predict how genetic modifications might improve TNT degradation. As an example, if a key enzyme is only weakly expressed, a model may suggest overexpressing it to enhance the degradation flux. Conversely, if certain pathways (like PQQ-dependent alcohol oxidation in Pseudomonas TNT3) are downregulated as an energy-saving strategy, knockout of those pathways might make the strain inherently more tolerant to TNT. Furthermore, systems biology can guide consortium design: by understanding the complete metabolic network affected by TNT, one can pair organisms with complementary capabilities (e.g., one that excels at reduction and another that can degrade the reduced products).'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='• Functional Genomics: Another omics-related approach is random mutagenesis or knockout studies combined with sequencing to find genes essential for TNT metabolism. Transposon mutagenesis, for instance, can create mutants of a TNT-degrader; those that lose the ability to transform TNT are then sequenced to identify the interrupted gene. Using this approach, researchers have confirmed the roles of specific reductases and transporters in TNT degradation. Functional gene arrays (FGA) have also been developed to monitor the expression of dozens of known nitroaromatic degradation genes in environmental samples, providing insight into which organisms and pathways are active in situ when TNT is present.'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='In summary, omics studies have expanded our understanding from individual enzymes to the global response of microorganisms to TNT. They reveal that TNT degradation is an orchestrated event involving stress mitigation, energy management, and coordinated expression of catabolic genes. Importantly, these insights can direct genetic engineering and adaptive laboratory evolution to create strains with enhanced TNT-degrading capabilities or robustness. They also assist in developing biomarkers (e.g., transcriptional reporters) to monitor bioremediation progress in the field. As omics technologies advance, we can expect even more detailed systems-level models of TNT biodegradation, possibly enabling predictive simulations of bioremediation outcomes under various environmental scenarios.\\n\\n\\n\\nReferences\\n\\n1. Ramos, J.L., et al. (2001). Biological degradation of 2,4,6-trinitrotoluene (TNT). Microbiology and Molecular Biology Reviews, 65(3): 335–352 .'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='2. Cabrera, M.Á., Márquez, S.L., & Pérez-Donoso, J.M. (2022). Comparative genomic analysis of Antarctic Pseudomonas isolates with 2,4,6-trinitrotoluene transformation capabilities reveals their unique features for xenobiotics degradation. Microorganisms, 10(8): 1581 .\\n\\n3. Cabrera, M.Á., Márquez, S.L., & Pérez-Donoso, J.M. (2024). New insights into xenobiotic tolerance of Antarctic bacteria: Transcriptomic analysis of Pseudomonas sp. TNT3 during 2,4,6-trinitrotoluene biotransformation. Environmental Science and Pollution Research, 31(11): 17256–17274 .\\n\\n4. Liao, H.Y., et al. (2018). Integrated transcriptomic and proteomic analysis of Citrobacter sp. reveals the biodegradation mechanism of 2,4,6-trinitrotoluene. Journal of Hazardous Materials, 349: 79–90 .\\n\\n5. Ziganshin, A.M., et al. (2007). Hydride-mediated reduction of 2,4,6-trinitrotoluene by yeasts as a way to its deep degradation. Mikrobiologiia (Microbiology), 76(6): 766–773 .\\n\\n6. EAWAG-BBD (1999). 2,4,6-Trinitrotoluene Degradation Pathway (Aerobic and Anaerobic). Eawag Biocatalysis/Biodegradation Database, University of Minnesota .'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='7. Hannink, N.K., et al. (2001). Phytodetoxification of TNT by transgenic plants expressing a bacterial nitroreductase. Nature Biotechnology, 19(12): 1168–1172 .\\n\\n8. Duque, E., et al. (1993). Identification of a Pseudomonas strain that mineralizes TNT and characterization of its metabolic pathway. Journal of Bacteriology, 175(8): 2278–2283 (First report of TNT mineralization by Pseudomonas).\\n\\n9. Vorbeck, C., et al. (1998). Initial reductive reactions in aerobic microbial metabolism of 2,4,6-trinitrotoluene. Applied and Environmental Microbiology, 64(1): 246–252 . (Describes hydride Meisenheimer complex formation by Rhodococcus and Mycobacterium).\\n\\n10. Rylott, E.L., et al. (2011). Engineering plants and microbes for biodegradation of explosives. Current Opinion in Biotechnology, 22(3): 434–440. (Review of advances in TNT/RDX phytoremediation and microbial degradation).\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\nExtra material: additional genes/pathways related to TNT-biodegradation\\n\\n\\n\\nNADPH dehydrogenase NamA'), Document(metadata={'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx'}, page_content='Extra material: additional genes/pathways related to TNT-biodegradation\\n\\n\\n\\nNADPH dehydrogenase NamA\\n\\nNADPH dehydrogenase NamA is\\xa0an enzyme that catalyzes the reduction of the double bond in various α,β-unsaturated aldehydes and ketones.\\xa0It also reduces the nitro group of nitroesters and nitroaromatic compounds.\\xa0This enzyme is thought to play a role in detoxification processes.\\xa0\\n\\nKey aspects of NADPH dehydrogenase NamA:\\n\\nFunction:\\xa0Catalyzes the reduction of α,β-unsaturated aldehydes and ketones, and the nitro group of nitroesters and nitroaromatic compounds.\\xa0\\n\\nRole:\\xa0Believed to be involved in detoxification pathways.\\xa0\\n\\nLocation:\\xa0Found in various organisms, including Bacillus subtilis.\\xa0\\n\\nOther names:\\xa0Can also be referred to as NADPH:(quinone-acceptor) oxidoreductase, reduced nicotinamide adenine dinucleotide phosphate (quinone), dehydrogenase, NADPH oxidase, and NADPH2 dehydrogenase (quinone).\\xa0\\n\\nCofactors:\\xa0May utilize FAD and a flavoprotein as cofactors.')]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Database updated with new documents\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/y_/2xtp4jn551g4kc50l69n92dr0000gn/T/ipykernel_90962/1031254841.py:135: LangChainDeprecationWarning: Since Chroma 0.4.x the manual persistence method is no longer supported as docs are automatically persisted.\n",
      "  self.vectordb.persist()\n"
     ]
    }
   ],
   "source": [
    "# update the vector db with new documents\n",
    "vectordb = local_rag_sys.update_vector_db(new_docs_path=workdir/\"new_docs_to_add\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1108dd35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contents of the documents dir: [PosixPath('/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf')]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00,  4.25it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00,  3.76it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 12 pages\n",
      "Total chunks created: 64\n",
      "[Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 0, 'page_label': '1'}, page_content='Journal of Hazardous Materials 416 (2021) 125928\\nAvailable online 24 April 2021\\n0304-3894/© 2021 The Author(s). Published by Elsevier B.V. This is an open access article under the CC BY-NC-ND license\\n(http://creativecommons.org/licenses/by-nc-nd/4.0/).\\nA marine bacterial community capable of degrading poly(ethylene \\nterephthalate) and polyethylene \\nRongrong Gao\\na , b , c , d\\n, Chaomin Sun\\na , b , d , * \\na\\nCAS and Shandong Province Key Laboratory of Experimental Marine Biology & Center of Deep Sea Research, Institute of Oceanology, Chinese Academy of Sciences, \\nQingdao, China \\nb\\nLaboratory for Marine Biology and Biotechnology, Qingdao National Laboratory for Marine Science and Technology, Qingdao, China \\nc\\nCollege of Earth Science, University of Chinese Academy of Sciences, Beijing, China \\nd\\nCenter of Ocean Mega-Science, Chinese Academy of Sciences, Qingdao, China   \\nARTICLE INFO  \\nEditor: Dr. Teresa A.P. Rocha-Santos  \\nKeywords: \\nOcean \\nPlastics \\nBacterial community reconstitution \\nDegradation \\nPollution \\nABSTRACT  \\nPlastic wastes are becoming the most common form of marine debris and present a growing global pollution'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 0, 'page_label': '1'}, page_content='ABSTRACT  \\nPlastic wastes are becoming the most common form of marine debris and present a growing global pollution \\nproblem. Here, we used a screening approach on hundreds of plastic waste-associated samples and discovered a \\nmarine bacterial community capable of efficiently colonizing and degrading both poly(ethylene terephthalate) \\n(PET) and polyethylene (PE). Using absolute quantitative 16S rRNA sequencing and cultivation methods, we \\nobtained corresponding abundance and purified cultures of three bacterial strains that mediated plastic degra -\\ndation. We further performed numerous techniques to characterize the efficient degradation of PET and PE by \\nthe reconstituted bacterial community containing these three bacteria. Additionally, we used liquid \\nchromatography-mass spectrometry to further demonstrate the degradation of PET and PE films by the recon -\\nstituted bacterial community. We conducted transcriptomic methods to investigate the plastic degradation \\nprocess and potential degradation mechanisms mediated by our reconstituted bacterial community. Lastly, we \\noverexpressed PE degradation enzymes based on transcriptomic results and verified their significant degradation'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 0, 'page_label': '1'}, page_content='overexpressed PE degradation enzymes based on transcriptomic results and verified their significant degradation \\neffects on the PE films. Overall, our study establishes a stable marine bacterial community that efficiently de -\\ngrades PET and PE and provides insights into plastic degradation pathways and their associated biological and \\nmechanistic processes — paving the way for developing microbial products against plastic wastes.   \\n1. Introduction \\nPlastics are widespread in the world ’ s oceans ( Law et al., 2010; Cozar \\net al., 2014; Jambeck et al., 2015; Lamb et al., 2018 ). It has been re -\\nported that roughly 4.8 – 12.7 million tons of plastic debris enter the \\nocean per year ( Barnes et al., 2009 ). Plastics in marine environments are \\nof increasing concern due to their persistence and effects on oceans \\n( Green et al., 2015 ), wildlife ( Li et al., 2016 ), and potentially humans \\n( Jambeck et al., 2015 ). An estimated one million birds and ten thou -\\nsands marine animals die each year as a result of plastic ingestion or by \\nbeing trapped in ocean plastics ( Green et al., 2015; Webb et al., 2013 ). \\nMoreover, after weathering, mechanical wear and ultraviolet radiation,'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 0, 'page_label': '1'}, page_content='Moreover, after weathering, mechanical wear and ultraviolet radiation, \\nlarge plastics can be broken into smaller fragments, with some smaller \\nthan 5 mm in diameter, a substance commonly defined as microplastics \\n( Hale et al., 2020 ). It is also notable that microplastics negatively impact \\nmarine biota and are ingested, accumulating within trophic webs where \\nthey even reach top predators ( Carson et al., 2013; Wang et al., 2016 ). \\nAmong the various types of plastic waste, PET and PE constitute a \\nmajority (46.5%) of the massive plastic pollution burden ( Geyer et al., \\n2017 ). PET is a type of semi-aromatic thermoplastic co-polymer resin \\nfrom the polyester family, which contains aromatic group heteroatoms \\nin their main chain ( Koshti et al., 2018 ). PE contains a carbon-carbon \\nbackbone and is highly resistant to various degradation processes, due \\nto its non-hydrolyzable covalent bonds ( Bombelli et al., 2017; Yang \\net al., 2014 ). As chemical compounds, both PET and PE have many \\nnon-biodegradable characteristics, such as long hydrocarbon chain, high \\nmolecular weight, high tensile strength, low gas permeability, and'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 0, 'page_label': '1'}, page_content='molecular weight, high tensile strength, low gas permeability, and \\ndurability to physical and chemical degradation ( Ghosh et al., 2013; \\nGewert et al., 2015; Joo et al., 2018 ). \\nLandfilling, incineration, recycling and biodegradation are the \\n* Corresponding author at: CAS and Shandong Province Key Laboratory of Experimental Marine Biology & Center of Deep Sea Research, Institute of Oceanology, \\nChinese Academy of Sciences, Qingdao, China. \\nE-mail address: sunchaomin@qdio.ac.cn (C. Sun).  \\nContents lists available at ScienceDirect \\nJournal of Hazardous Materials \\njournal homep age: www. elsevier.com/ locate/jhaz mat \\nhttps://doi.org/10.1016/j.jhazmat.2021.125928 \\nReceived 1 February 2021; Received in revised form 15 April 2021; Accepted 16 April 2021'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 1, 'page_label': '2'}, page_content='Journal of Hazardous Materials 416 (2021) 125928\\n2\\nprincipal strategies for solving the current plastic wastes problem ( Webb \\net al., 2013 ). However, perhaps the most promising and environmen -\\ntally friendly strategy for accomplishing eventual bioremediation of \\nplastic wastes lies in microorganism-mediated degradation. Microbial \\ndegradation of plastics is typically an enzymatic process catalyzing \\ncleavage of polymer bonds into monomers ( Yoshida et al., 2016; Danso \\net al., 2019 ). Thus far, PET hydrolyzing activity has been described for \\nmembers of the cutinase, lipase and esterase families ( Danso et al., \\n2019 ). PE is one of the most abundant plastics in the ocean and has clear \\npotential for degradation when incubated with specific microorganisms \\nin controlled laboratory settings ( Paco et al., 2017; Harshvardhan and \\nJha, 2013; Restrepo-Florez et al., 2014 ). However, as compared to the \\nextensive studies into PET-degrading bacteria and enzymes ( Koshti \\net al., 2018; Joo et al., 2018; Yoshida et al., 2016; Danso et al., 2018; \\nEberl et al., 2009; Brueckner et al., 2008 ), research into \\nmicroorganism-mediated PE degradation lags well behind. Further -'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 1, 'page_label': '2'}, page_content='microorganism-mediated PE degradation lags well behind. Further -\\nmore, degradation efficiencies for single strains or single enzymes are \\nstill too low to meet industrial application requirements. Alternatively, a \\nmicrobial community may be an attractive choice for degrading PE, \\nowing to the many robust functions and synergistic effects inherent \\nwithin a multiple species approach ( Pinto et al., 2020 ). New environ -\\nment bioremediation horizons for removing biodegradably-challenging \\ncompounds have been opened by constructing artificial microbial con -\\nsortia ( Qian et al., 2020 ). \\nHere, using a large-scale screening approach, we isolated and \\ndefined a marine bacterial community that efficiently degraded both \\nPET and PE. We isolated three bacterial strains capable of plastic \\ndegradation and reconstituted them to form an artificial bacterial \\ncommunity with similar plastic degradation capabilities to that of the \\nbiologically native community. Using various techniques, we further \\nclarified the degradation effects and possible released products of PET \\nand PE degradation by this reconstituted bacterial community. Lastly,'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 1, 'page_label': '2'}, page_content='and PE degradation by this reconstituted bacterial community. Lastly, \\nwe used transcriptomic methods to clarify the potential degradation \\nprocesses and enzymes associated with these processes. \\n2. Materials and methods \\n2.1. Screening microbial communities for PET and PE degradation \\nRoughly 300 surface sedimentary samples mainly consisting of \\nplastic debris were collected from the upper tidal locations in the Hui -\\nquan Bay (Qingdao, China) and kept in plastic containers at room \\ntemperatures (about 26 \\n◦\\nC) until used for screening. Three kinds of PET \\nplastic were used for degradation assays, including: plastic drinking \\nbottle, type ES301450 (0.25 mm in thickness) and type ES301005 \\n(0.0005 mm in thickness), and the latter two types of films were pur -\\nchased from the Good Fellow Company (UK). Similarly, three kinds of \\nPE plastic were used for degradation assays, including commercial PE \\nbags, type ET311350 (0.25 mm in thickness) and type ET311126 (0.025 \\nmm in thickness), the latter two types of films were also purchased from \\nthe Good Fellow Company (UK). Additive-free plastic films were pur -\\nchased from the Good Fellow Company according to the manufacturer ’ s'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 1, 'page_label': '2'}, page_content='chased from the Good Fellow Company according to the manufacturer ’ s \\nstandard (Q/SH3180014). All PET and PE films used in this study were \\ntreated with 75% ethanol and air-dried in a laminar-flow clean bench \\nprior to use. For screening microbial communities capable of degrading \\nplastics, PET or PE films were all cut into 30 mm × 20 mm square sheets \\nand incubated in the minimal medium (0.5 g yeast extract, 1 g peptone, \\nin 1 L filtered sea water, and pHis adjusted to 7.0.) containing different \\nplastics-contaminated bacterial samples. Degradation effects were \\nchecked after several weeks to months. \\n2.2. Scanning Electron Microscopy (SEM) observations \\nPET and PE films were collected from each sample and degradation \\ndefects were observed using SEM. Following medium- or bacteria \\ntreatment, plastic films were soaked in 5% glutaraldehyde for cell \\nfixation and were then dehydrated with 30 – 100% graded ethanol for 10 \\nmin each and critical-point-dried with CO\\n2 \\n( Mumtaz et al., 2010 ). Dried \\nspecimens were sputter coated for 5 min with gold and platinum (10 nm) \\nusing a Hitachi MC1000 Ion Sputter (Japan). Observation was done'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 1, 'page_label': '2'}, page_content='using a Hitachi MC1000 Ion Sputter (Japan). Observation was done \\nusing a field emission scanning electron microscope (Hitachi S-3400 N, \\nJapan) operating at an accelerating voltage of 5 kV. Medium- and bac -\\nteria treated PET and PE films were washed in ultrasonic cleaner with \\n1% SDS, distilled water, and then 75% ethanol ( Yoshida et al., 2016 ). \\nFilms were air-dried, coated with gold and platinum (10 nm) using a \\nHitachi MC1000 Ion Sputter, and subjected to SEM observation. \\n2.3. Absolute quantification of individual bacterium in the community \\nThe original bacterial community cultivated in minimal medium \\nwithout supplemental plastics served as a control group. Correspond -\\ningly, cultures supplemented with PET (type ES301450) and PE (type \\nET311350) served as experimental groups. Samples were collected in \\nfive different growth stages (stage I: OD\\n600 \\n= 0.3, cultivated for 4.5 h; \\nstage II: OD\\n600 \\n= 0.58, cultivated for 7.5 h, stage III: OD\\n600 \\n= 0.85, \\ncultivated for 8.5 h; stage IV: OD\\n600 \\n= 1.05, cultivated for 10.5 h; stage V: \\nOD\\n600 \\n= 1.13, cultivated for 7 d). Three replicate groups were used for'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 1, 'page_label': '2'}, page_content='= 1.05, cultivated for 10.5 h; stage V: \\nOD\\n600 \\n= 1.13, cultivated for 7 d). Three replicate groups were used for \\nabsolute quantification of 16 S rRNA sequences. Analyses were per -\\nformed by Genesky Biotechnologies Inc. (Shanghai, China). Briefly, total \\ngenomic DNA from different samples was extracted using the QIAamp \\nRapid DNA Kit (Qiagen, Germany). DNA concentration and integrity \\nwere measured by a NanoDrop2000 spectrophotometer (Thermo Fisher \\nScientific, USA) and agarose gel electrophoresis. Spike-in sequences \\ncontaining conserved regions identical to those of naturally-occurring \\n16S rRNA genes and artificial variable regions were distinct from \\nthose found in public nucleotide sequence databases. These sequences \\nserved as internal standards and facilitated absolute quantification \\nacross samples. Appropriate mixtures with known copy numbers of \\nspike-in sequences were added to the sample DNA mixtures. The V4-V5 \\nregions of 16S rRNA genes and spike-in sequences were amplified and \\nsequenced using an Illumina HiSeq machine. \\n2.4. Isolation and genome sequencing of three plastic degrading bacteria \\nTo isolate plastic degrading bacteria within the community, biofilms'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 1, 'page_label': '2'}, page_content='To isolate plastic degrading bacteria within the community, biofilms \\nattached to the plastic films were collected and plated on the 2216E solid \\nmedium (containing 5 g/L peptone, 1 g/L yeast extract, 1 L filtered \\nseawater, 15 g agar, pH adjusted to 7.4 – 7.6). Visible colonies were \\nfurther purified several times until they were considered to be axenic. \\nBacterial strain purity was confirmed by repeated partial sequencing of \\n16 S rRNA. Thereafter, the plastic degrading ability of each pure culture \\nwas checked by SEM, and the isolates possessing significant degrading \\neffect were chosen for further study. Genomic sequencing of bacterial \\nstrains was performed as described previously ( Zhang et al., 2020 ). \\nSequencing libraries were generated using NEBNext ® Ultra ™ DNA Li -\\nbrary Prep Kit for Illumina (NEB, USA) following manufacturer ’ s rec -\\nommendations and index codes were added to attribute sequences to \\neach sample ( Bankevich et al., 2012 ). Seven databases were used to \\npredict gene functions, including GO ( Ashburner et al., 2000 ), KEGG \\n( Kanehisa et al., 2004 ), COG ( Galperin et al., 2015 ), NR ( Saier et al.,'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 1, 'page_label': '2'}, page_content='( Kanehisa et al., 2004 ), COG ( Galperin et al., 2015 ), NR ( Saier et al., \\n2014 ) and Swiss-Prot ( Bairoch and Apweiler, 2000 ). \\n2.5. Fourier Transform Infrared (FTIR) analysis \\nPET and PE films were recovered after a four-week microbial con -\\nsortia incubation period. Films were then carefully rinsed in ultrasonic \\ncleaner with 1% SDS, distilled water, and then 75% ethanol ( Sammon \\net al., 2000; Kardas et al., 2011 ). After air drying, medium- and bacteria \\ntreated plastic films were recorded over the wavelength range of \\n450 – 4000 cm\\n\\x00 1 \\nat a resolution of 1 cm\\n\\x00 1 \\nusing a Nicolet-360 FTIR \\n(Waltham, USA) spectrometer operating in ATR mode ( Bonhomme \\net al., 2003 ). Thirty two scans were taken for each spectrum. \\nR. Gao and C. Sun'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 2, 'page_label': '3'}, page_content='Journal of Hazardous Materials 416 (2021) 125928\\n3\\n2.6. Gel Permeation Chromatography (GPC) analysis \\nThe molecular weight of medium- and bacteria treated PET films was \\ndetermined by GPC. GPC was carried out on a Shimadzu GPC-20A \\n(Japan) equipped with LC20 columns and operating at 35 \\n◦\\nC ( Shiono, \\n1979 ). Trichlorobenzene was used as a mobile phase (1 mL/min) after \\ncalibration with polystyrene standards of known molecular mass. A \\nsample concentration of 1 mg/mL was used ( Collins et al., 1991 ). The \\nmolecular weight of medium- and bacteria treated PE films was deter -\\nmined by GPC on an Agilent PL-GPC220 equipped with Agilent PLgel \\nOlexis 300 × 7.5 mm columns and operating at 150 \\n◦\\nC ( Huang et al., \\n2015 ). Trichlorobenzene was used as a mobile phase (1 mL/min) after \\ncalibration with polystyrene standards of known molecular mass. A \\nsample concentration of 1 mg/mL was used ( Bonhomme et al., 2003; \\nTung and Buckser, 1959 ). \\n2.7. X-Ray Diffraction (XRD) analysis \\nXRD was performed using a Bruker D8 Advance instrument with a \\nwavelength of 1.5406 Å of CuK α ray. The XRD tube current was set as 40 \\nmA, and the tube voltage was set as 40 kV. Measurements for PET were'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 2, 'page_label': '3'}, page_content='mA, and the tube voltage was set as 40 kV. Measurements for PET were \\nset in the angle range from 2 θ = 5\\n◦\\nto 2 θ = 45\\n◦\\nat a rate of 1\\n◦\\n/min \\n( Gaonkar et al., 2020; Farrow and Preston, 1960 ). Measurements for PE \\nwere set in the angle range from 2 θ = 3\\n◦\\nto 2 θ = 50\\n◦\\nat a rate of 1\\n◦\\n/min \\n( Xiao and Akpalu, 2007 ). \\n2.8. High-performance liquid chromatography-mass spectrometry (HPLC- \\nMS) analysis \\nThe treatment process for PET ’ s HPLC-MS samples was performed as \\nfollowing. Briefly, after 14-day incubation, 100 mL cultures containing \\nthe reconstituted bacterial community in the presence or absence of PET \\nwere collected (12,000 g , 20 min, 4 \\n◦\\nC) to obtain the supernatant. The \\nsupernatant was adjusted to pH 2.0 using 2 M HCl and extracted with 10 \\nmL ethyl acetate. The ethyl acetate layer was recovered and evaporated. \\nThe residue was dissolved in 250 L of 16 mM phosphate buffer (pH 2.5) \\n( Yoshida et al., 2016 ). HPLC-MS analysis of products released from PET \\nwas performed on an API QTRAP 5500 LCMS system equipped with a \\nXB-C18 100 A analytical column (2.1 × 250 mm, 2.6 µ m). The mobile \\nphase was methanol at a flow rate of 0.5 mL/min, and effluent was'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 2, 'page_label': '3'}, page_content='phase was methanol at a flow rate of 0.5 mL/min, and effluent was \\nmonitored at a wavelength of 240 nm. The typical elution condition was \\nfollowed at: 0 – 1 min, 10% (v/v) methanol; 1 – 2 min, 10 – 50% methanol \\nlinear gradient; 2 – 2.5 min, 50 – 95% (v/v) methanol; 4.5 – 4.6 min, \\n95 – 10% methanol linear gradient. The reaction mixture supernatant \\nwas diluted with the mobile phase toward the calibration range, acidi -\\nfied with concentrated HCl (37%) and centrifuged to remove any pre -\\ncipitation ( Yoshida et al., 2016; Barth et al., 2015 ). Standard curves of \\nTPA and BHET were prepared in a concentration range from 0.17 to 1 \\nmM and showed no significant difference. \\nThe treatment process for PE ’ s HPLC-MS samples was performed as \\nfollowing. Briefly, after 14-day incubation, 100 mL cultures containing \\nthe reconstituted bacterial community in the presence or absence of PE \\nwere collected (12,000 g , 20 min, 4 \\n◦\\nC) to obtain the supernatant. The \\nsupernatant was freeze-dried and then re-dissolved in 1 mL of 50% \\nethanol, and centrifuged again at 12,000 g for 20 min after performing \\nultra sound for 10 min. Then, 2 µ L of supernatant was used for LC-MS'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 2, 'page_label': '3'}, page_content='ultra sound for 10 min. Then, 2 µ L of supernatant was used for LC-MS \\nanalysis (Waters ACQUITY SQD, USA) ( Ren et al., 2019 ). Degradation \\nproducts were detected by LC-MS equipped with a ACQUITY UPLC BEH \\nC18 analytical column (2.1 × 50 mm, 1.7 µ m) at 50 – 1500 m/z ( Bombelli \\net al., 2017 ). During operation the column temperature was 30 \\n◦\\nC and \\nthe flow rate was 0.300 mL/min. The typical elution condition was set \\nas: 0 – 3 min, 90% water, 10% acetonitrile; 3 – 4 min, 10 – 90% acetonitrile \\nlinear gradient; 4 – 8 min, 90% (v/v) acetonitrile; 8 – 8.5 min, 90 – 10% \\nacetonitrile linear gradient; 8.5 – 10 min, 10% (v/v) acetonitrile ( Ren \\net al., 2019 ). The MS peak difference between medium- and bacteria \\ntreated traces was observed at 0.581 min. \\n2.9. Transcriptomic analyses \\nThe reconstituted bacterial community cultivated in the minimal \\nmedium without supplemental plastics was used as a control group. \\nCorrespondingly, the cultures supplemented with PET (type ES301450) \\nor PE (type ET311350) were used as experimental groups. Control or \\nexperimental samples were collected at 8 h, 7 d and 14 d, respectively.'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 2, 'page_label': '3'}, page_content='experimental samples were collected at 8 h, 7 d and 14 d, respectively. \\nTranscriptomic analyses were performed by Novogene (Tianjin, China). \\nBriefly, total RNAs from each sample were extracted using TRIzol re -\\nagent (Invitrogen, USA) with DNA contamination removed using MEGA \\nclear Kit (Life technologies, USA). Sequencing libraries were generated \\nusing NEBNext ® Ultra ™ Directional RNA Library Prep Kit for Illumina ® \\n(NEB, USA) following manufacturer ’ s recommendations and index \\ncodes were added to attribute sequences to each sample. rRNA was \\nremoved that leaves the mRNA. mRNA was purified from the total RNA \\nusing poly-T oligo-attached magnetic beads. Fragmentation was carried \\nout using divalent cations under elevated temperature in NEBNext First \\nStrand Synthesis Reaction Buffer (5 × ) (NEB, USA). Clustering of index- \\ncoded samples was performed on a cBot Cluster Generation System using \\nTruSeq PE Cluster Kit v3-cBot-HS (Illumia) according to the manufac -\\nturer ’ s instructions ( Cock et al., 2010 ). After cluster generation, library \\npreparations were sequenced on an Illumina Hiseq platform and \\npaired-end reads were generated. Raw read FASTQ data were processed'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 2, 'page_label': '3'}, page_content='paired-end reads were generated. Raw read FASTQ data were processed \\nusing in-house Perl scripts to remove reads containing adapters, reads \\ncontaining poly-N sequences, and low quality reads. Q20, Q30 and GC \\ncontent was also calculated during this step. These high-quality data \\nwere used for all downstream analyses. Differential expression analysis \\nof two conditions/groups was performed using the DESeq R package \\n(1.18.0). KOBAS software was used to test the statistical enrichment of \\ndifferentially expressed genes in KEGG pathways. GO enrichment \\nanalysis of differentially expressed genes was performed using the \\nGOseq R package, which includes gene length bias corrections. GO terms \\nwith corrected P value less than 0.05 were considered significantly \\nenriched by differential expressed genes ( Subramanian et al., 2005; \\nGarber et al., 2011 ). More details about statistical analyses for tran -\\nscriptomic data were put in the Supplementary information . \\n2.10. Expression, purification and functional assay of potential PE- \\ndegrading enzymes \\nTo test the indeed degradation function of potential enzymes pro -\\nposed by transcriptomic results, we purified three overexpressed en -'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 2, 'page_label': '3'}, page_content='posed by transcriptomic results, we purified three overexpressed en -\\nzymes in Escherichia coli and detected their degradation effects in vitro. \\nBriefly, three genes (IR195_09930, IR195_04495 and IR196_16650) \\npotentially encoding PE degradation enzymes were amplified by KOD \\nOne ™ PCR Master Mix (TOYOBO, Japan) and inserted into the 2GT \\nvector (expressing GST fused protein) ( Sun et al., 2011 ). The expression \\nvectors together with 2GT were eventually transformed into the \\nRosetta2(DE3)pLysS E. coli cell line (EMD Biosciences, USA) for protein \\nexpression. When E. coli cultures containing respective expression vector \\nwere grown to an OD600 of 0.5 – 0.6 at 37 \\n◦\\nC, isopropyl β -D-1-thioga -\\nlactopyranoside (IPTG) was added at a final concentration of 0.2 mM. \\nThen corresponding cultures were shifted to 16 \\n◦\\nC and cultured for \\nanother 12 h. Cells were harvested by centrifugation at 8000 g for 20 \\nmin, and resuspended in the appropriate lysis buffer A (20 mM HEPES, \\npH = 7.5, 300 mM KCl, 1 mM DTT,10% glycerol, 400 µ g lysozyme, 40 µ g \\nPMSF). Cells were lysed by sonication on ice and the lysate was clarified'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 2, 'page_label': '3'}, page_content='PMSF). Cells were lysed by sonication on ice and the lysate was clarified \\nby centrifugation at 12,000 g in a Beckman Ti-45 rotor for 30 min. The \\nsupernatants were filtered by polyethersulfone membranes with a pore \\nsize of 0.22 µ m (Merck Millpore, USA). The clarified lysate was loaded \\non a GSTrap HP column (GE Healthcare, USA) and the GST-tagged \\nproteins (or GST alone) were eluted on an \\n¨\\nAKTA purifier system (Agi -\\nlent, USA) by using a linear gradient with buffer B (20 mM HEPES, pH =\\n7.5, 300 mM KCl, 10 mM reduced glutathione, 1 mM DTT, 10% glyc -\\nerol). Afterwards, we checked the purity of four proteins (three GST \\nfusion proteins and GST) by using SDS-PAGE assay. The degradation \\nR. Gao and C. Sun'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 3, 'page_label': '4'}, page_content='Journal of Hazardous Materials 416 (2021) 125928\\n4\\neffects were detected in a solution containing respective protein at a \\nfinal concentration of 0.1 mg/mL and PE films (type ET311350, 0.25 \\nmm in thickness) at 30 \\n◦\\nC for 24 h. The surface morphology of PE films \\nwas observed by SEM as described above. \\n2.11. Data availability \\nComplete genome sequences of Halomonas sp. and Exiguobacterium \\nsp. have been deposited at GenBank under the accession numbers \\nCP065135 and CP065134, respectively. Complete genome sequence of \\nOchrobactrum sp. has been deposited at GenBank under the accession \\nnumbers CP064062 (chromosome 1), CP064063 (chromosome 2) and \\nCP064064 (plasmid). 16S RNA sequences used for absolute quantifica -\\ntion have been deposited at GSA database under the accession number \\nCRA003472. Raw sequencing reads for transcriptomic analysis have \\nbeen deposited at NCBI (accession numbers: PRJNA675535 and \\nPRJNA675655). \\n3. Results \\n3.1. Discovery of a marine bacterial community that efficiently degrades \\nboth PET and PE \\nTo obtain marine bacteria capable of degrading PET or PE, we \\ncollected roughly 300 plastic debris contaminated ocean sediment'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 3, 'page_label': '4'}, page_content='collected roughly 300 plastic debris contaminated ocean sediment \\nsamples from different locations in an ocean bay of China. Using these \\nsamples, we screened for microorganisms capable of using PET- \\ncontaining plastic drinking bottles or commercial PE bags as their \\nmajor carbon source. Using this approach, within one of our plastic \\ndebris samples, we derived a distinct consortium (named CAS6) capable \\nof efficiently colonizing both PET ( Fig. 1 B) and PE films ( Fig. 1 D). \\nScanning electronic microscopy (SEM) observations confirmed the \\ncolonization of PET and PE films by consortium CAS6 ( Fig. 1 F and 1 I). Of \\nnote, on the surface of PET or PE film consortium CAS6 formed a clear \\nbiofilm layer, and bacterial cells in this biofilm closely interacted each \\nother with filament-like structures ( Fig. 1 G and 1 J). When removing the \\nmicrobial film layer, we observed significant morphological changes in \\nboth PET and PE films ( Fig. 2 ) by SEM. In particular, we observed \\nnumerous heavy cracks and deep holes on the PE surface and even inside \\nthe films ( Fig. 2 F – H). \\nSince most commercial plastics contain dyes, plasticizers, antistatic'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 3, 'page_label': '4'}, page_content='the films ( Fig. 2 F – H). \\nSince most commercial plastics contain dyes, plasticizers, antistatic \\nagents and other additives, it is necessary to ensure our consortium is in \\nfact degrading plastics and not merely additives. Therefore, we repeated \\nour consortium degradation tests using PET and PE films that do not \\ncontain any additives. Consistently, both PET and PE additive-free films \\nwere clearly visibly degraded by consortium CAS6 after 4 weeks of \\ntreatment ( Fig. 3 B and 3 D). Furthermore, the four corners of both films \\nlost sharp morphology as compared to controls ( Fig. 3 A and 3 C). Similar \\nto our results using additive-containing plastics, SEM revealed efficient \\ncolonization of film surfaces by the consortium ( Fig. 3 F and 3 I) and clear \\ndegradation pits, cracks or holes on the surface and within the PET and \\nPE films ( Fig. 3 G and 3 J). We found consortium CAS6 preferred \\ndegrading pure PE ( Fig. 3 J) over pure PET ( Fig. 3 G). Overall, we have \\nidentified and characterized the functions of a bacterial consortium \\ncapable of degrading both PET and PE films, findings that warrant \\nfurther study.'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 3, 'page_label': '4'}, page_content='capable of degrading both PET and PE films, findings that warrant \\nfurther study. \\nFig. 1. Colonization of consortium CAS6 on additive-containing PET and PE films. (A) Morphology of PET films treated by the medium for 7 days. (B) Morphology of \\nPET films treated by consortium CAS6 for 7 days. (C) Morphology of PE films treated by the medium for 7 days. (D) Morphology of PE films treated by consortium \\nCAS6 for 7 days. (E) SEM observation of PET films treated by the medium for 7 days. (F, G) SEM observation of the colonization of consortium CAS6 on PET films \\nafter 7 days incubation. (H) SEM observation of PE films treated by the medium for 7 days. (I , J) SEM observation of the colonization of consortium CAS6 on PE films \\nafter 7 days incubation. \\nR. Gao and C. Sun'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 4, 'page_label': '5'}, page_content='Journal of Hazardous Materials 416 (2021) 125928\\n5\\n3.2. Isolation of key degraders and reconstitution of a functional \\ncommunity capable of plastic degradation \\nTo further characterize the composition and plastic degradation \\ndynamics of the consortium CAS6, we performed an absolute quantita -\\ntive analysis of 16 S rRNA sequences on this microbial flora. Growth \\ncurves showed it took this consortium roughly 10 h to enter stationary \\nphase growth and retained a stable population quantity after a more \\nthan 14 days of incubation in the medium supplemented with PET and \\nPE films ( Fig. 4 A). While the growth of this consortium quickly dropped \\ndown to a very low level after 7 days incubation in the medium absent of \\nplastics ( Fig. 4 A), strongly indicating bacteria within consortium CAS6 \\ncould utilize plastic as a nutrition source. We also monitored plastic \\ndegradation effects in six different growth stages by SEM. In these SEM \\nobservations we found that the thickness and density of the microbial \\nlayer on both PET ( Fig. 4 B) and PE films ( Fig. 4 C) increased with'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 4, 'page_label': '5'}, page_content='layer on both PET ( Fig. 4 B) and PE films ( Fig. 4 C) increased with \\nFig. 2. Detection of the degradation effects of consortium CAS6 on additive-containing PET and PE films. (A) SEM observation of PET films treated by the medium \\nfor 7 days. (B – - D) SEM observation of the degradation effects of consortium CAS6 on PET films after 7 days incubation. (E) SEM observation of PE films treated by the \\nmedium for 7 days. (F – H) SEM observation of the degradation effects of consortium CAS6 on PE films after 7 days incubation. \\nFig. 3. Observation of the colonization and degradation effects of consortium CAS6 on additive-free PET and PE films. (A) Morphology of PET films treated by the \\nmedium for 4 weeks. (B) Morphology of PET films treated by consortium CAS6 for 4 weeks. (C) Morphology of PE films treated by the medium for 4 weeks. (D) \\nMorphology of PE films treated by consortium CAS6 for 4 weeks. (E) SEM observation of PET films treated by the medium for 4 weeks. (F) SEM observation of the \\ncolonization of consortium CAS6 on PET films after 4 weeks incubation. (G) SEM observation of the degradation effects of PET films by consortium CAS6 after four'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 4, 'page_label': '5'}, page_content='weeks incubation. (H) SEM observation of PE films treated by the medium for 4 weeks. (I) SEM observation of the colonization of consortium CAS6 on PE films after 4 \\nweeks incubation. (J) SEM observation of the degradation effects of PE films by consortium CAS6 after 4 weeks incubation. PET films were ES301450 (0.25 mm in \\nthickness). PE films were ET311350 (0.25 mm in thickness). \\nR. Gao and C. Sun'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 5, 'page_label': '6'}, page_content='Journal of Hazardous Materials 416 (2021) 125928\\n6\\nculturing time. Obvious cracks and signs of plastic film degradation were \\nobserved after 7 ( Fig. 4 B and 4 C, stage V) and 14 days ( Fig. 4 B and 4 C, \\nstage VI) of incubation. From this, we performed an absolute quantita -\\ntive analysis of 16 S rRNA sequences of bacteria within consortium CAS6 \\nat the 7-day incubation time point. Sequencing results revealed five \\npredominant bacterial general at this time point, including: Idiomarina \\n(~50%), Marinobacter (~28%), Exiguobacterium (~18%), Halomonas \\n(~2%), Ochrobactrum (~1%) ( Fig. 5 A and Supplementary Table 1 ). To \\nobtain pure cultures of these five bacterial genera, we isolated cells \\nattached to films. Interestingly, upon further testing we found that only \\nthree bacterial isolates ( Exiguobacterium sp., Halomonas sp. , Ochrobac -\\ntrum sp.) belonging to the above five predominant genera showed sig -\\nnificant degradation of PET ( Fig. 5 B, 5 D, 5 F) and PE films ( Fig. 5 C, 5 E, \\n5 G). These results indicated Idiomarina and Marinobacter general \\nbacteria might function to help form surface biofilms, but likely do not'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 5, 'page_label': '6'}, page_content='bacteria might function to help form surface biofilms, but likely do not \\ndegrade plastics, despite their high abundance within the bacterial \\ncommunity. Notably, our SEM observations on the whole treated films \\nindicate the mixture of Exiguobacterium sp., Halomonas sp. and Ochro -\\nbactrum sp. had a greater degradation efficiency on both PET and PE \\nfilms as compared to single isolates, and the representative pictures were \\nshown in Fig. 5 H and 5 I. From this, we reconstituted a novel bacterial \\ncommunity containing these three bacteria in a 1:1:1 ratio and further \\ntested plastic degradation efficiency. \\nSurprisingly, after 2 weeks of incubation, both PET ( Fig. 6 B) and PE \\nfilms ( Fig. 6 D) were fully degraded into small pieces, with PE film being \\nheavily degraded. SEM observations further confirmed efficient coloni -\\nzation and degradation capabilities on both PET ( Fig. 6 E – -G) and PE \\nfilms ( Fig. 6 H – J). \\nFig. 4. Growth, colonization and \\nplastic-degradation of consortium CAS6 \\nin different time points. (A) Growth \\nassay of consortium CAS6 in the pres -\\nence or absence of PET and PE films in \\nsix (I to VI) different stages. (B) SEM'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 5, 'page_label': '6'}, page_content='assay of consortium CAS6 in the pres -\\nence or absence of PET and PE films in \\nsix (I to VI) different stages. (B) SEM \\nobservations of the colonization of \\nconsortium CAS6 on the surface of PET \\nfilm after cultivation for 4.5 h (I), 7.5 h \\n(II), 8.5 h (III), 10.5 h (IV), respectively. \\nSEM observations of the degradation \\neffects of consortium CAS6 on the PET \\nfilm after cultivation for 7 d (V) and 14 \\nd (VI), respectively. I to VI embedded in \\nthe figures are the same to those shown \\nin panel A. PET films used were \\nES301450 (0.25 mm in thickness). (C) \\nSEM observations of the colonization of \\nconsortium CAS6 on the surface of PE \\nfilm after cultivation for 4.5 h (I), 7.5 h \\n(II), 8.5 h (III), 10.5 h (IV), respectively. \\nSEM observations of the degradation \\neffects of consortium CAS6 on the PE \\nfilm after cultivation for 7 d (V) and 14 \\nd (VI), respectively. I to VI embedded in \\nthe figures are the same to those shown \\nin panel A. PE films used were \\nET311350 (0.25 mm in thickness).   \\nR. Gao and C. Sun'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 6, 'page_label': '7'}, page_content='Journal of Hazardous Materials 416 (2021) 125928\\n7\\n3.3. Verification of PET and PE films degradation by our reconstituted \\nbacterial community \\nWe next used multiple techniques to further verify the PET and PE \\ndegradation functions of our reconstituted bacterial community. First, \\nwe cultured our reconstituted bacterial community with PET and PE \\nfilms for four weeks and then used Fourier Transform Infrared (FTIR) \\nimaging to analyze the changes of surface chemical components and \\nfunctional groups. In bacteria-treated PET cultures, FTIR spectra showed \\na distinct peak at a wave number of 3318 cm\\n\\x00 1 \\n( Fig. 7 A), indicating \\ncarboxylic acid and alcohol functional groups (R-OH stretching, \\n3000 – 3500 cm \\n\\x00 1\\n) ( Sammon et al., 2000 ). In bacteria-treated PE cul -\\ntures, FTIR spectra showed two distinct peaks ( Fig. 7 E). One peak was \\nobserved in the vicinity of 1715 cm\\n\\x00 1\\n, indicating carbonyl bonds \\n(-C ˭ O-), while the other peak was observed at a wave number of \\n3318 cm\\n\\x00 1 \\nand was attributed to hydroxyl groups ( Bombelli et al., \\n2017 ) ( Fig. 7 E). Due to this dissociation of more key chemical bonds, we'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 6, 'page_label': '7'}, page_content='2017 ) ( Fig. 7 E). Due to this dissociation of more key chemical bonds, we \\nwere able to conclude that our reconstituted bacterial community more \\nefficiently degrades PE than PET film. Overall, our FTIR spectra data \\nsuggested a hydrolysis reaction of PET and PE films, including the for -\\nmation of carboxylic acid end groups and carbonyl bonds, leading to a \\nreduced molecular weight of the main polymer and cleavage of PET \\npolymer ester bonds and PE polymer carbon – carbon bonds. \\nWe further analyzed molecular weight distribution (MWD) changes \\nfollowing four weeks of culturing PET and PE films with our recon -\\nstituted bacterial community. Using Gel Permeation Chromatography \\n(GPC) we found the MWD of treated PET film showed a clear depoly -\\nmerization trend ( Fig. 7 C), with two peaks appearing in the curve. One \\npeak represented the range of molecular weights from \\n98,451 – 399,162 Da (accounting for 1.55%), while the other peak rep -\\nresented the range of molecular weights from 126 to 33,575 Da (ac -\\ncounting for 98.44%). In comparison, the MWD curve for medium- \\ntreated cultures had three peaks ( Fig. 7 B), representing the range of'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 6, 'page_label': '7'}, page_content='treated cultures had three peaks ( Fig. 7 B), representing the range of \\nmolecular weights from 2824 – 472,417 Da (accounting for 75%), \\n170 – 1617 Da (accounting for 10.64%) and 51 – 140 Da (accounting for \\n14.17%). Our GPC results clearly indicated treated PET film had a sig -\\nnificant reduced proportion of large molecules, with a corresponding \\nincrease in small molecules. Similarly, our MWD results for treated PE \\nfilm show molecular weights decreasing from 231,017 Da to 122,388 Da \\n( Fig. 7 F, G). We also found that bacteria-treated PET and PE films for \\nfour weeks showed a reduced relative crystallinity degree, as measured \\nby peak-differentiating and imitating calculations analyzed by X-Ray \\nDiffraction (XRD). Bacteria-treated PET film decreased from 92.55% to \\n89.85% ( Fig. 7 D), while bacteria-treated PE film decreased from 49.10% \\nto 29.50% ( Fig. 7 H). From our SEM, FTIR, GPC and XRD analyses, we \\nconclude that our reconstituted bacterial community has potential PET \\nand PE degradation capabilities. \\nWe next analyzed bacterial community-mediated degradation \\nproducts by High-performance liquid chromatography-Mass spectrom -'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 6, 'page_label': '7'}, page_content='products by High-performance liquid chromatography-Mass spectrom -\\netry (HPLC-MS). For PET, terephthalic acid (TPA), ethylene glycol (EG), \\nmono-(2-hydroxyethyl) terephthalate (MHET) and bis (2-hydroxyethyl) \\nterephthalate (BHET) have been identified as the main enzymatic \\ndegradation products ( Yoshida et al., 2016 ). Consistent with these \\nfindings, we found that after four weeks of bacteria incubation, TPA \\n( Fig. 8 A, B) and MHET ( Fig. 8 C, 8D) are the major PET hydrolysis \\nproducts. We used HPLC-MS to analyze the degradation products of PE \\nfilm following 14 days of bacteria incubation. Our results showed sig -\\nnificant differences in the abundance and categories of eluted com -\\npounds when compared to the medium-treated group ( Fig. 8 E). In \\nparticular, differences were observed at the end of the mass spectrum \\ngraph, strongly suggesting PE film was decomposed into novel chemical \\nsubstances. Unfortunately, we were unable to determine the chemical \\nidentity of these degradation compounds due to a lack of standardized \\nsamples. However, their presence supports the hypothesis that our \\nreconstituted bacterial community is capable of degrading PE film.'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 6, 'page_label': '7'}, page_content='reconstituted bacterial community is capable of degrading PE film. \\nFig. 5. Abundance quantification and plastic degradation of three core bacteria derived from consortium CAS6. (A) Absolute abundance quantification of bacteria \\nwithin consortium CAS6 using 16S rRNA sequencing, following a 7-day incubation with both PET and PE films. The top 5 most abundant general names are shown. \\nGenus abundance is indicated following each name. The genera names for the three core bacteria are shown in red. SEM observations of PET (B) and PE (C) treated \\nwith Exiguobacterium sp. for 14 days. SEM observations of PET (D) and PE (E) treated with Halomonas sp. for 14 days. SEM observations of PET (F) and PE (G) treated \\nwith Ochrobactrum sp. for 14 days. SEM observations of PET (H) and PE (I) treated with a mixture of Exiguobacterium sp., Halomonas sp. and Ochrobactrum sp. for \\n7 days. \\nR. Gao and C. Sun'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 7, 'page_label': '8'}, page_content='Journal of Hazardous Materials 416 (2021) 125928\\n8\\nFig. 6. Colonization and degradation effects of the reconstituted bacterial community on PET and PE films. (A) Morphology of PET films treated by the medium for \\n14 days. (B) Morphology of PET films treated by the reconstituted bacterial community for 14 days. (C) Morphology of PE films treated by the medium for 14 days. \\n(D) Morphology of PE films treated by the reconstituted bacterial community for 14 days. (E) SEM observation of PET films treated by the medium for 14 days. (F) \\nSEM observation of the colonization of the reconstituted bacterial community on PET films after 14 days incubation. (G) SEM observation of the degradation effects \\nof the reconstituted bacterial community on PET films after 14 days incubation. (H) SEM observation of PE films treated by the medium for 14 days. (I) SEM \\nobservation of the colonization of the reconstituted bacterial community on PE films after 14 days incubation. (J) SEM observation of the degradation effects of the \\nreconstituted bacterial community on PE films after 14 days incubation. PET films used were ES301005 (0.0005 mm in thickness). PE films used were ET311126'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 7, 'page_label': '8'}, page_content='(0.025 mm in thickness). \\nFig. 7. Validation of PET and PE degradation by the reconstituted bacteria community. (A) FTIR analysis of PET films treated by medium or reconstituted bacterial \\ncommunity for 4 weeks. (B) GPC analysis of PET films treated by medium for 4 weeks. (C) GPC analysis of reconstituted bacterial community treated PET films after 4 \\nweeks incubation. (D) XRD analysis of PET films treated by medium or reconstituted bacterial community for 4 weeks. (E) FTIR analysis of PE films treated by \\nmedium or reconstituted bacterial community for 4 weeks. (F) GPC analysis of medium-treated PE films. (G) GPC analysis of reconstituted bacterial community \\ntreated PE films after 4 weeks incubation. (H) XRD analysis of PE films treated by medium or reconstituted bacterial community for 4 weeks. PET films used were \\nES301005 (0.0005 mm in thickness). PE films used were ET311350 ET311126 (0.025 mm in thickness). \\nR. Gao and C. Sun'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 8, 'page_label': '9'}, page_content='Journal of Hazardous Materials 416 (2021) 125928\\n9\\n3.4. Transcriptomic profiling of the plastic degradation process and \\nmechanisms as carried out by our reconstituted bacterial community \\nTo explore the plastic degradation process and potential mechanisms \\nmediated by our reconstituted bacterial community, we performed a \\nmacro transcriptome analysis of this flora in the presence of PET or PE \\nfilm. Given our previous findings, we performed transcriptomic analysis \\nat 8 h, 7 d and 14 d time-points. Using differential gene expression an -\\nalyses at 8 h incubation with PET or PE film, we found genes associated \\nwith energy production and cell growth, including citrate cycle and ri -\\nbosomal biosynthesis genes, to be significantly upregulated in all three \\nbacterial species ( Exiguobacterium sp., Halomonas sp. , Ochrobactrum sp.) \\n( Fig. 9 A). Clearly, during the first 8 h of incubation, bacteria quickly \\nmultiplied, mostly utilizing the readily available nutrients contained in \\nthe original minimal medium. When extending the incubation time to 7 \\nd and 14 d, the most markedly up-regulated genes were closely related'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 8, 'page_label': '9'}, page_content='d and 14 d, the most markedly up-regulated genes were closely related \\nto biofilm formation (such as quorum sensing, bacterial chemotaxis, \\nflagellar assembling and two-component system), bacterial secretion \\nsystem, and cell growth/reproduction (such as citrate cycle, carbon \\nmetabolism, fatty acid degradation and ribosomal biosynthesis), \\nregardless of the plastic type and incubation time ( Fig. 9 B). Notably, the \\nexpressions of many putative plastic degrading enzymes such as lipases, \\nesterases, cutinases and hydrolases were evidently up-regulated ( Fig. 9 C \\nand D), suggesting their potential functions for plastic degradation. \\nTo further verify the transcriptomic results, we overexpressed three \\nputative PE degrading enzymes (including IR195_09930_esterase, \\nIR195_04495_hydrolase and IR196_16650_hydrolyase) in E. coli cells \\n( Fig. S61 ), and checked their respective degradation effects on PE film. \\nNotably, all putative enzymes showed significant degradation effects on \\nPE film ( Fig. 9 E) compared to the control (treated by GST), which \\nfurther verified the transcriptomic results and provided promising \\nenzyme candidates for future development of plastic degradation bio -'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 8, 'page_label': '9'}, page_content='enzyme candidates for future development of plastic degradation bio -\\nproducts. Taken together, our findings suggested this bacterial com -\\nmunity likely obtained energy for growth and reproduction through the \\ndegradation and utilization of plastics ( Fig. 9 ). \\n4. Discussion \\nPlastics have become a global concern as their accumulation in the \\nworld ’ s oceans has had significant impacts on marine organisms and \\nhuman health ( Cozar et al., 2014; Jambeck et al., 2015; Lamb et al., \\n2018; Ostle et al., 2019 ). Therefore, significant effort has been made to \\nreduce ocean plastic waste. To remove plastic waste and recycle \\nplastic-based materials, biocatalytic degradation could be applied as an \\nenvironmentally friendly approach ( Joo et al., 2018; Yoshida et al., \\n2016 ). Microbes have the potential to degrade plastics with ester bonds \\nvia enzymatic hydrolysis and through colonization of material surfaces \\n( Joo et al., 2018; Yoshida et al., 2016; Paco et al., 2017 ). As some marine \\nbacteria can degrade hydrocarbons, with similar chemical structures to \\nplastics, it has been suggested certain microbial communities could'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 8, 'page_label': '9'}, page_content='plastics, it has been suggested certain microbial communities could \\ncolonize and degrade plastics, using them as a carbon matrix ( Rosato \\net al., 2020; Xu et al., 2019 ). However, it remains unclear to what extent \\nFig. 8. HPLC analysis of products released from PET and PE films following treatment with the reconstituted bacterial community. (A) HPLC spectrum of the \\nstandard terephthalic acid (TPA). (B) HPLC spectrum of products (prediction as TPA) released from PET films treated with our reconstituted bacterial community for \\n4 weeks. (C) HPLC spectrum of the standard mono-(2-hydroxyethyl) terephthalate (MHET). (D) HPLC spectrum of products (prediction as MHET) released from PET \\nfilms treated with our reconstituted bacterial community for 4 weeks. (E) HPLC-MS spectrum of products released from the PE films treated with our reconstituted \\nbacterial community or medium for 2 weeks. Increased magnification of the given area as indicated by the dashed rectangle. PET films used were ES301005 \\n(0.0005 mm in thickness). PE films used were ET311126 (0.025 mm in thickness). \\nR. Gao and C. Sun'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 9, 'page_label': '10'}, page_content='Journal of Hazardous Materials 416 (2021) 125928\\n10\\nmarine prokaryotic communities are capable of degrading ocean plas -\\ntics. To our knowledge, our study is the first to investigate the com -\\nmunity structures of marine microbial biofilms formed on two different \\nplastics (PET and PE) ( Fig. 5 and Supplementary Table 1 ). Here, we \\nsuccessfully reconstituted a bacterial community specifically tailored \\ntowards plastic degradation ( Figs. 5 and 6 ). We further investigated the \\nplastic degradation process and characterized plastic degradation \\nproducts ( Figs. 8 and 9 ). In total, our study addresses whether oceanic \\nbacteria are capable of degrading plastics, and demonstrates the great \\npotential for using these bacteria to develop plastic degradation \\nbio-products. \\nWe obtained a functional bacterial community that efficiently \\ndegraded plastics from an ocean bay heavily contaminated by plastic \\nwaste. Meanwhile, we failed to identify any potential plastic-degrading \\nbacterial flora from the deep-sea sediments, suggesting that screening \\npotential plastic degraders within plastic polluted ocean locations may \\nbe a preferred approach. Similarly, several research groups have'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 9, 'page_label': '10'}, page_content='be a preferred approach. Similarly, several research groups have \\nobtained bacterial communities associated with microplastics that also \\nhave plastic-degrading functions. One such example is that of plasti -\\nsphere bacterial communities found on PET film surfaces in the North \\nSea ( Oberbeckmann et al., 2016 ). Additionally, bacterial biofilms asso -\\nciated with PE, PP and glass were found during an in-situ incubation \\nexperiment in the northern Adriatic Sea ( Pinto et al., 2020 ). However, \\nthese studies mainly focused on community structures and bacteria \\ndynamics on the plastic surface, and did not define the metabolic \\npathways or mechanistic processes involved in the biodegradation of \\nplastics. In our present study, based on the growth assay results, we \\nfound a bacterial community CAS6 could maintain a better growth \\nstatus in the medium supplemented with plastics for 7 d or longer when \\ncompared to that incubated in the medium without supplement of \\nplastics ( Fig. 4 A), suggesting the bacterial community can obtain suffi -\\ncient nutrients to support cell growth. Given the sparse carbon source \\nprovided in the original minimal medium, it is reasonable to infer that'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 9, 'page_label': '10'}, page_content='provided in the original minimal medium, it is reasonable to infer that \\nthe bacterial community colonizes, degrades and utilizes environmental \\nFig. 9. Transcriptomic analyses of plastic degradation processes and mechanisms as carried out by the reconstituted bacterial community. (A) A proposed \\ntranscriptome-based model of growth status of the reconstituted bacterial community cultured in the minimal medium supplemented with PET or PE film for 8 h. (B) \\nA proposed transcriptome-based model of plastic degradation and utilization as mediated by the reconstituted bacterial community cultured in the minimal medium \\nsupplemented with PET or PE film for 7 or 14 days. Growth and metabolic pathways proposed by the transcriptomic results. PET films used were ES301005 \\n(0.0005 mm in thickness). PE films used were ET311126 (0.025 mm in thickness). (C) A heat map showing predicted PET-degradation enzyme gene expression \\nwithin the reconstituted bacterial community. (D) A heat map showing predicted PE-degradation enzyme gene expression within the reconstituted bacterial com -'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 9, 'page_label': '10'}, page_content='munity. All the transcriptomic data associated with this Figure are shown in the Supplementary information (Supplementary Fig. 1 -60) and have been uploaded to \\nthe public database. (E) SEM observation of degradation effects of GST and GST fused enzymes (including IR195_09930_esterase, IR195_04495_hydrolase and \\nIR196_16650_hydrolyase) on PE films (type ET311350, 0.25 mm in thickness). GST: PE films treated by 0.1 mg/mL GST at 30\\n◦\\nC for 24 h; GST-IR196_16650: PE films \\ntreated by 0.1 mg/mL GST-IR196_16650 at 30\\n◦\\nC for 24 h; GST-IR195_04495: PE films treated by 0.1 mg/mL GST-IR195_04495 at 30\\n◦\\nC for 24 h; GST-IR195_09930: \\nPE films treated by 0.1 mg/mL GST-IR195_09930 at 30\\n◦\\nC for 24 h. \\nR. Gao and C. Sun'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 10, 'page_label': '11'}, page_content='Journal of Hazardous Materials 416 (2021) 125928\\n11\\nplastics as a nutrient source. To efficiently occupy plastic surfaces, the \\nbacterial community appeared to form a biofilm (Figs. 1–4), which is \\nconsistent with our observations of large amount of biofilms on plastic \\nsurfaces and the observed upregulation of genes associated with biofilm \\nformation (Fig. 9B). After successful bacterial community colonization \\nof plastic targets, bacteria might secret diverse enzymes for degrading \\nthese plastics (Fig. 9). Overall, we defined plastic degradation details by \\nthis bacterial community through transcriptomic methods, revealing \\nthree potential steps for biodegradation, including: (1) biofilm forma -\\ntion, (2) plastic degradation and (3) plastic utilization by bacterial es -\\nterases, cutinases and hydrolases (Fig. 9). Our study provides a \\ncompelling candidate biodegradation community and a framework for \\nobtaining plastic degraders from marine environments in the future. \\nIn our present study, we utilized a community approach, as a com -\\nmunity of three bacteria had greater plastic degradation function than \\nthat of any single bacterium group (Fig. 5). Indeed, the efficiency for'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 10, 'page_label': '11'}, page_content='that of any single bacterium group (Fig. 5). Indeed, the efficiency for \\nbiodegradation and bioremediation of environmental pollutants using \\nsingle bacterial strains is still very limited (Qian et al., 2020). In contrast, \\ndue to their synergistic effects, mixed floras have greater environmental \\nadaptability, higher degradation efficiencies and a broader usage scope \\nfor bioremediation (Qian et al., 2020). Given these advantages, recent \\nattention has shifted towards microbial consortia. There have also been \\nimpressive bioremediation achievements using microbial consortia, \\nsuch as the treatment of wastewater eutrophication (Mujtaba et al., \\n2017), the removal of heavy metals (Shafique et al., 2017), and the \\ndegradation of dyes (Shanmugam et al., 2017). In contrast, there are few \\nreports on the plastic degradation potential of bacterial communities, \\nespecially in the context of a reconstituted functional consortium. Our \\ncurrent study pioneers methods to advance the development of highly \\nefficient, stable and controllable synthetic microbial consortia for the \\ndegradation of plastic waste. Notably, the three bacteria within our'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 10, 'page_label': '11'}, page_content='degradation of plastic waste. Notably, the three bacteria within our \\nreconstituted bacterial community (Exiguobacterium sp., Halomonas sp. \\nand Ochrobactrum sp.) are all fast growing and easily cultured, providing \\nsignificant advantages for future applications. \\nAdditionally, our bacterial community preferred to degrade and \\nutilize PE films (Fig. 6D), demonstrating a great potential for future PE \\nbioremediation. PE is largely utilized in packaging, representing roughly \\n40% of the total demand for plastic products worldwide (www.plasticse \\nurope.org), with over a trillion plastic bags used every year (Bombelli \\net al., 2017; Yang et al., 2014). PE is comprised of a linear and highly \\nstable carbon-carbon (C-C) backbone, making it resistant to degradation \\nand creating a significant challenge for plastic waste biodegradation. PE \\nbiodegradation has been observed following extremely long incubation \\ntimes (multiple months) and only under certain conditions. For example, \\nmodest degradation of PE was observed after a combination of nitric \\nacid treatment and a 3-month incubation with the fungus Penicillium \\nsimplicissimum (Yamada-Onodera et al., 2001). Even slower PE degra -'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 10, 'page_label': '11'}, page_content='simplicissimum (Yamada-Onodera et al., 2001). Even slower PE degra -\\ndation was recorded after 4–7 months of exposure to the bacterium \\nNocardia asteroids (Bonhomme et al., 2003). Excitingly, our recon -\\nstituted bacterial community brought about significant PE degradation \\n(e.g. ubiquitous cracks on the surface and damage to the four PE film \\ncorners) within several days, and near total PE degradation within 2 \\nweeks (e.g. PE film degraded into very small pieces). In particular, the \\npotential PE degradation enzymes showed significant degradation ef -\\nfects within 24 h (Fig. 9E). Given the rapid rate of biodegradation re -\\nported here, our findings undoubtedly have significant potential \\nbiotechnological applications. Further investigation will be necessary to \\ndetermine the detailed nature of PE degradation products by our \\nreconstituted bacterial community as well as associated enzymes, and \\nimprove the degradation effects by optimizing the community compo -\\nsition and enzyme activity. \\nAuthor contributions \\nR.G. and C.S. conceived and designed the study. R.G. performed all \\nthe experiments. R.G. and C.S. analyzed the data. R.G. wrote the'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 10, 'page_label': '11'}, page_content='the experiments. R.G. and C.S. analyzed the data. R.G. wrote the \\nmanuscript. C.S. revised the manuscript. All authors read and approved \\nthe final manuscript. \\nDeclaration of Competing Interest \\nThe authors have no conflict of interest. \\nAcknowledgements \\nThis work is funded by the China Ocean Mineral Resources R&D \\nAssociation Grant (Grant No. DY135-B2–14), the Major Research Plan of \\nthe National Natural Science Foundation (Grant No. 92051107), Key \\nDeployment Projects of Center of Ocean Mega-Science of the Chinese \\nAcademy of Sciences (Grant No. COMS2020Q04), National Key R and D \\nProgram of China (Grant No. 2018YFC0310800), Strategic Priority \\nResearch Program of the Chinese Academy of Sciences (Grant No. \\nXDA22050301), the Taishan Young Scholar Program of Shandong \\nProvince (tsqn20161051), and Qingdao Innovation Leadership Program \\n(Grant No. 18–1-2–7-zhc) for Chaomin Sun. \\nAppendix A. Supporting information \\nSupplementary data associated with this article can be found in the \\nonline version at doi:10.1016/j.jhazmat.2021.125928. \\nReferences \\nAshburner, M., Ball, C.A., Blake, J.A., Botstein, D., Butler, H., Cherry, J.M., Davis, A.P.,'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 10, 'page_label': '11'}, page_content='References \\nAshburner, M., Ball, C.A., Blake, J.A., Botstein, D., Butler, H., Cherry, J.M., Davis, A.P., \\nDolinski, K., Dwight, S.S., Eppig, J.T., Harris, M.A., Hill, D.P., Issel-Tarver, L., \\nKasarskis, A., Lewis, S., Matese, J.C., Richardson, J.E., Ringwald, M., Rubin, G.M., \\nSherlock, G., Consortium, G.O., 2000. Gene ontology: tool for the unification of \\nbiology. Nat. Genet. 25, 25–29. \\nBairoch, A., Apweiler, R., 2000. The SWISS-PROT protein sequence database and its \\nsupplement TrEMBL in 2000. Nucleic Acids Res. 28, 45–48. \\nBankevich, A., Nurk, S., Antipov, D., Gurevich, A.A., Dvorkin, M., Kulikov, A.S., Lesin, V. \\nM., Nikolenko, S.I., Pham, S., Prjibelski, A.D., Pyshkin, A.V., Sirotkin, A.V., \\nVyahhi, N., Tesler, G., Alekseyev, M.A., Pevzner, P.A., 2012. SPAdes: a new genome \\nassembly algorithm and its applications to single-cell sequencing. J. Comput. Biol. \\n19, 455–477. \\nBarnes, D.K., Galgani, F., Thompson, R.C., Barlaz, M., 2009. Accumulation and \\nfragmentation of plastic debris in global environments. Philos. Trans. R. Soc. Lond. B \\nBiol. Sci. 364, 1985–1998. \\nBarth, M., Oeser, T., Wei, R., Then, J., Schmidt, J., Zimmermann, W., 2015. Effect of'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 10, 'page_label': '11'}, page_content='Biol. Sci. 364, 1985–1998. \\nBarth, M., Oeser, T., Wei, R., Then, J., Schmidt, J., Zimmermann, W., 2015. Effect of \\nhydrolysis products on the enzymatic degradation of polyethylene terephthalate \\nnanoparticles by a polyester hydrolase from Thermobifida fusca. Biochem. Eng. J. \\n93, 222–228. \\nBombelli, P., Howe, C.J., Bertocchini, F., 2017. Polyethylene bio-degradation by \\ncaterpillars of the wax moth Galleria mellonella. Curr. Biol. 27, R292–R293. \\nBonhomme, S., Cuer, A., Delort, A.M., Lemaire, J., Sancelme, M., Scott, G., 2003. \\nEnvironmental biodegradation of polyethylene. Polym. Degrad. Stabil. 81, 441–452. \\nBrueckner, T., Eberl, A., Heumann, S., Rabe, M., Guebitz, G.M., 2008. Enzymatic and \\nchemical hydrolysis of poly(ethylene terephthalate) fabrics. J. Polym. Sci. Pol. \\nChem. 46, 6435–6443. \\nCarson, H.S., Nerheim, M.S., Carroll, K.A., Eriksen, M., 2013. The plastic-associated \\nmicroorganisms of the North Pacific Gyre. Mar. Pollut. Bull. 75, 126–132. \\nCock, P.J.A., Fields, C.J., Goto, N., Heuer, M.L., Rice, P.M., 2010. The Sanger FASTQ file \\nformat for sequences with quality scores, and the Solexa/Illumina FASTQ variants. \\nNucleic Acids Res. 38, 1767–1771.'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 10, 'page_label': '11'}, page_content='format for sequences with quality scores, and the Solexa/Illumina FASTQ variants. \\nNucleic Acids Res. 38, 1767–1771. \\nCollins, M.J., Zeronian, S.H., Marshall, M.L., 1991. Analysis of the molecular-weight \\ndistributions of aminolyzed poly(ethylene-terephthalate) by using gel-permeation \\nchromatography. J. Macromol. Sci. Chem. A28, 775–792. \\nCozar, A., Echevarria, F., Gonzalez-Gordillo, J.I., Irigoien, X., Ubeda, B., Hernandez- \\nLeon, S., Palma, A.T., Navarro, S., Garcia-de-Lomas, J., Ruiz, A., Fernandez-de- \\nPuelles, M.L., Duarte, C.M., 2014. Plastic debris in the open ocean. Proc. Natl. Acad. \\nSci. USA 111, 10239–10244. \\nDanso, D., Chow, J., Streit, W.R., 2019. Plastics: environmental and biotechnological \\nperspectives on microbial degradation. Appl. Environ. Microbiol. 85. \\nDanso, D., Schmeisser, C., Chow, J., Zimmermann, W., Wei, R., Leggewie, C., Li, X.Z., \\nHazen, T., Streit, W.R., 2018. New insights into the function and global distribution \\nof polyethylene terephthalate (PET)-degrading bacteria and enzymes in marine and \\nterrestrial metagenomes. Appl. Environ. Microbiol. 84. \\nEberl, A., Heumann, S., Bruckner, T., Araujo, R., Cavaco-Paulo, A., Kaufmann, F.,'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 10, 'page_label': '11'}, page_content='Eberl, A., Heumann, S., Bruckner, T., Araujo, R., Cavaco-Paulo, A., Kaufmann, F., \\nKroutil, W., Guebitz, G.M., 2009. Enzymatic surface hydrolysis of poly(ethylene \\nterephthalate) and bis(benzoyloxyethyl) terephthalate by lipase and cutinase in the \\npresence of surface active molecules. J. Biotechnol. 143, 207–212. \\nFarrow, G., Preston, D., 1960. Measurement of crystallinity in drawn polyethylene \\nterephthalate fibres by X-Ray diffraction. Br. J. Appl. Phys. 11, 353–358. \\nR. Gao and C. Sun'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 11, 'page_label': '12'}, page_content='Journal of Hazardous Materials 416 (2021) 125928\\n12\\nGalperin, M.Y., Makarova, K.S., Wolf, Y.I., Koonin, E.V., 2015. Expanded microbial \\ngenome coverage and improved protein family annotation in the COG database. \\nNucleic Acids Res. 43, D261–D269. \\nGaonkar, A.A., Murudkar, V.V., Deshpande, V.D., 2020. Comparison of crystallization \\nkinetics of polyethylene terephthalate (PET) and reorganized PET. Thermochim. \\nActa 683, 178472. \\nGarber, M., Grabherr, M.G., Guttman, M., Trapnell, C., 2011. Computational methods for \\ntranscriptome annotation and quantification using RNA-seq. Nat. Methods 8, \\n469–477. \\nGewert, B., Plassmann, M.M., MacLeod, M., 2015. Pathways for degradation of plastic \\npolymers floating in the marine environment. Environ. Sci. Proc. Imp. 17, \\n1513–1521. \\nGeyer, R., Jambeck, J.R., Law, K.L., 2017. Production, use, and fate of all plastics ever \\nmade. Sci. Adv. 3, 1700782. \\nGhosh, S.K., Pal, S., Ray, S., 2013. Study of microbes having potentiality for \\nbiodegradation of plastics. Environ. Sci. Pollut. R. 20, 4339–4355. \\nGreen, D.S., Boots, B., Blockley, D.J., Rocha, C., Thompson, R., 2015. Impacts of'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 11, 'page_label': '12'}, page_content='Green, D.S., Boots, B., Blockley, D.J., Rocha, C., Thompson, R., 2015. Impacts of \\ndiscarded plastic bags on marine assemblages and ecosystem functioning. Environ. \\nSci. Technol. 49, 5380–5389. \\nHale, R.C., Seeley, M.E., La Guardia, M.J., Mai, L., Zeng, E.Y., 2020. A global perspective \\non microplastics. J. Geophys. Res.: Oceans 125. \\nHarshvardhan, K., Jha, B., 2013. Biodegradation of low-density polyethylene by marine \\nbacteria from pelagic waters, Arabian Sea, India. Mar. Pollut. Bull. 77, 100–106. \\nHuang, H.H., Guo, M.F., Wei, D., Suarez, I., Coto, B., Lopez, E., Ortin, A., Yau, W.W., \\n2015. Direct comparison of IR and DRI detector for HT-GPC of polyolefins. \\nMacromol. Symp. 356, 95–109. \\nJambeck, J.R., Geyer, R., Wilcox, C., Siegler, T.R., Perryman, M., Andrady, A., \\nNarayan, R., Law, K.L., 2015. Plastic waste inputs from land into the ocean. Science \\n347, 768–771. \\nJoo, S., Cho, I.J., Seo, H., Son, H.F., Sagong, H.Y., Shin, T.J., Choi, S.Y., Lee, S.Y., Kim, K. \\nJ., 2018. Structural insight into molecular mechanism of poly (ethylene \\nterephthalate) degradation. Nat. Commun. 9, 382. \\nKanehisa, M., Goto, S., Kawashima, S., Okuno, Y., Hattori, M., 2004. The KEGG resource'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 11, 'page_label': '12'}, page_content='Kanehisa, M., Goto, S., Kawashima, S., Okuno, Y., Hattori, M., 2004. The KEGG resource \\nfor deciphering the genome. Nucleic Acids Res. 32, D277–D280. \\nKardas, I., Lipp-Symonowicz, B., Sztajnowski, S., 2011. The influence of enzymatic \\ntreatment on the surface modification of PET fibers. J. Appl. Polym. Sci. 119, \\n3117–3126. \\nKoshti, R., Mehta, L., Samarth, N., 2018. Biological recycling of Polyethylene \\nTerephthalate: a mini-review. J. Polym. Environ. 26, 3520–3529. \\nLamb, J.B., Willis, B.L., Fiorenza, E.A., Couch, C.S., Howard, R., Rader, D.N., True, J.D., \\nKelly, L.A., Ahmad, A., Jompa, J., Harvell, C.D., 2018. Plastic waste associated with \\ndisease on coral reefs. Science 359, 460–462. \\nLaw, K.L., Moret-Ferguson, S., Maximenko, N.A., Proskurowski, G., Peacock, E.E., \\nHafner, J., Reddy, C.M., 2010. Plastic accumulation in the North Atlantic subtropical \\ngyre. Science 329, 1185–1188. \\nLi, W.C., Tse, H.F., Fok, L., 2016. Plastic waste in the marine environment: a review of \\nsources, occurrence and effects. Sci. Total Environ. 566, 333–349. \\nMujtaba, G., Rizwan, M., Lee, K., 2017. Removal of nutrients and COD from wastewater'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 11, 'page_label': '12'}, page_content='Mujtaba, G., Rizwan, M., Lee, K., 2017. Removal of nutrients and COD from wastewater \\nusing symbiotic co-culture of bacterium Pseudomonas putida and immobilized \\nmicroalga Chlorella vulgaris. J. Ind. Eng. Chem. 49, 145–151. \\nMumtaz, T., Khan, M.R., Hassan, M.A., 2010. Study of environmental biodegradation of \\nLDPE films in soil using optical and scanning electron microscopy. Micron 41, \\n430–438. \\nOberbeckmann, S., Osborn, A.M., Duhaime, M.B., 2016. Microbes on a bottle: substrate, \\nseason and geography influence community composition of microbes colonizing \\nmarine plastic debris. Plos One 11, 0159289. \\nOstle, C., Thompson, R.C., Broughton, D., Gregory, L., Wootton, M., Johns, D.G., 2019. \\nThe rise in ocean plastics evidenced from a 60-year time series. Nat. Commun. 10, \\n1622. \\nPaco, A., Duarte, K., da Costa, J.P., Santos, P.S.M., Pereira, R., Pereira, M.E., Freitas, A.C., \\nDuarte, A.C., Rocha-Santos, T.A.P., 2017. Biodegradation of polyethylene \\nmicroplastics by the marine fungus Zalerion maritimum. Sci. Total Environ. 586, \\n10–15. \\nPinto, M., Zenner, P.P., Langer, T.M., Harrison, J., Simon, M., Varela, M.M., Herndl, G.J.,'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 11, 'page_label': '12'}, page_content='10–15. \\nPinto, M., Zenner, P.P., Langer, T.M., Harrison, J., Simon, M., Varela, M.M., Herndl, G.J., \\n2020. Putative degraders of low-density polyethylene-derived compounds are \\nubiquitous members of plastic-associated bacterial communities in the marine \\nenvironment. Environ. Microbiol. 22, 4779–4793. \\nQian, X.J., Chen, L., Sui, Y., Chen, C., Zhang, W.M., Zhou, J., Dong, W.L., Jiang, M., \\nXin, F.X., Ochsenreither, K., 2020. Biotechnological potential and applications of \\nmicrobial consortia. Biotechnol. Adv. 40, 107500. \\nRen, L., Men, L.N., Zhang, Z.W., Guan, F.F., Tian, J., Wang, B., Wang, J.H., Zhang, Y.H., \\nZhang, W., 2019. Biodegradation of polyethylene by Enterobacter sp. D1 from the \\nguts of wax moth Galleria mellonella. Int. J. Environ. Res. Public Health 16, 1941. \\nRestrepo-Florez, J.M., Bassi, A., Thompson, M.R., 2014. Microbial degradation and \\ndeterioration of polyethylene - a review. Int. Biodeter. Biodegr. 88, 83–90. \\nRosato, A., Barone, M., Negroni, A., Brigidi, P., Fava, F., Xu, P., Candela, M., Zanaroli, G., \\n2020. Microbial colonization of different microplastic types and biotransformation'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 11, 'page_label': '12'}, page_content='2020. Microbial colonization of different microplastic types and biotransformation \\nof sorbed PCBs by a marine anaerobic bacterial community. Sci. Total Environ. 705, \\n135790. \\nSaier, M.H., Reddy, V.S., Tamang, D.G., Vastermark, A., 2014. The transporter \\nclassification database. Nucleic Acids Res. 42, D251–D258. \\nSammon, C., Yarwood, J., Everall, N., 2000. A FTIR-ATR study of liquid diffusion \\nprocesses in PET films: comparison of water with simple alcohols. Polymer 41, \\n2521–2534. \\nShafique, M., Jawaid, A., Rehman, Y., 2017. As(V) reduction, As(III) oxidation, and Cr \\n(VI) reduction by multi-metal-resistant Bacillus subtilis, Bacillus safensis, and Bacillus \\ncereus species isolated from wastewater treatment plant. Geomicrobiol. J. 34, \\n687–694. \\nShanmugam, B.K., Easwaran, S.N., Lakra, R., Deepa, P.R., Mahadevan, S., 2017. \\nMetabolic pathway and role of individual species in the bacterial consortium for \\nbiodegradation of azo dye: a biocalorimetric investigation. Chemosphere 188, \\n81–89. \\nShiono, S., 1979. Separation and identification of poly(ethylene-terephthalate) oligomers \\nby gel-permeation chromatography. J. Polym. Sci. Pol. Chem. 17, 4123–4127.'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 11, 'page_label': '12'}, page_content='by gel-permeation chromatography. J. Polym. Sci. Pol. Chem. 17, 4123–4127. \\nSubramanian, A., Tamayo, P., Mootha, V.K., Mukherjee, S., Ebert, B.L., Gillette, M.A., \\nPaulovich, A., Pomeroy, S.L., Golub, T.R., Lander, E.S., Mesirov, J.P., 2005. Gene set \\nenrichment analysis: A knowledge-based approach for interpreting genome-wide \\nexpression profiles. Proc. Natl. Acad. Sci. USA 102, 15545–15550. \\nSun, C.M., Todorovic, A., Querol-Audi, J., Bai, Y., Villa, N., Snyder, M., Ashchyan, J., \\nLewis, C.S., Hartland, A., Gradia, S., Fraser, C.S., Doudna, J.A., Nogales, E., Cate, J. \\nH.D., 2011. Functional reconstitution of human eukaryotic translation initiation \\nfactor 3 (eIF3). Proc. Natl. Acad. Sci. USA 108, 20473–20478. \\nTung, L.H., Buckser, S., 1959. The effect of molecular weight on the crystallinity of \\npolyethylene. J. Phys. Chem. 62, 1530–1534. \\nWang, J.D., Tan, Z., Peng, J.P., Qiu, Q.X., Li, M.M., 2016. The behaviors of microplastics \\nin the marine environment. Mar. Environ. Res. 113, 7–17. \\nWebb, H.K., Arnott, J., Crawford, R.J., Ivanova, E.P., 2013. Plastic degradation and its \\nenvironmental implications with special reference to Poly(ethylene terephthalate). \\nPolymers 5, 1–18.'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 11, 'page_label': '12'}, page_content='environmental implications with special reference to Poly(ethylene terephthalate). \\nPolymers 5, 1–18. \\nXiao, Z.C., Akpalu, Y.A., 2007. New insights into the characteristics of early stage \\ncrystallization of a polyethylene. Polymer 48, 5388–5397. \\nXu, X.Y., Wang, S., Gao, F.L., Li, J.X., Zheng, L., Sun, C.J., He, C.F., Wang, Z.X., Qu, L.Y., \\n2019. Marine microplastic-associated bacterial community succession in response to \\ngeography, exposure time, and plastic type in China’s coastal seawaters. Mar. Pollut. \\nBull. 145, 278–286. \\nYamada-Onodera, K., Mukumoto, H., Katsuyaya, Y., Saiganji, A., Tani, Y., 2001. \\nDegradation of polyethylene by a fungus, Penicillium simplicissimum YK. Polym. \\nDegrad. Stabil. 72, 323–327. \\nYang, J., Yang, Y., Wu, W.M., Zhao, J., Jiang, L., 2014. Evidence of polyethylene \\nbiodegradation by bacterial strains from the guts of plastic-eating waxworms. \\nEnviron. Sci. Technol. 48, 13776–13784. \\nYoshida, S., Hiraga, K., Takehana, T., Taniguchi, I., Yamaji, H., Maeda, Y., Toyohara, K., \\nMiyamoto, K., Kimura, Y., Oda, K., 2016. A bacterium that degrades and assimilates \\npoly(ethylene terephthalate). Science 351, 1196–1199.'), Document(metadata={'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creator': 'Elsevier', 'creationdate': '2021-06-12T16:00:41+00:00', 'moddate': '2021-06-12T16:49:48+00:00', 'title': 'A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene', 'keywords': 'Ocean,Plastics,Bacterial community reconstitution,Degradation,Pollution', 'subject': 'Journal of Hazardous Materials, 416 (2021) 125928. doi:10.1016/j.jhazmat.2021.125928', 'author': 'Rongrong Gao', 'source': '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf', 'total_pages': 12, 'page': 11, 'page_label': '12'}, page_content='poly(ethylene terephthalate). Science 351, 1196–1199. \\nZhang, J., Liu, R., Xi, S.C., Cai, R.N., Zhang, X., Sun, C.M., 2020. A novel bacterial \\nthiosulfate oxidation pathway provides a new clue about the formation of zero- \\nvalent sulfur in deep sea. ISME J. 14, 2261–2274. \\nR. Gao and C. Sun')]\n",
      "\n",
      "Database updated with new documents\n"
     ]
    }
   ],
   "source": [
    "# update the vector db with new documents\n",
    "vectordb = local_rag_sys.update_vector_db(new_docs_path=workdir/\"new_pdfs_to_add\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b74d3637",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Polyethylene Degradation by a Rhodococcous Strain Isolated from Naturally Weathered Plastic Waste Enrichment .pdf', '/Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx', '/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove.pdf', '/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation.pdf', '/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf', '/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Using artificial intelligence to document the hidden RNA virosphere.pdf', '/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Diffusion Sequence Models for Enhanced Protein Representation and Generation.pdf', '/Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/xTrimoPGLM- unified 100-billion-parameter pretrained transformer for deciphering the language of proteins.pdf', '/Users/danid/Downloads/RAG_personal_knowledge_base/new_pdfs_to_add/A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene.pdf'}\n"
     ]
    }
   ],
   "source": [
    "# List existing documents in the vector store\n",
    "docs_in_db = local_rag_sys.list_existing_documents()\n",
    "docs = []\n",
    "for doc in docs_in_db:\n",
    "    docs.append(doc['source'])\n",
    "print(set(docs))  # Display unique document titles in the vector store"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e673d0b",
   "metadata": {},
   "source": [
    "### Single Prompts Chain (not conversation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "61d41e64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No split documents available for BM25Retriever creation. Need to build vector DB first.\n"
     ]
    }
   ],
   "source": [
    "# Complete workflow with existing documents vector db, running QA-chain with a query to the model\n",
    "# Single prompts chain\n",
    "qa_chain = local_rag_sys.build_qa_chain(retriever_type=\"vec_semantic\", \n",
    "                                        use_conversation=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "642937c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "result = qa_chain.invoke('''How does the SaProt protein-LLM differs from other protein-LLMs such as ESM-2? \n",
    "                         Two new protein LLMs models called: Diffusion Sequence Model (DSM) and xTrimoPGLM are suppose to be state-of-the-art. Compare them to SaProt. \n",
    "                        What are the key architectural novelties of these models?\n",
    "                        Use the given context (source documents) to answer the question.\n",
    "                        Give the name of the source documents and metadata that supports your answer.''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "4c9d84aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== ANSWER ===\n",
      "**What makes SaProt different from other protein‑LLMs**\n",
      "\n",
      "| Feature | SaProt (the “structure‑aware” LLM) | ESM‑2 (baseline LLM) |\n",
      "|---------|-----------------------------------|----------------------|\n",
      "| **Core novelty** | Introduces a *Structure‑Aware (SA) vocabulary* that concatenates the usual residue token with a structure token derived from Foldseek‑encoded 3‑D information. The model therefore sees “SA‑tokens” instead of pure amino‑acid letters. | Pure sequence‑only tokenisation; no explicit structural token. |\n",
      "| **Back‑bone** | Uses the exact same transformer architecture, size and pre‑training data as ESM‑2 (≈650 M parameters), but the input embedding layer is enlarged to accommodate the extra structure tokens. | Standard ESM‑2 transformer. |\n",
      "| **Training data** | ~40 M protein sequences *paired* with predicted structures (Foldseek/AlphaFold). This is far larger in terms of structural examples than any previous PLM. | Only raw amino‑acid sequences; no paired structures. |\n",
      "| **Resulting behaviour** | Consistently higher zero‑shot and fine‑tuned performance on 10 protein‑function tasks (Thermostability, HumanPPI, Metal‑Ion Binding, DeepLoc, etc.) and markedly better contact‑map prediction, showing that SaProt encodes more accurate structural information. | Good baseline but out‑performed by SaProt on all the same benchmarks. |\n",
      "| **Source document** | *“SAPROT: PROTEIN LANGUAGE MODELING WITH STRUCTURE‑AWARE VOCABULARY”* (Westlake University, arXiv 2025) – sections “Abstract”, “4.2.3 Results”, Table 1/2 and the discussion of the SA‑token. | Same document (comparison statements) and the background description of ESM‑2 in the same paper. |\n",
      "\n",
      "---\n",
      "\n",
      "### Diffusion Sequence Model (DSM)\n",
      "\n",
      "| Architectural novelties |\n",
      "|--------------------------|\n",
      "| **Diffusion‑based pre‑training** – DSM adds a *masked forward diffusion process* on top of the standard MLM objective. Tokens are progressively corrupted and then denoised, which teaches the model to reconstruct sequences even at very high mask rates (up to 90 %). |\n",
      "| **New language‑modeling head** – beyond the usual ESM‑2 token classifier, DSM introduces a diffusion head that predicts the original residues from noisy inputs. |\n",
      "| **Generative capability** – because the diffusion process is invertible, DSM can *sample* complete protein sequences in one forward pass, something pure MLM models cannot do. |\n",
      "| **Back‑bone reuse** – The underlying transformer architecture is taken directly from ESM‑2 (same layer count and hidden size), so DSM inherits the strong representation power of ESM‑2 while adding diffusion. |\n",
      "| **Performance notes** – DSM‑650 achieved the highest F1 scores among similarly sized PLMs on a broad suite of tasks, only slightly below the much larger ProtT5; it also showed a clear advantage over autoregressive (ProtCLM) and discrete‑diffusion (DPLM) baselines. |\n",
      "| **Source document** | *“Diffusion Sequence Models for Enhanced Protein Representation and Generation”* (arXiv 2025‑06‑09, authors Hallee et al.) – sections “4.4 DSM can be fine‑tuned…”, “4.3 DSM produced high‑quality embeddings…” and Table 3/Figure 4 describing the diffusion training objective. |\n",
      "\n",
      "---\n",
      "\n",
      "### xTrimoPGLM\n",
      "\n",
      "| Architectural novelties |\n",
      "|--------------------------|\n",
      "| **Two‑stage pre‑training** – First a massive MLM phase (≈400 B tokens) for rapid “warm‑up”, then a second phase mixing *20 % MLM* with *80 % General‑Linear‑Model (GLM)* objectives. The GLM component supplies an autoregressive generative head while retaining bidirectional context. |\n",
      "| **Scale** – 100 B parameters, trained on ≈1 trillion tokens, making it orders of magnitude larger than SaProt (650 M) or DSM (≈650 M). |\n",
      "| **Evoformer‑style modules** – The model contains *single‑representation* and *pair‑representation* streams, a recycling module (run three times), and a folding module reminiscent of AlphaFold’s Evoformer. This explicit pairwise processing is absent from SaProt/ESM‑2. |\n",
      "| **MSA‑free but multi‑modal** – Although it does not use multiple sequence alignments, the architecture can ingest residue sequences together with optional structural cues (through the pair stream) and learns to predict 3‑D geometry directly. |\n",
      "| **Benchmark superiority** – Outperforms all prior baselines on 18 protein‑understanding tasks across four categories, and attains lower perplexity on CAMEO/CASP15 than ESMFold’s PLM. |\n",
      "| **Source document** | The later part of the provided context describing *xTrimoPGLM* (sections “Evaluation of protein understanding benchmarks”, the diagram with “Input sequence → … Evoformer Structure module → Recycling module → Folding module”, and the performance summary). No separate citation line is given, but it appears in the same multi‑document dump together with the DSM paper. |\n",
      "\n",
      "---\n",
      "\n",
      "## Comparative Summary\n",
      "\n",
      "| Model | Core architectural idea | How it diverges from SaProt |\n",
      "|-------|------------------------|-----------------------------|\n",
      "| **SaProt** | Adds explicit structure tokens (SA‑vocab) to a standard ESM‑2 transformer. | Baseline for comparison – no diffusion, no GLM head, far smaller than xTrimoPGLM. |\n",
      "| **DSM** | Keeps the same transformer but replaces the MLM loss with a *masked diffusion* objective and adds a diffusion decoding head, enabling one‑step generative sampling. | Diffusion training vs. SaProt’s pure masked‑language‑modeling; no structure tokens (still uses ESM‑2 backbone). |\n",
      "| **xTrimoPGLM** | Massive 100 B‑parameter model with dual MLM/GLM pre‑training, Evoformer‑style single/pair streams, recycling and folding modules. | Introduces pairwise processing and a generative GLM head; scale is >150× SaProt; no explicit SA‑vocab but learns structure implicitly via the pair stream. |\n",
      "\n",
      "All three newer models aim to improve on the ESM‑2 baseline (which SaProt also improves upon) but take **different architectural routes**:\n",
      "\n",
      "* **SaProt → structural token enrichment**  \n",
      "* **DSM → diffusion‑based denoising and generation**  \n",
      "* **xTrimoPGLM → ultra‑large, dual‑objective, Evoformer‑inspired architecture**\n",
      "\n",
      "These differences are directly supported by the cited source documents.\n",
      "\n",
      "=== SOURCES ===\n",
      "Content: • SaProt outperforms ESM-2 in all protein-level tasks. Specifically, SaProt shows remark-\n",
      "able enhancements over ESM-2 in the Thermostability, HumanPPI, Metal Ion Binding, and\n",
      "DeepLoc tasks. This outc...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 7\n",
      "\n",
      "Content: arXiv:2506.08293v1  [q-bio.BM]  9 Jun 2025\n",
      "Diffusion Sequence Models for Enhanced Protein Representation and\n",
      "Generation\n",
      "Logan Hallee1, 2, Nikolaos Rafailidis1, David B. Bichara2, and Jason P. Gleghorn...\n",
      "Metadata title: Diffusion Sequence Models for Enhanced Protein Representation and Generation\n",
      "\n",
      "Metadata page number: 0\n",
      "\n",
      "Content: embeddings and valuable downstream tasks (datasets described in Supplemental Figure S3). DSM650 produced the highest\n",
      "quality embeddings among similarly sized pLMs, generating consistently high F1 scor...\n",
      "Metadata title: Diffusion Sequence Models for Enhanced Protein Representation and Generation\n",
      "\n",
      "Metadata page number: 6\n",
      "\n",
      "Content: ESM-2, given that residue-based PLMs also implicitly contain structure information (Rao et al.,\n",
      "2020)? To answer it, we conduct an additional structure prediction task, namely contact map predic-\n",
      "tion...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 7\n",
      "\n",
      "Content: tion 4.1.3. Its impressive performance positions it as a compelling alternative to the ESM family.\n",
      "5 A NALYSIS\n",
      "We conduct insightful ablation studies by dissecting SaProt. One can find more analysis i...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 7\n",
      "\n",
      "Content: Model ClinVar ProteinGym\n",
      "w/o MSA retrieval w/ MSA retrieval\n",
      "AUC Spearman’s ρ Spearman’s ρ\n",
      "ESM-2 (15B) 0.843 0.436 0.479\n",
      "SaProt 0.909 0.478 0.489\n",
      "Table 6: Zero-shot comparison with ESM-2 15B version. C...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 16\n",
      "\n",
      "Content: process. DSM extends ESM2 ( 33) with a novel language modeling head and training objective, enabling robust denoising\n",
      "across high corruption rates and sequence generation with global context. After tr...\n",
      "Metadata title: Diffusion Sequence Models for Enhanced Protein Representation and Generation\n",
      "\n",
      "Metadata page number: 1\n",
      "\n",
      "Content: non-redundant version (P IDE <40%) of the SCOPe (Chandonia et al., 2018) database. For the\n",
      "alpha and beta proteins, the representations generated by ESM-2 are intertwined, whereas those gen-\n",
      "erated by...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 8\n",
      "\n",
      "Content: Model ClinV\n",
      "ar ProteinGym(w/o MSA retrieval) DeepLoc\n",
      "Subcellular\n",
      "Binary\n",
      "A\n",
      "UC Spearman’s ρ Acc% Acc%\n",
      "Ev\n",
      "oformer-inspired ESM(35M) 0.589\n",
      "0.178 66.11 89.25\n",
      "MIF 0.638\n",
      "0.256 61.90 85.90\n",
      "SaProt(35M) 0.754\n",
      "0...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 18\n",
      "\n",
      "Content: tures. Through extensive evaluation, our SaProt model surpasses well-established\n",
      "and renowned baselines across 10 significant downstream tasks, demonstrating\n",
      "its exceptional capacity and broad applica...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 0\n",
      "\n",
      "Content: 4.1.3 R ESULTS\n",
      "Table 1 showcases the zero-shot results on the ProteinGym and ClinVar datasets. Based on the\n",
      "results, we draw several noteworthy conclusions:\n",
      "• SaProt outperforms all residue sequence-b...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 6\n",
      "\n",
      "Content: Interestingly, the diffusion-based pLMs showcased a distinct sequence reconstruction scaling law that became clear when\n",
      "contrasting DSM and DPLM compared to ESM2 (Figure 4). Even at 90% masking, the d...\n",
      "Metadata title: Diffusion Sequence Models for Enhanced Protein Representation and Generation\n",
      "\n",
      "Metadata page number: 6\n",
      "\n",
      "Content: ESM-2\n",
      "All alpha proteins All beta proteins\n",
      "SaProt\n",
      "All alpha proteins All beta proteins\n",
      "Figure 5: Embedding visualizations of ESM-2 and SaProt on SCOPe database.\n",
      "ACKNOWLEDGMENTS\n",
      "We thank Sergey Ovchinn...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 9\n",
      "\n",
      "Content: SAPROT: P ROTEIN LANGUAGE MODELING WITH\n",
      "STRUCTURE -AWARE VOCABULARY\n",
      "Jin Su, Chenchen Han, Yuyang Zhou, Junjie Shan, Xibin Zhou, Fajie Yuan∗\n",
      "Westlake University\n",
      "{sujin, hanchenchen, zhouyuyang, shanjun...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 0\n",
      "\n",
      "Content: phase. This highlights the effectiveness of SaProt, even in situations where protein structures are not\n",
      "available.\n",
      "But in general, in this specific scenario, SaProt may not exhibit a clear advantage o...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 19\n",
      "\n",
      "Content: allows for seamless integration with advances in large-scale foundation AI models, such as\n",
      "BERT (Devlin et al., 2018), BART (Lewis et al., 2019), GPT (Brown et al., 2020), etc.\n",
      "• By utilizing the SA-t...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 1\n",
      "\n",
      "Content: We compare SaProt with two types of baselines: sequence-based models and structure-based mod-\n",
      "els. For sequence-based models, we include ESM-1b (Rives et al., 2019), ESM-1v (Meier et al.,\n",
      "2021) (the r...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 5\n",
      "\n",
      "Content: Like the ESM models, SaProt aims to contribute to the advancement of the biological community.\n",
      "This study has several limitations: (1) The performance of the proposed SA vocabulary heavily\n",
      "depends on ...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 8\n",
      "\n",
      "Content: Proteins play vital roles in the sustenance, growth and defense mecha-\n",
      "nisms of living organisms. They provide structural support for many \n",
      "essential biological processes such as synthesizing enzymes,...\n",
      "Metadata title: xTrimoPGLM: unified 100-billion-parameter pretrained transformer for deciphering the language of proteins\n",
      "\n",
      "Metadata page number: 0\n",
      "\n",
      "Content: In summary, DSM closes the gap between understanding and creating proteins in a single architecture, enabled by a few\n",
      "Python line changes, providing a foundation for rapid, scalable, and biologically ...\n",
      "Metadata title: Diffusion Sequence Models for Enhanced Protein Representation and Generation\n",
      "\n",
      "Metadata page number: 7\n",
      "\n",
      "Content: stage. Note that, the empirical studies in the Methods confirm that the  \n",
      "GLM model, when continued from MLM pretraining, converges faster \n",
      "than GLM model pretrained from scratch. Thus, the initial 40...\n",
      "Metadata title: xTrimoPGLM: unified 100-billion-parameter pretrained transformer for deciphering the language of proteins\n",
      "\n",
      "Metadata page number: 4\n",
      "\n",
      "Content: 44. Xingyi Cheng, Bo Chen, Pan Li, Jing Gong, Jie Tang, and Le Song. Training compute-optimal protein language models. bioRxiv, 2024. doi: 10.1101/2024.06.06.597716.\n",
      "Pages: 2024.06.06.597716 Section: ...\n",
      "Metadata title: Diffusion Sequence Models for Enhanced Protein Representation and Generation\n",
      "\n",
      "Metadata page number: 9\n",
      "\n",
      "Content: of objectives, which has led to a strategy for training xTrimoPGLM at an \n",
      "unprecedented scale of 100 billion parameters and 1 trillion training tokens. Our \n",
      "extensive experiments reveal that (1) xTrim...\n",
      "Metadata title: xTrimoPGLM: unified 100-billion-parameter pretrained transformer for deciphering the language of proteins\n",
      "\n",
      "Metadata page number: 0\n",
      "\n",
      "Content: while CASP15 consists of 56 publicly available proteins (released in  \n",
      "May 2022). Initially, we observed the perplexity (the lower the better)  \n",
      "of these two datasets in PLMs, which served as the back...\n",
      "Metadata title: xTrimoPGLM: unified 100-billion-parameter pretrained transformer for deciphering the language of proteins\n",
      "\n",
      "Metadata page number: 7\n",
      "\n",
      "Content: Fig. S5. Identical mask rate comparison plot to Figure 4 except with checkpoints for DSM 650 and ESMC models added ( 45, 46). The 40k and 80k step checkpoints\n",
      "demonstrate the gradual increase in perfo...\n",
      "Metadata title: Diffusion Sequence Models for Enhanced Protein Representation and Generation\n",
      "\n",
      "Metadata page number: 16\n",
      "\n",
      "Content: In general, SaProt exhibits slightly better performance on AF2 structures, while SaProt-PDB\n",
      "achieves better accuracy on real structures. This outcome is expected as training and testing are\n",
      "consistent...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 8\n",
      "\n",
      "Content: 85.80 74.44 0.886 0.672 0.381 0.435 84.16 93.63\n",
      "Table 2: Experimental results on 8 downstream tasks.\n",
      "4.2.1 D ATASETS\n",
      "For protein-level tasks, we evaluate SaProt on a diverse set of datasets from sever...\n",
      "Metadata title: SaProt: Protein Language Modeling with Structure-aware Vocabulary\n",
      "\n",
      "Metadata page number: 6\n",
      "\n",
      "Content: 4.2 DSM demonstrated substantially improved sequence reconstruction compared to MLM coun-\n",
      "terparts\n",
      "In line with more direct comparisons to existing MLM-based pLMs, we examined sequence reconstruction ...\n",
      "Metadata title: Diffusion Sequence Models for Enhanced Protein Representation and Generation\n",
      "\n",
      "Metadata page number: 6\n",
      "\n",
      "Content: the return (Fig. 1b). We speculate that protein sequences, unlike natural \n",
      "language, are less smooth in semantic space and more akin to a factual \n",
      "nature. Currently, PLM-based methods still struggle t...\n",
      "Metadata title: xTrimoPGLM: unified 100-billion-parameter pretrained transformer for deciphering the language of proteins\n",
      "\n",
      "Metadata page number: 9\n",
      "\n",
      "Content: for predicting subsequent words, the prefix-encoding portion \n",
      "remains bidirectional.\n",
      "The framework of xTrimoPGLM, along with its application in down-\n",
      "stream tasks, is depicted in Fig. 1a. Motivated by...\n",
      "Metadata title: xTrimoPGLM: unified 100-billion-parameter pretrained transformer for deciphering the language of proteins\n",
      "\n",
      "Metadata page number: 1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"\\n=== ANSWER ===\")\n",
    "print(result['result'])   \n",
    "print(\"\\n=== SOURCES ===\")\n",
    "for doc in result['source_documents']:\n",
    "    print(f\"Content: {doc.page_content[:200]}...\")\n",
    "    print(f\"Metadata title: {doc.metadata['title']}\\n\")\n",
    "    print(f\"Metadata page number: {doc.metadata['page']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7c9c4bb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "result = qa_chain.invoke('''Which microorganisms (bacteria and fungi) are capable of degradation polyethylene (PE)?. \n",
    "                        By using which methods were they isolated and from which environments? \n",
    "                        Is Lysinibacillus and also Rhodococcous strains are capable of degrading PE? \n",
    "                        Use the given context (source documents) to answer the question.\n",
    "                        Give the name of the source documents and metadata that supports your answer.''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "050d66e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== ANSWER ===\n",
      "**Microorganisms that have been shown in the supplied documents to degrade polyethylene (PE)**  \n",
      "\n",
      "| Domain | Species / genus (examples) reported as PE‑degraders | How they were obtained (isolation method & source environment) | Document(s) & bibliographic details |\n",
      "|--------|------------------------------------------------------|--------------------------------------------------------------------------|-------------------------------------|\n",
      "| **Bacteria** | • *Streptomyces* spp. ( *badius, setonia, viridosporus* )  <br>• *Bacillus* spp. (many strains cited in several surveys)  <br>• *Lysinibacillus sp.* JJY0216  <br>• *Lysinibacillus fusiformis*  <br>• *Methylobacterium paraoxydans*  <br>• *Brevibacillus borstelensis* strain 707  <br>• *Pseudomonas* spp.  <br>• *Arthrobacter* sp. (HDPE degrader)  <br>• *Priestia megaterium* (formerly *Bacillus megaterium*)  <br>• *Klebsiella pneumoniae*  <br>• *Pseudomonas fluorescens*  <br>• *Enterobacter ludwigii*  <br>• *Chryseobacterium* sp. | • **Enrichment of a mixed culture** in a minimal medium (1 % glucose + 1 % yeast extract) with pieces of LDPE/OPP; after several days the culture was plated on LB agar, single colonies were picked and identified by 16S‑rRNA sequencing. The original inoculum came from **soil collected under a kenaf–plastic composite cover** in a grove (i.e., a landfill‑type environment).  <br>• Some isolates required **pre‑treatments of the plastic** (e.g., UV irradiation for *Brevibacillus*; 65 % nitric acid for *M. paraoxydans*) before enrichment.  <br>• *Streptomyces* strains were taken from earlier literature (Pometto et al., 1992) that used **soil‑derived actinomycetes**.  <br>• *Arthrobacter* and *Pseudomonas* isolates were reported from **landfill soils** in previous studies cited in the text.  <br>• The *Rhodococcus* strain (see below) was isolated after a **609‑day enrichment of naturally weathered plastic waste** in seawater microcosms. | • “Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation”, *Microbiological Research* 267 (2023).  <br>• “Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove”, *Polymer Degradation and Stability* 191 (2021) 109662.  <br>• Table 1 in the same selection paper (lists *M. paraoxydans*, *B. borstelensis* etc.).  <br>• Pometto et al., 1992 (cited within the text).  <br>• Zheng et al., 2005 (cited for *Arthrobacter*). |\n",
      "| **Fungi** | • *Aspergillus* spp. (e.g., *A. niger, A. terreus*)  <br>• *Fusarium* spp. (*F. solani, F. oxysporum*)  <br>• *Penicillium* spp. (e.g., *P. simplicissimum*)  <br>• *Lecanicillium* sp.  <br>• *Trichoderma* sp.  <br>• *A. flavus*, *F. oxysporum*, *Penicillium granulatum* (reported as PP degraders but also listed among PE‑degraders in the review) | • **Direct plating of landfill/soil samples** onto agar media, followed by incubation on polyethylene granules or films; formation of a biofilm and SEM‑observed surface cracks were taken as evidence of degradation.  <br>• Some isolates came from **municipal‑landfill soils** (e.g., strains *L. araneicola*, *F. solani* mentioned in the text).  <br>• Others were obtained from **culture collections/previous studies** that had screened fungal isolates for plastic degradation (references to Koutny et al., 2006; Manzur et al., 2004; Seneviratne et al., 2006, etc.). | • Same “Selection of microorganisms…” paper (Microbiological Research 2023) – the discussion on fungal PE degraders.  <br>• The paragraph beginning “It is generally accepted that fungi degrade polyethylene more efficiently than bacteria” (cites Aspergillus, Fusarium, Penicillium).  <br>• References within that paragraph (Koutny et al., 2006; Manzur et al., 2004; Seneviratne et al., 2006) provide the original source of those isolates. |\n",
      "\n",
      "---\n",
      "\n",
      "### Are **Lysinibacillus** and **Rhodococcus** strains capable of degrading PE?\n",
      "\n",
      "**Yes.**\n",
      "\n",
      "* **Lysinibacillus sp.** – Both the mixed‑culture study and the pure‑culture work with strain **JJY0216** showed a **9 % weight loss of LDPE after 28 days** (pure culture) and **≈4 % loss of polypropylene** (same strain). The degradation was observed without any physicochemical pretreatment, confirming that this *Lysinibacillus* isolate can directly degrade PE.  \n",
      "  *Source:* “Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation”, Microbiol. Res. 267 (2023); also “Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216…”, Polymer Degrad. & Stability 191 (2021).\n",
      "\n",
      "* **Rhodococcus sp.** – The article **“Polyethylene Degradation by a Rhodococcous Strain Isolated from Naturally Weathered Plastic Waste Enrichment”**, *Environmental Science & Technology* 57 (2023), reports that after 30 days incubation the Rhodococcus strain A34 achieved **≈1 % weight loss of PE** together with chemical evidence (C–O, C=O groups, pits on the surface). This demonstrates that a Rhodococcus isolate can also degrade polyethylene.  \n",
      "  *Source:* “Polyethylene Degradation by a Rhodococcous Strain Isolated from Naturally Weathered Plastic Waste Enrichment”, Environ. Sci. Technol. 57 (2023).\n",
      "\n",
      "---\n",
      "\n",
      "### Summary  \n",
      "\n",
      "- A broad suite of **bacterial genera** (*Streptomyces, Bacillus, Lysinibacillus, Methylobacterium, Brevibacillus, Pseudomonas, Arthrobacter, Priestia, Klebsiella, Enterobacter, Chryseobacterium*) and **fungal genera** (*Aspergillus, Fusarium, Penicillium, Lecanicillium, Trichoderma, etc.) have been documented in the supplied texts as being able to break down polyethylene.  \n",
      "- The isolates were obtained mainly by **enrichment culturing of soil or landfill material** (including a kenaf‑plastic composite site) in minimal media containing PE as the sole carbon source, sometimes after **pre‑treatment of the polymer**, followed by plating and molecular identification (16S rRNA for bacteria; ITS/ morphological ID for fungi). Additional isolates came from **long‑term natural weathering enrichments** (Rhodococcus) or from previously published collections.  \n",
      "- Both **Lysinibacillus sp. JJY0216** and a **Rhodococcus sp. strain A34** are explicitly shown in the provided documents to degrade PE, confirming their capability.  \n",
      "\n",
      "=== SOURCES ===\n",
      "Content: and domestic applications. Polyethylene and polypropylene account for more than 50% of total plastic \n",
      "usage, and they resist decomposition in a natural environment. In this study, Lysinibacillus sp., ...\n",
      "Metadata title: Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove\n",
      "\n",
      "Metadata page number: 0\n",
      "\n",
      "Content: than the number of bacterial strains. It is generally accepted that fungi \n",
      "degrade polyethylene more efficiently than bacteria ( Muhonja et al., \n",
      "2018 ). The first reason for this difference is the ab...\n",
      "Metadata title: Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation\n",
      "\n",
      "Metadata page number: 6\n",
      "\n",
      "Content: the degradation process ( Orr et al., 2004 ). Structural changes were also \n",
      "analyzed using SEM. \n",
      "Based on the literature, biodegradation of polyethylene by micro -\n",
      "organisms was investigated as early ...\n",
      "Metadata title: Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation\n",
      "\n",
      "Metadata page number: 6\n",
      "\n",
      "Content: dation experiments. A minimal medium, with continuous feeding \n",
      "with glucose and yeast extract, was used to examine degradation. \n",
      "The growth of the mixed culture reached a stationary phase in four \n",
      "day...\n",
      "Metadata title: Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove\n",
      "\n",
      "Metadata page number: 2\n",
      "\n",
      "Content: the tested polymer ( Restrepo-Fl ´orez et al., 2014 ) but does not apply to \n",
      "the P. megaterium (former Bacillus megaterium ) strain that was isolated \n",
      "from soil samples taken at a landfill. S. marcesc...\n",
      "Metadata title: Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation\n",
      "\n",
      "Metadata page number: 6\n",
      "\n",
      "Content: Microbiological Research 267 (2023) 127251\n",
      "Available online 8 November 2022\n",
      "0944-5013/© 2022 Published by Elsevier GmbH. This article is made available under the Elsevier license (http://www.elsevier....\n",
      "Metadata title: Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation\n",
      "\n",
      "Metadata page number: 0\n",
      "\n",
      "Content: Polymer Degradation and Stability 191 (2021) 109662 \n",
      "Contents lists available at ScienceDirect \n",
      "Polymer Degradation and Stability \n",
      "journal homepage: www.elsevier.com/locate/polymdegradstab \n",
      "Biodegrada...\n",
      "Metadata title: Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove\n",
      "\n",
      "Metadata page number: 0\n",
      "\n",
      "Content: In conclusion, the isolation of plastic-degrading bacteria and fungi from waste landfills represents an effective \n",
      "strategy for the collection of microorganisms with high potential for PE and PP degra...\n",
      "Metadata title: Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation\n",
      "\n",
      "Metadata page number: 0\n",
      "\n",
      "Content: Polyethylene Degradation by a Rhodococcous Strain Isolated from\n",
      "Naturally Weathered Plastic Waste Enrichment\n",
      "Xuanyu Tao, Huanrong Ouyang, Aifen Zhou,* Dongyu Wang, Hagan Matlock, Josiah S. Morgan,\n",
      "Abi...\n",
      "Metadata title: Polyethylene Degradation by a Rhodococcous Strain Isolated from Naturally Weathered Plastic Waste Enrichment\n",
      "\n",
      "Metadata page number: 0\n",
      "\n",
      "Content: Biodeterior. Biodegradation 113 (2016) 244–255 . \n",
      "[14] S. Mukherjee , U.R. Chowdhuri , P.P. Kundu , Bio-degradation of polyethylene \n",
      "waste by simultaneous use of two bacteria: bacillus licheniformis f...\n",
      "Metadata title: Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove\n",
      "\n",
      "Metadata page number: 7\n",
      "\n",
      "Content: biodegradation of polymers. \n",
      "We assumed that natural biological material collected from a landfill \n",
      "will contain some microorganisms capable of utilizing polymers (PE and \n",
      "PP) as the primary source of...\n",
      "Metadata title: Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation\n",
      "\n",
      "Metadata page number: 1\n",
      "\n",
      "Content: UV light before inoculation with Bacillus flexus and reveald 2.5 % \n",
      "reduction in polymer weight, compared to non-inoculated control \n",
      "variant. It was confirmed that PE known for its high density (HDPE)...\n",
      "Metadata title: Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation\n",
      "\n",
      "Metadata page number: 1\n",
      "\n",
      "Content: J.-M. Jeon, S.-J. Park, T.-R. Choi et al. Polymer Degradation and Stability 191 (2021) 109662 \n",
      "dation in natural environments. Several microbial species, includ- \n",
      "ing Bacillus sp., Brevibacillus borst...\n",
      "Metadata title: Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove\n",
      "\n",
      "Metadata page number: 1\n",
      "\n",
      "Content: J.-M. Jeon, S.-J. Park, T.-R. Choi et al. Polymer Degradation and Stability 191 (2021) 109662 \n",
      "Fig. 5. Scheme of biological degradation process for plastic waste. \n",
      "study can be used as a basis for fur...\n",
      "Metadata title: Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove\n",
      "\n",
      "Metadata page number: 6\n",
      "\n",
      "Content: J.-M. Jeon, S.-J. Park, T.-R. Choi et al. Polymer Degradation and Stability 191 (2021) 109662 \n",
      "Fig. 3. Biodegradation ability of polyethylene and polypropylene by Lysinibacillus sp. JJY0216. Growth of...\n",
      "Metadata title: Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove\n",
      "\n",
      "Metadata page number: 3\n",
      "\n",
      "Content: pounds, which are secreted and broken down by the last group of \n",
      "microbes ( Shah et al., 2008 ). It is worth noting that the efficiency of \n",
      "microorganisms ’ degradation capacity varies under different...\n",
      "Metadata title: Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation\n",
      "\n",
      "Metadata page number: 1\n",
      "\n",
      "Content: degradation of plastic waste. Notably, the three bacteria within our \n",
      "reconstituted bacterial community (Exiguobacterium sp., Halomonas sp. \n",
      "and Ochrobactrum sp.) are all fast growing and easily cultu...\n",
      "Metadata title: A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene\n",
      "\n",
      "Metadata page number: 10\n",
      "\n",
      "Content: J.-M. Jeon, S.-J. Park, T.-R. Choi et al. Polymer Degradation and Stability 191 (2021) 109662 \n",
      "Fig. 1. Biodegradation of polyethylene and polypropylene by mixed culture. Growth of mixed culture and we...\n",
      "Metadata title: Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove\n",
      "\n",
      "Metadata page number: 2\n",
      "\n",
      "Content: ture of polypropylene. \n",
      "Moreover, the crystallinity of each LDPE and OPP was calcu- \n",
      "lated. Compared with the 46% crystallinity of control LDPE, LDPE \n",
      "biodegraded by Lysinibacillus sp. JJY0216 and mix...\n",
      "Metadata title: Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove\n",
      "\n",
      "Metadata page number: 4\n",
      "\n",
      "Content: ( Benedict et al., 1983 ); however, its pathogenic nature means that it \n",
      "cannot be viewed as a potential agent used in applied research. \n",
      "5. Conclusion \n",
      "Incubation of biological materials extracted fr...\n",
      "Metadata title: Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation\n",
      "\n",
      "Metadata page number: 9\n",
      "\n",
      "Content: stabilization. Macromol. Symp. 115, 143 – 163. https://doi.org/10.1002/ \n",
      "masy.19971150110 . \n",
      "Pramila, R., Ramesh, K., 2017. Biodegradation of low density polyethylene (LDPE) by \n",
      "fungi isolated from mu...\n",
      "Metadata title: Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation\n",
      "\n",
      "Metadata page number: 10\n",
      "\n",
      "Content: proximately 7.5% on day 28 ( Fig. 3 B). The OPP ﬁlm’s weight was \n",
      "decreased by approximately 3.1% on day 28, a result consistent \n",
      "with the mixed culture ﬁndings ( Fig. 3 B). These results are more \n",
      "th...\n",
      "Metadata title: Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove\n",
      "\n",
      "Metadata page number: 2\n",
      "\n",
      "Content: microorganism-mediated PE degradation lags well behind. Further -\n",
      "more, degradation efficiencies for single strains or single enzymes are \n",
      "still too low to meet industrial application requirements. Al...\n",
      "Metadata title: A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene\n",
      "\n",
      "Metadata page number: 1\n",
      "\n",
      "Content: shown that microorganisms, including bacteria and fungi as\n",
      "well as insects and the associated gut microorganisms, can\n",
      "degrade various types of plastics such as polyethylene (PE),\n",
      "polystyrene (PS), pol...\n",
      "Metadata title: Polyethylene Degradation by a Rhodococcous Strain Isolated from Naturally Weathered Plastic Waste Enrichment\n",
      "\n",
      "Metadata page number: 0\n",
      "\n",
      "Content: ethylene (PE) and polypropylene (PP) degradation and assessed the scale of polymer degradation by microbial \n",
      "populations. \n",
      "We isolated seven bacterial strains (three Priestia megaterium strains, Klebs...\n",
      "Metadata title: Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation\n",
      "\n",
      "Metadata page number: 0\n",
      "\n",
      "Content: chloride coating under the effect of microorganisms. Prot. Met. 23, 173 – 175 . \n",
      "Mohanan, N., Montazer, Z., Sharma, P.K., Levin, D.B., 2020. Microbial and enzymatic \n",
      "degradation of synthetic plastics....\n",
      "Metadata title: Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation\n",
      "\n",
      "Metadata page number: 10\n",
      "\n",
      "Content: and PE degradation by this reconstituted bacterial community. Lastly, \n",
      "we used transcriptomic methods to clarify the potential degradation \n",
      "processes and enzymes associated with these processes. \n",
      "2. M...\n",
      "Metadata title: A marine bacterial community capable of degrading poly(ethylene terephthalate) and polyethylene\n",
      "\n",
      "Metadata page number: 1\n",
      "\n",
      "Content: ing in rarely biodegradable in the absence of pretreatment pro- \n",
      "cesses such as extended time exposure to high temperature or \n",
      "UV exposure [1] . These C –C chains can be broken by alkane de- \n",
      "hydrogen...\n",
      "Metadata title: Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove\n",
      "\n",
      "Metadata page number: 1\n",
      "\n",
      "Content: Microbiological Research 267 (2023) 127251\n",
      "10\n",
      "Aspergillus niger, A. terreus, Aureobasidium pullulans, Chaetomium \n",
      "globosum, Paecilomyces variotii, and Scopulariopsis brevicaulis are fungal \n",
      "strains th...\n",
      "Metadata title: Selection of microorganisms capable of polyethylene (PE) and polypropylene (PP) degradation\n",
      "\n",
      "Metadata page number: 9\n",
      "\n",
      "Content: tured at 30 ◦C by inoculating 0.1 g of soil sample with 50 mL \n",
      "of Luria–Bertani (LB) medium. The same conditions were used for \n",
      "seed culture to assess biodegradability. To isolate a single strain \n",
      "fro...\n",
      "Metadata title: Biodegradation of polyethylene and polypropylene by Lysinibacillus species JJY0216 isolated from soil grove\n",
      "\n",
      "Metadata page number: 1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"\\n=== ANSWER ===\")\n",
    "print(result['result'])   \n",
    "print(\"\\n=== SOURCES ===\")\n",
    "for doc in result['source_documents']:\n",
    "    print(f\"Content: {doc.page_content[:200]}...\")\n",
    "    print(f\"Metadata title: {doc.metadata['title']}\\n\")\n",
    "    print(f\"Metadata page number: {doc.metadata['page']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "61030a02",
   "metadata": {},
   "outputs": [],
   "source": [
    "result = qa_chain.invoke('''Summarize the main group of TNT degrading microbial species, separate into Aerobic vs. Anaerobic bacteria, what are the main differences? \n",
    "                        Use the given context (source documents) to answer the question.\n",
    "                        Give the name of the source documents and metadata that supports your answer.''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e78ed0b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== ANSWER ===\n",
      "**Main groups of TNT‑degrading microbes (as described in the document “Comprehensive analysis of microbial TNT degradation”)**\n",
      "\n",
      "|                     | Representative taxa mentioned in the text | Typical metabolic features |\n",
      "|---------------------|-------------------------------------------|----------------------------|\n",
      "| **Aerobic bacteria** | • *Pseudomonas* spp. (e.g., *P. fluorescens*, *P. putida*)  <br>• *Bacillus* spp.<br>• *Staphylococcus* spp.<br>• *Mycobacterium* spp.<br>• *Rhodococcus* spp. (e.g., *R. erythropolis*)<br>• *Serratia marcescens* <br>• *Stenotrophomonas* isolates | – Operate in the presence of O₂.  <br>– Initiate TNT degradation by **nitro‑group reduction** (nitroreductases, azoreductases) producing hydroxylamino‑ and amino‑dinitrotoluenes (ADNTs, DANTs).<br>– Some strains form Meisenheimer σ‑complexes that release nitrite (denitration) and generate dinitrotoluene intermediates that can be further attacked by dioxygenases.<br>– Frequently generate **azoxy dimers** or polymeric products; complete mineralisation of the aromatic ring is rare.  <br>– Often use TNT only as a nitrogen source, not for growth energy. |\n",
      "| **Anaerobic bacteria** | • *Veillonella alkalescens*<br>• *Escherichia coli* (facultative)<br>• *Clostridium* spp. (*C. bifermentans*, *C. acetobutylicum*, *C. pasterianum*, *C. sordellii*)<br>• *Desulfovibrio* spp.<br>• Methanogenic archaea (e.g., *Methanococcus*) | – Function when O₂ is absent or limited.  <br>– Carry out **sequential six‑electron reductions** of each nitro group, converting TNT → TAT (2,4,6‑triaminotoluene). <br>– Reduction proceeds through nitroso → hydroxylamino → amine intermediates; often mediated by oxygen‑insensitive nitroreductases. <br>– The fully reduced product (TAT) is chemically unstable and tends to polymerise or adsorb, but the pathway removes all nitro groups – a step that aerobic cells rarely achieve.  <br>– Some anaerobes can use TNT as an **electron acceptor** (e.g., *Desulfovibrio* spp.) or as a nitrogen source; the process does not generate much energy for growth but alleviates toxicity. |\n",
      "\n",
      "### Main differences between aerobic and anaerobic TNT‑degrading bacteria  \n",
      "\n",
      "| Aspect | Aerobic bacteria | Anaerobic bacteria |\n",
      "|-------|------------------|--------------------|\n",
      "| **Electron acceptor** | O₂ is available; oxygenases can act after nitro‑group reduction. | No O₂; reduction relies on electron donors such as H₂, NADH, or ferredoxin. |\n",
      "| **Extent of nitro‑group removal** | Typically removes one or two nitro groups (ADNTs/DANTs); complete three‑step reduction to TAT is uncommon. | Generally reduces **all three** nitro groups → TAT; the pathway can be “complete” with respect to denitration. |\n",
      "| **Aromatic ring fate** | Ring often remains intact; further oxidation to catechol or mineralisation occurs only after conversion to dinitrotoluene (e.g., via Meisenheimer complex). | The aromatic ring may stay intact as TAT, which can polymerise or be immobilised; true mineralisation of the carbon skeleton is still rare. |\n",
      "| **Typical products** | ADNTs, DANTs, azoxy dimers, nitrite, occasional dinitrotoluene that enters known dioxygenase pathways. | TAT (2,4,6‑triaminotoluene), sometimes aminophenols or toluene (via speculative deamination). |\n",
      "| **Physiological role** | Often use TNT as a nitrogen source; the process is largely co‑metabolic and can be coupled to growth on other carbon sources. | TNT may serve as an electron sink/terminal electron acceptor or as a nitrogen source; reduction does not provide substantial energy but reduces toxicity. |\n",
      "| **Representative environments** | Oxic soils, composts, aerobic bioreactors. | Anoxic sediments, flooded soils, anaerobic digesters, sulfate‑reducing zones. |\n",
      "\n",
      "### Source documents & metadata supporting the answer  \n",
      "\n",
      "1. **Section “TNT‑Degrading Microbial Species – Aerobic Bacteria”** (within the main document).  \n",
      "   *Mentions*: *Pseudomonas spp.*, *Bacillus*, *Staphylococcus*, *Mycobacterium*, *Rhodococcus erythropolis*, *Serratia marcescens*, *Stenotrophomonas* – describes aerobic nitro‑group reduction, Meisenheimer complexes and limited mineralisation.\n",
      "\n",
      "2. **Section “Other Anaerobes”** (within the same document).  \n",
      "   *Mentions*: *Veillonella alkalescens*, *Escherichia coli* (facultative), methanogenic archaea (*Methanococcus*) – outlines anaerobic complete reduction to TAT.\n",
      "\n",
      "3. **Section “Clostridium spp.”** under “Multi‑Omics Approaches and Systems Biology Insights”.  \n",
      "   *Details*: *Clostridium bifermentans*, *C. acetobutylicum*, etc., converting TNT → DANTs → TAT; notes Bamberger rearrangement, azoxy polymer formation.\n",
      "\n",
      "4. **Section “Anaerobic Bacteria”** and **“Complete Reduction (Anaerobic Pathway)”** subsections.  \n",
      "   *Describes*: sequential six‑electron reductions, role of sulfate‑reducers (*Desulfovibrio* spp.), use of TNT as electron acceptor.\n",
      "\n",
      "5. **Reference list entries** that give bibliographic metadata:  \n",
      "\n",
      "   - Ramos, J.L., et al. (2001). *Biological degradation of 2,4,6‑trinitrotoluene (TNT)*. **Microbiology and Molecular Biology Reviews**, 65(3):335–352. *(Provides background on TNT biodegradation.)*  \n",
      "   - Cabrera, M.Á., Márquez, S.L., & Pérez‑Donoso, J.M. (2022). *Comparative genomic analysis of Antarctic Pseudomonas isolates with 2,4,6‑trinitrotoluene transformation capabilities*. **Microorganisms**, 10(8):1581. *(Supports aerobic Pseudomonas data.)*  \n",
      "   - Ziganshin, A.M., et al. (2007). *Hydride‑mediated reduction of 2,4,6‑trinitrotoluene by yeasts as a way to its deep degradation*. **Mikrobiologiia**, 76(6):766–773. *(Although about yeasts, it reinforces the contrast between aerobic nitro‑reduction and deeper anaerobic‐type reductions.)*  \n",
      "\n",
      "These sections collectively provide the taxonomy of aerobic vs. anaerobic TNT‑degrading bacteria and the mechanistic differences highlighted above.\n",
      "\n",
      "=== SOURCES ===\n",
      "Content: Scope of Document: This document provides a comprehensive analysis of microbial TNT degradation. We first catalog the known TNT-degrading species across bacteria, fungi, and yeasts, highlighting their...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Aerobic Bacteria: Under oxic conditions, many soil bacteria can initiate TNT degradation by partial reduction of its nitro groups or removal of nitrite. Notable examples include:\n",
      "\n",
      "• Pseudomonas spp.: ...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Other Anaerobes: Veillonella alkalescens was one of the first organisms noted to reduce TNT to TAT under strictly anaerobic conditions. Even Escherichia coli (facultative anaerobe) can perform compl...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Microbial Bioremediation: The use of microorganisms to detoxify and remove TNT from the environment offers a promising alternative to costly physicochemical cleanup methods . Certain bacteria and fung...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: In summary, TNT degradation pathways are complex and can follow multiple routes depending on the organism and conditions. Aerobic processes excel at initial transformations and co-metabolic turnover o...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Clostridium spp.: These obligate anaerobes are among the most effective TNT degraders in anoxic conditions. Clostridium bifermentans strains transform TNT through successive reductions to diaminonit...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Other Aerobes: Bacillus and Staphylococcus species can reduce TNT, typically to monoamino dinitrotoluene (e.g., 2-amino-4,6-dinitrotoluene) with nitrite release. Mycobacterium and Rhodococcus (Gram-...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Comparative Genomic Insights: Genome sequencing of TNT-degrading strains has shed light on their evolutionary adaptations. For example, three Antarctic Pseudomonas isolates (TNT3, TNT11, TNT19) were s...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Complete Reduction (Anaerobic Pathway): In oxygen-free environments, TNT can be fully reduced to TAT (2,4,6-triaminotoluene) by certain bacteria. This occurs via sequential addition of six electrons (...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: with other anaerobic processes (e.g., use of TNT as a terminal electron acceptor). In remediation strategies, an anaerobic phase is sometimes used to reduce TNT to TAT, followed by an aerobic phase to...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: . These azoxy compounds are intensely colored, highly recalcitrant, and represent dead-end immobilization products in many TNT-fed cultures . In summary, aerobic nitro-reduction rapidly eliminates TNT...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Anaerobic Degradation Routes: In oxygen-free environments, TNT undergoes more complete reduction. The hallmark of anaerobic TNT metabolism is the production of triaminotoluene (TAT). This results from...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Sulfate-Reducers: Desulfovibrio species can anaerobically respire TNT. For example, Desulfovibrio sp. strain B uses TNT as a nitrogen source and may produce toluene as a downstream product . Other D...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: In summary, the genetic toolkit for TNT biodegradation comprises mostly reductive enzymes that either remove nitro groups or destabilize the aromatic ring, along with supporting systems to manage stre...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Genomic and Metabolic Insights into Microbial Biodegradation of TNT\n",
      "\n",
      "\n",
      "\n",
      "Abstract...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: on TNT as sole nitrogen source and was suggested to sequentially remove all three nitro substituents, as evidenced by detection of DNTs, mononitrotoluene, and toluene in culture . The release of nitri...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: In summary, omics studies have expanded our understanding from individual enzymes to the global response of microorganisms to TNT. They reveal that TNT degradation is an orchestrated event involving s...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Genomic Insights: Key Genes and Enzymes\n",
      "\n",
      "\n",
      "\n",
      "Microbial TNT degradation relies on a suite of specialized enzymes that initiate TNT transformation despite its inert aromatic structure. Genomic and biochem...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: to its electron transport chain, certain metabolic pathways, and motility functions during TNT exposure. This likely reflects a shift to a slower-growing, stress-resistant state: the bacterium conserv...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Nitroreductases (Type I): A common entry point for TNT metabolism is the stepwise reduction of nitro (-NO₂) groups to nitroso (-NO), hydroxylamino (-NHOH), and amino (-NH₂) groups. These reactions a...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Azoreductases: An unexpected finding from genomic studies is the involvement of azoreductase enzymes in TNT degradation. Azoreductases normally reduce azo bonds (–N=N–) in dyes, but some can also re...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: 7. Hannink, N.K., et al. (2001). Phytodetoxification of TNT by transgenic plants expressing a bacterial nitroreductase. Nature Biotechnology, 19(12): 1168–1172 .\n",
      "\n",
      "8. Duque, E., et al. (1993). Identifi...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Key Intermediates and Fate: Across these pathways, the recurring intermediates are: hydroxylamino-dinitrotoluenes, amino-dinitrotoluenes (ADNTs), diaminonitrotoluenes (DANTs), and triaminotoluene (TAT...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Transcriptomic Profiling: Genome-wide RNA expression analyses reveal which genes are turned on or off in the presence of TNT. A recent study examined the transcriptome of Pseudomonas sp. TNT3 (an An...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Enterobacteriaceae: Enterobacter cloacae PB2 is capable of denitrating TNT. It reduces TNT via formation of a Meisenheimer complex and releases nitrite . An NADPH-dependent enzyme called pentaerythr...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Accessory and Regulatory Genes: Alongside catabolic enzymes, microbes possess regulatory and support genes that enable TNT degradation. These include: (1) oxidative stress response genes – TNT redox...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Candida and Geotrichum: Environmental yeast isolates Candida sp. AN-L15 and Geotrichum sp. AN-Z4 degrade TNT through an alternative pathway dominated by hydride attack on the aromatic ring . These y...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Aerobic Degradation Routes: In the presence of oxygen, TNT is not readily mineralized, but it undergoes a series of reductive transformations: the dominant pathway is sequential nitro-group reduction....\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: 2. Cabrera, M.Á., Márquez, S.L., & Pérez-Donoso, J.M. (2022). Comparative genomic analysis of Antarctic Pseudomonas isolates with 2,4,6-trinitrotoluene transformation capabilities reveals their unique...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Yarrowia lipolytica: This robust yeast is known for degrading hydrocarbons and was also reported to degrade TNT. Studies found that factors like pH influence TNT transformation by Y. lipolytica, and...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"\\n=== ANSWER ===\")\n",
    "print(result['result'])   \n",
    "print(\"\\n=== SOURCES ===\")\n",
    "for doc in result['source_documents']:\n",
    "    print(f\"Content: {doc.page_content[:200]}...\")\n",
    "    print(f\"Metadata title: {doc.metadata['source']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "cffa12c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "result = qa_chain.invoke('''What are the main genes and enzymes involved in TNT degradation? \n",
    "                        Make a brief summary of the mechanism of action of each such gene/enzyme.\n",
    "                        Use the given context (source documents) to answer the question.\n",
    "                        Give the name of the source documents and metadata that supports your answer.''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "41c80e6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== ANSWER ===\n",
      "**Main genes / enzymes that have been identified in the microbial degradation of 2‑4‑6‑trinitrotoluene (TNT)**  \n",
      "\n",
      "| Gene / Enzyme (example) | Primary catalytic/mechanistic role in TNT turnover | Where this information appears in the supplied context |\n",
      "|--------------------------|---------------------------------------------------|--------------------------------------------------------|\n",
      "| **Type I nitroreductases** – *nfsA, nfsB, nemA* (e.g., NemA from a Citrobacter sp.) | Two‑electron reduction of TNT’s nitro groups (–NO₂ → –NO → –NHOH → –NH₂) using FMN/FAD and NAD(P)H as electron donors. This generates aminodinitrotoluene (ADNT) intermediates but does **not** cleave the aromatic ring. | “Nitroreductases (Type I)… genes encoding such enzymes (e.g., nfsA, nfsB … nemA…) have been found… NemA was most up‑regulated… converts TNT to 4‑ADNT and 2‑ADNT.” – *Genomic Insights: Key Genes and Enzymes*; also in the “TNT‑Degrading Microbial Species” section. |\n",
      "| **Azoreductase** (azoreductase‑encoding gene highly induced in *Pseudomonas sp.* TNT3) | Reduces nitro‑aromatic or hydroxylamino intermediates that arise during TNT catabolism; may attack TNT itself or its hydroxylamine derivatives, helping to prevent accumulation of toxic intermediates. The enzyme is an FMN‑dependent reductase that uses NAD(P)H. | “Azoreductases… a recent transcriptomic analysis … revealed an azoreductase‑encoding gene as the most highly induced gene (>1000‑fold…) during TNT exposure.” – *Genomic Insights: Key Genes and Enzymes*; also reiterated in “Transcriptomic Profiling”. |\n",
      "| **Old Yellow Enzyme (OYE) family flavoprotein reductases** – e.g., PETN reductase (*petnR / pnrA*), XenA/B, NEM reductases | Catalyze hydride transfer from NADPH to the electron‑poor aromatic ring of TNT, forming a σ‑Meisenheimer complex that spontaneously eliminates nitrite (denitration). This both reduces nitro groups and destabilises the ring, opening pathways toward ring cleavage. | “Old Yellow Enzyme … enzymes … include PETN reductase (gene petnR, also known as pnrA)… can perform hydride ion attack on the aromatic ring of TNT, yielding a σ‑complex that spontaneously eliminates nitrite.” – *Genomic Insights: Key Genes and Enzymes*; further discussed under “Old Yellow Enzyme (OYE) Family & Flavoprotein Reductases”. |\n",
      "| **NADPH dehydrogenase NamA** | Supplies reducing power (NADPH) to the above reductases, thereby supporting the two‑electron reductions of nitro groups and Meisenheimer‑complex formation. | “NADPH dehydrogenase NamA” is listed under *Extra material: additional genes/pathways related to TNT‑biodegradation*. |\n",
      "| **Oxidative‑stress enzymes** – Superoxide dismutases (SOD), peroxidases, catalases | Scavenge reactive oxygen species generated by TNT redox cycling; protect cellular macromolecules while the reductive catabolism proceeds. | “Accessory and Regulatory Genes… oxidative stress response genes – … superoxide dismutases and peroxidases are often up‑regulated” – *Genomic Insights: Key Genes and Enzymes*. |\n",
      "| **Nitrosative‑stress enzymes** – Nitric‑oxide dioxygenase, DNA‑repair proteins | Detoxify nitric oxide / nitroso intermediates that arise from nitro‑group reduction; repair damage to nucleic acids. | Same paragraph as above (oxidative‑stress genes). |\n",
      "| **Assimilatory nitrite reductase (NasB) & related nitrogen‑assimilation enzymes** (e.g., glutamine synthetase) | Reduce nitrite released during denitration to ammonium and incorporate it into amino‑acid biosynthesis, allowing the cell to use TNT as a nitrogen source. | “Nitrite Reductases and Assimilatory Enzymes… NasB … together convert TNT‑derived nitrite into ammonium for incorporation” – *Introduction* section. |\n",
      "| **Multidrug efflux pumps / transporters** (e.g., RND family) | Export TNT, its reduced products or toxic intermediates out of the cytoplasm, lowering intracellular concentrations and alleviating stress. | “Transporters and efflux pumps… multidrug efflux systems … likely to reduce intracellular toxin accumulation” – *Genomic Insights: Key Genes and Enzymes*. |\n",
      "| **Regulatory proteins** – LysR‑type, MarR‑type transcriptional regulators (often plasmid‑borne) | Sense TNT or its metabolites and activate expression of the catabolic operons (nitroreductases, azoreductase, efflux pumps). | “Regulators … specialized regulators (e.g., LysR‑type or MarR‑type regulators) may sense TNT… Some TNT‑degrading plasmids carry such regulatory genes” – *Genomic Insights: Key Genes and Enzymes*. |\n",
      "\n",
      "---\n",
      "\n",
      "### How each component fits into the overall degradation mechanism  \n",
      "\n",
      "1. **Initial reduction of nitro groups** – Type I nitroreductases (e.g., NemA) perform sequential two‑electron reductions, converting TNT → 4‑ADNT/2‑ADNT → amino‑dinitrotoluenes.  \n",
      "2. **Alternative / complementary reduction** – Azoreductase and OYE‑family enzymes act in parallel: the azoreductase can reduce hydroxylamine intermediates; OYE enzymes (PETN reductase, XenA/B) add a hydride to the aromatic ring forming a Meisenheimer σ‑complex that ejects nitrite, yielding dinitrotoluene (DNT) or other de‑nitro products.  \n",
      "3. **Support of redox balance** – NADPH‑generating enzymes such as NamA keep the reductases supplied with reducing equivalents.  \n",
      "4. **Stress mitigation** – Superoxide dismutase, peroxidases, nitric‑oxide detoxifiers, and DNA‑repair proteins protect the cell from ROS/RNS generated during TNT reduction.  \n",
      "5. **Nitrogen assimilation** – NasB reduces released nitrite to ammonia; downstream ammonium‑assimilating enzymes incorporate this nitrogen into biomass.  \n",
      "6. **Detoxification & export** – Multidrug efflux pumps pump out residual TNT or toxic intermediates, reducing intracellular toxicity.  \n",
      "7. **Regulation** – LysR/MarR‐type regulators (often on the same plasmid as catabolic genes) sense TNT or its reduced metabolites and up‑regulate the whole suite of degradation genes only when needed.\n",
      "\n",
      "---\n",
      "\n",
      "### Source documents & metadata used\n",
      "\n",
      "| Document / Section | Type / Year (where provided) |\n",
      "|--------------------|------------------------------|\n",
      "| **Genomic Insights: Key Genes and Enzymes** – multiple paragraphs on nitroreductases, azoreductase, OYE family, stress genes, transporters, regulators. | Unspecified report (part of the supplied “Genomic Insights” chapter). |\n",
      "| **Transcriptomic Profiling** – description of >1000‑fold azoreductase induction in *Pseudomonas sp.* TNT3. | Same source as above; internal heading. |\n",
      "| **TNT‑Degrading Microbial Species** – discussion of nitroreductases (nemA, nfsA/B) and their role. | Same overarching document. |\n",
      "| **Old Yellow Enzyme (OYE) Family & Flavoprotein Reductases** – details on PETN reductase, XenA/B, Meisenheimer complex formation. | Same source; internal heading. |\n",
      "| **Introduction** (section “Nitrite Reductases and Assimilatory Enzymes”). | Same source. |\n",
      "| **Extra material: additional genes/pathways related to TNT‑biodegradation** – mention of NADPH dehydrogenase NamA. | Same source. |\n",
      "| **References cited in the text** (e.g., Cabrera et al. 2024; Liao et al. 2018) – provide bibliographic metadata for specific studies that reported azoreductase induction, nitroreductase up‑regulation, etc. | Author(s), year as shown in reference list. |\n",
      "\n",
      "All the above points are drawn **exclusively** from the supplied context; no external knowledge was introduced.\n",
      "\n",
      "=== SOURCES ===\n",
      "Content: Genomic Insights: Key Genes and Enzymes\n",
      "\n",
      "\n",
      "\n",
      "Microbial TNT degradation relies on a suite of specialized enzymes that initiate TNT transformation despite its inert aromatic structure. Genomic and biochem...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Functional Genomics: Another omics-related approach is random mutagenesis or knockout studies combined with sequencing to find genes essential for TNT metabolism. Transposon mutagenesis, for instanc...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Accessory and Regulatory Genes: Alongside catabolic enzymes, microbes possess regulatory and support genes that enable TNT degradation. These include: (1) oxidative stress response genes – TNT redox...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Azoreductases: An unexpected finding from genomic studies is the involvement of azoreductase enzymes in TNT degradation. Azoreductases normally reduce azo bonds (–N=N–) in dyes, but some can also re...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Systems Biology and Modeling: With large omics datasets, scientists have begun constructing regulatory and metabolic network models for TNT-degrading organisms. For instance, integrating the transcr...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Transcriptomic Profiling: Genome-wide RNA expression analyses reveal which genes are turned on or off in the presence of TNT. A recent study examined the transcriptome of Pseudomonas sp. TNT3 (an An...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: In summary, the genetic toolkit for TNT biodegradation comprises mostly reductive enzymes that either remove nitro groups or destabilize the aromatic ring, along with supporting systems to manage stre...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Comparative Genomic Insights: Genome sequencing of TNT-degrading strains has shed light on their evolutionary adaptations. For example, three Antarctic Pseudomonas isolates (TNT3, TNT11, TNT19) were s...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Scope of Document: This document provides a comprehensive analysis of microbial TNT degradation. We first catalog the known TNT-degrading species across bacteria, fungi, and yeasts, highlighting their...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Metabolomics and Pathway Confirmation: While less reported than transcript/proteome data, metabolomic analyses (identifying intermediate compounds) and isotope tracing have been used to confirm TNT ...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Nitroreductases (Type I): A common entry point for TNT metabolism is the stepwise reduction of nitro (-NO₂) groups to nitroso (-NO), hydroxylamino (-NHOH), and amino (-NH₂) groups. These reactions a...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: In summary, omics studies have expanded our understanding from individual enzymes to the global response of microorganisms to TNT. They reveal that TNT degradation is an orchestrated event involving s...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Proteomic Analyses: Complementing transcriptomics, proteomics examines changes in the protein complement of a cell in response to TNT. In one study, researchers performed an integrated proteome and ...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Genomic and Metabolic Insights into Microbial Biodegradation of TNT\n",
      "\n",
      "\n",
      "\n",
      "Abstract...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: to its electron transport chain, certain metabolic pathways, and motility functions during TNT exposure. This likely reflects a shift to a slower-growing, stress-resistant state: the bacterium conserv...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: 7. Hannink, N.K., et al. (2001). Phytodetoxification of TNT by transgenic plants expressing a bacterial nitroreductase. Nature Biotechnology, 19(12): 1168–1172 .\n",
      "\n",
      "8. Duque, E., et al. (1993). Identifi...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Microbial Bioremediation: The use of microorganisms to detoxify and remove TNT from the environment offers a promising alternative to costly physicochemical cleanup methods . Certain bacteria and fung...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: In summary, TNT degradation pathways are complex and can follow multiple routes depending on the organism and conditions. Aerobic processes excel at initial transformations and co-metabolic turnover o...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Dioxygenases and Ring-Cleavage Enzymes: Direct oxidative ring cleavage of TNT is very challenging due to its electron-poor ring, and no dedicated TNT dioxygenase has been identified. However, once T...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Abstract\n",
      "\n",
      "2,4,6-Trinitrotoluene (TNT) is a recalcitrant explosive pollutant whose breakdown in the environment is facilitated by diverse microbes. This document summarizes key findings on TNT biodegra...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Nitrite Reductases and Assimilatory Enzymes: Though not directly modifying TNT, enzymes like assimilatory nitrite reductase (NasB) and ammonium-assimilating enzymes (e.g., glutamine synthetase) are ...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Monooxygenases and Dioxygenases: Enzymes that incorporate oxygen into aromatic rings (like ring-hydroxylating dioxygenases or monooxygenases) are central to degrading simpler nitroaromatics (e.g., din...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Aerobic Bacteria: Under oxic conditions, many soil bacteria can initiate TNT degradation by partial reduction of its nitro groups or removal of nitrite. Notable examples include:\n",
      "\n",
      "• Pseudomonas spp.: ...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Partial Oxidative Pathways: Although TNT’s nitro substituents strongly inhibit direct oxygenase attack, there are instances of cometabolic oxidation of TNT. One example is oxidation of the methyl grou...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • White-Rot Fungi (Basidiomycetes): Phanerochaete chrysosporium, a well-studied ligninolytic fungus, can mineralize a portion of TNT under nutrient-limited (lignin-degrading) conditions . Fungal TNT d...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: of part of the TNT molecule can proceed . Overall, this pathway highlights how some microbes use unusual enzymatic reactions to destabilize TNT’s ring, releasing nitrite as a nutrient and facilitating...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: 2. Cabrera, M.Á., Márquez, S.L., & Pérez-Donoso, J.M. (2022). Comparative genomic analysis of Antarctic Pseudomonas isolates with 2,4,6-trinitrotoluene transformation capabilities reveals their unique...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Old Yellow Enzyme (OYE) Family & Flavoprotein Reductases: Enzymes of the OYE family are broad-spectrum flavoprotein oxidoreductases that can attack nitroaromatic compounds in alternative ways. Sever...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Key Intermediates and Fate: Across these pathways, the recurring intermediates are: hydroxylamino-dinitrotoluenes, amino-dinitrotoluenes (ADNTs), diaminonitrotoluenes (DANTs), and triaminotoluene (TAT...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: for further metabolism of TNT . Once a DNT or mononitrotoluene forms, bacteria can channel those into known catabolic pathways (many strains readily metabolize DNTs via dioxygenases that cleave the ri...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"\\n=== ANSWER ===\")\n",
    "print(result['result'])   \n",
    "print(\"\\n=== SOURCES ===\")\n",
    "for doc in result['source_documents']:\n",
    "    print(f\"Content: {doc.page_content[:200]}...\")\n",
    "    print(f\"Metadata title: {doc.metadata['source']}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce563e40",
   "metadata": {},
   "source": [
    "### Conversational Chain (with ConversationBufferMemory)\n",
    "For follow‑up calls use the same qa_chain object, and the memory will automatically include everything you said before.\n",
    "\n",
    "If you want to start a brand‑new chat without creating a new chain object:\n",
    "\n",
    "`qa_chain.memory.clear()`   # wipes the stored chat_history\n",
    "\n",
    "Or just instantiate a fresh chain:\n",
    "\n",
    "```\n",
    "qa_chain = local_rag_sys.build_qa_chain(\n",
    "    retriever_type=\"vec_semantic\",\n",
    "    use_conversation=True,\n",
    ")\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "6471ac04",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No split documents available for BM25Retriever creation. Need to build vector DB first.\n"
     ]
    }
   ],
   "source": [
    "# Complete workflow with existing documents vector db, running QA-chain with a query to the model\n",
    "# Build a conversational chain\n",
    "qa_chain = local_rag_sys.build_qa_chain(retriever_type=\"vec_semantic\", \n",
    "                                        use_conversation=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "e1bb75e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "🤖 **How SaProt differs from other protein‑LLMs such as ESM‑2**\n",
      "\n",
      "| Aspect | SaProt (SaProt paper) | ESM‑2 (ESM family) |\n",
      "|--------|----------------------|--------------------|\n",
      "| **Vocabulary** | Uses a *structure‑aware (SA) vocabulary* that fuses the ordinary amino‑acid token with a Foldseek‑derived 3Di structural token at each position. The embedding layer therefore contains 441 tokens (21 residues × 21 structure symbols). | Uses only the conventional 20 residue tokens (plus special tokens). |\n",
      "| **Input representation** | Each protein is encoded as an *SA‑token sequence* (e.g., “s₁f₁, s₂f₂ …”), where the second component carries explicit 3‑D conformation information. | Input consists solely of the linear amino‑acid string; any structural signal must be inferred implicitly from the sequence. |\n",
      "| **Model size & architecture** | Identical transformer backbone and parameter count to ESM‑2 650M; the only architectural change is the expanded embedding matrix for the SA tokens. (SaProt “shares an identical network architecture, model size, and training examples with ESM‑2, with the key difference lying in its structure‑aware vocabulary.”) | Same transformer backbone but with the original residue‑only embedding layer. |\n",
      "| **Training data** | Pre‑trained on ~40 M proteins together with ∼40 M predicted structures from AlphaFoldDB (encoded by Foldseek). | Pre‑trained only on protein sequences (no explicit structural tokens). |\n",
      "| **Performance on structure‑related tasks** | Outperforms ESM‑2 in contact‑map prediction (Table 3: SaProt P@L = 86.15 vs 55.30 for short‑range contacts, etc.) and shows superior zero‑shot mutational‑effect scores (Table 1: ClinVar AUC 0.909 vs 0.862). | Lower scores on the same tasks. |\n",
      "| **Embedding geometry** | In t‑SNE visualisations of the SCOPe non‑redundant set, SaProt’s embeddings for α and β proteins are *separated* according to structure type, while ESM‑2’s embeddings are *intertwined*. (Figure 5 caption: “For the alpha and beta proteins, the representations generated by ESM‑2 are intertwined, whereas those generated by SaProt are separated based on structure type.”) | Same backbone but without explicit structural separation. |\n",
      "\n",
      "### Source documents & metadata supporting the answer  \n",
      "\n",
      "| # | Document (title) | Authors / Year | Relevant excerpt |\n",
      "|---|------------------|----------------|-----------------|\n",
      "| 1 | **“SaProt: Protein language modeling with structure‑aware vocabulary”** | Jin Su, Chenchen Han, Yuyang Zhou, Junjie Shan, Xibin Zhou, Fajie Yuan (2023) – bioRxiv | *“We introduce the concept of a ‘structure‑aware vocabulary’ that integrates residue tokens with structure tokens… SaProt shares an identical network architecture, model size, and training examples with ESM‑2, **with the key difference lying in its structure‑aware vocabulary**.”* |\n",
      "| 2 | Same SaProt paper (Section 3.1 & Figure 5) | — | *“For the alpha and beta proteins, the representations generated by **ESM‑2 are intertwined**, whereas those generated by **SaProt are separated based on structure type**.”* |\n",
      "| 3 | Same SaProt paper (Table 3) | — | *Contact‑map results: “SaProt 86.15 … vs ESM‑2 55.30” showing superior structural prediction.* |\n",
      "| 4 | Same SaProt paper (Table 1 & discussion) | — | *Zero‑shot mutational effect: “SaProt AUC 0.909 vs ESM‑2 0.862”* |\n",
      "| 5 | Same SaProt paper (Data description) | — | *“≈40 million protein sequences and structures… encoded by Foldseek into 3Di tokens, forming SA‑token sequences.”* |\n",
      "\n",
      "These excerpts collectively demonstrate that **SaProt’s main distinction from ESM‑2 is the incorporation of an explicit structure‑aware token vocabulary**, which allows it to ingest 3‑D structural information directly during pre‑training and leads to consistently better performance on tasks that depend on protein structure.\n",
      "\n",
      "🤖 **Performance comparison**\n",
      "\n",
      "| Model | What the provided documents say about its performance | How it stacks up against **SaProt** |\n",
      "|-------|------------------------------------------------------|-----------------------------------|\n",
      "| **SaProt** | *“SaProt outperforms all residue‑sequence‑based and structure‑based models on both tasks”* (SaProt – Protein Language Modeling with Structure‑Aware Vocabulary, Westlake University) and “exhibits consistent improvement for predicting mutational effects … higher accuracy compared to MIF‑ST” (same source). | Baseline in the comparison – it is already state‑of‑the‑art among 35 M‑parameter PLMs that use only residue sequences. |\n",
      "| **Diffusion Sequence Model (DSM)** | *“DSM models match or outperform MLM‑based and discrete diffusion pLMs of the same size as well as an AR pLM almost twice DSM’s size”* (Diffusion Sequence Models for Enhanced Protein Representation and Generation, arXiv 2506.08293v1 9 Jun 2025).  <br> *“DSM650 produced the highest quality embeddings among similarly sized pLMs … only overtaken by the much larger ProtT5”* (same source). | DSM’s embeddings are **on par with or better than** those of SaProt for models of comparable size, and it even beats a much larger autoregressive model.  The paper does not give a direct head‑to‑head numeric comparison to SaProt, but the statements make clear that DSM reaches *similar* or *higher* representation quality while also adding strong generative abilities. |\n",
      "| **xTrimoPGLM** | *“xTrimoPGLM substantially outperforms other advanced baselines in 18 protein understanding benchmarks across four categories”* (xTrimoPGLM: unified 100‑billion‑parameter pretrained transformer for deciphering the language of proteins, Nature Methods Vol 22 May 2025). <br> *“xTrimoPGLM achieved a perplexity of 4.01 on CAMEO and 4.45 on CASP15, lower than ESMFold’s 5.21/6.18”* (same source). | With **100 B parameters**, xTrimoPGLM is far larger than SaProt (35–650 M) and therefore delivers higher absolute performance on a wide set of tasks. The document explicitly says it “substantially outperforms … baselines”, which include the ESM family that SaProt already beats, indicating that xTrimoPGLM exceeds SaProt’s performance when model scale is taken into account.\n",
      "\n",
      "**Design comparison**\n",
      "\n",
      "| Aspect | SaProt | DSM – what is new | xTrimoPGLM – what is new |\n",
      "|--------|--------|-------------------|--------------------------|\n",
      "| **Core architecture** | Uses the ESM‑2 transformer backbone (same size as the corresponding ESM model). | Extends the same ESM‑2 checkpoint but **adds a novel language‑modeling head** (extra linear layer, soft‑logit cap) and *ties* the final projection to the token‑embedding matrix. | A **unified transformer** of up to 100 B parameters that is trained jointly on two objectives (see below). |\n",
      "| **Input representation** | Introduces a **structure‑aware vocabulary** that concatenates residue tokens with Foldseek‑derived structure tokens, forming “SA‑tokens”. | Keeps the ordinary residue token space; the novelty lies in the **masked diffusion training objective**, not in new tokens. | Still works on plain amino‑acid sequences (no special SA‑vocab), but the model is designed to handle both *understanding* and *generation* tasks within a single network. |\n",
      "| **Training objective** | Standard masked language modeling (MLM) on the combined residue + structure token stream. | **Masked forward diffusion**: each training step randomly masks up to 90 % of tokens, then the model denoises them in a *single* forward pass. This enables “robust denoising across high corruption rates and sequence generation with global context” (Diffusion Sequence Models… arXiv 2506.08293v1). | **Curriculum‑style two‑phase pre‑training**: first 400 B tokens with pure MLM, then a second stage where *20 %* of batches use MLM and *80 %* use a generalized language modeling (GLM) objective, thereby unifying representation learning and autoregressive generation. |\n",
      "| **Generative capability** | Not designed for de‑novo sequence generation; focuses on representation quality. | Provides **unconditional generative sampling** that produces realistic protein sequences in one forward pass, matching natural amino‑acid k‑mer statistics and predicted structures/functions. | Supports both *de‑novo* and *conditional* design (e.g., binder generation) through the same model, leveraging its GLM head for autoregressive decoding. |\n",
      "| **Scale** | 35 M – 650 M parameters (the paper reports a 650 M version). | Same size as the underlying ESM‑2 checkpoint it extends (e.g., DSM‑150 M, DSM‑650 M). | 100 B parameters—two orders of magnitude larger than SaProt. |\n",
      "\n",
      "**Key architectural novelties**\n",
      "\n",
      "* **DSM**\n",
      "  * A **new language‑modeling head** (extra linear transformation + soft‑logit cap) tied to the embedding matrix (Diffusion Sequence Models…, arXiv 2506.08293v1).\n",
      "  * **Masked diffusion training** that treats masking as a forward diffusion process, allowing the model to denoise highly corrupted inputs and generate full sequences in one pass (same source).\n",
      "\n",
      "* **xTrimoPGLM**\n",
      "  * **Joint MLM + GLM pre‑training** with a curriculum that first minimizes MLM loss then blends 20 % MLM / 80 % GLM, enabling the same network to excel at both understanding and generation tasks (xTrimoPGLM…, Nature Methods Vol 22 May 2025).\n",
      "  * **Massive scale** (100 B parameters) combined with a unified objective, leading to lower perplexity on structure‑prediction benchmarks than dedicated PLMs such as ESMFold (same source).\n",
      "\n",
      "In summary, SaProt advances performance by enriching the token vocabulary with structural information while keeping a conventional MLM training regime. DSM builds on the same backbone but introduces a diffusion‑based denoising head that yields representations competitive with SaProt and adds strong generative power. xTrimoPGLM goes further in design by unifying two complementary language‑modeling objectives within an extremely large transformer, achieving substantially higher benchmark scores than both SaProt and DSM.\n"
     ]
    }
   ],
   "source": [
    "# First turn\n",
    "ans1 = qa_chain.invoke({\"question\": '''How does the SaProt protein-LLM differs from other protein-LLMs such as ESM-2? \n",
    "                        Use the given context (source documents) to answer the question.\n",
    "                        Give the name of the source documents and metadata that supports your answer.'''})\n",
    "print(\"\\n🤖\", ans1[\"answer\"])\n",
    "\n",
    "# Second turn – the model sees the previous answer in its context now\n",
    "ans2 = qa_chain.invoke({\"question\": '''Two new protein LLMs models called: Diffusion Sequence Model (DSM) and xTrimoPGLM are suppose to be state-of-the-art. \n",
    "                        Compare them to the model you discussed in the previous prompt.\n",
    "                        What are the key architectural novelties of these models?\n",
    "                        Use the given context (source documents) to answer the question.\n",
    "                        Give the name of the source documents and metadata that supports your answer. '''})\n",
    "print(\"\\n🤖\", ans2[\"answer\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "b21345fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Content: process. DSM extends ESM2 ( 33) with a novel language modeling head and training objective, enabling robust denoising\n",
      "across high corruption rates and sequence generation with global context. After tr...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Diffusion Sequence Models for Enhanced Protein Representation and Generation.pdf\n",
      "\n",
      "Content: arXiv:2506.08293v1  [q-bio.BM]  9 Jun 2025\n",
      "Diffusion Sequence Models for Enhanced Protein Representation and\n",
      "Generation\n",
      "Logan Hallee1, 2, Nikolaos Rafailidis1, David B. Bichara2, and Jason P. Gleghorn...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Diffusion Sequence Models for Enhanced Protein Representation and Generation.pdf\n",
      "\n",
      "Content: embeddings and valuable downstream tasks (datasets described in Supplemental Figure S3). DSM650 produced the highest\n",
      "quality embeddings among similarly sized pLMs, generating consistently high F1 scor...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Diffusion Sequence Models for Enhanced Protein Representation and Generation.pdf\n",
      "\n",
      "Content: • SaProt outperforms ESM-2 in all protein-level tasks. Specifically, SaProt shows remark-\n",
      "able enhancements over ESM-2 in the Thermostability, HumanPPI, Metal Ion Binding, and\n",
      "DeepLoc tasks. This outc...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf\n",
      "\n",
      "Content: of objectives, which has led to a strategy for training xTrimoPGLM at an \n",
      "unprecedented scale of 100 billion parameters and 1 trillion training tokens. Our \n",
      "extensive experiments reveal that (1) xTrim...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/xTrimoPGLM- unified 100-billion-parameter pretrained transformer for deciphering the language of proteins.pdf\n",
      "\n",
      "Content: SAPROT: P ROTEIN LANGUAGE MODELING WITH\n",
      "STRUCTURE -AWARE VOCABULARY\n",
      "Jin Su, Chenchen Han, Yuyang Zhou, Junjie Shan, Xibin Zhou, Fajie Yuan∗\n",
      "Westlake University\n",
      "{sujin, hanchenchen, zhouyuyang, shanjun...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf\n",
      "\n",
      "Content: tures. Through extensive evaluation, our SaProt model surpasses well-established\n",
      "and renowned baselines across 10 significant downstream tasks, demonstrating\n",
      "its exceptional capacity and broad applica...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf\n",
      "\n",
      "Content: Like the ESM models, SaProt aims to contribute to the advancement of the biological community.\n",
      "This study has several limitations: (1) The performance of the proposed SA vocabulary heavily\n",
      "depends on ...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf\n",
      "\n",
      "Content: non-redundant version (P IDE <40%) of the SCOPe (Chandonia et al., 2018) database. For the\n",
      "alpha and beta proteins, the representations generated by ESM-2 are intertwined, whereas those gen-\n",
      "erated by...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf\n",
      "\n",
      "Content: the return (Fig. 1b). We speculate that protein sequences, unlike natural \n",
      "language, are less smooth in semantic space and more akin to a factual \n",
      "nature. Currently, PLM-based methods still struggle t...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/xTrimoPGLM- unified 100-billion-parameter pretrained transformer for deciphering the language of proteins.pdf\n",
      "\n",
      "Content: In summary, DSM closes the gap between understanding and creating proteins in a single architecture, enabled by a few\n",
      "Python line changes, providing a foundation for rapid, scalable, and biologically ...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Diffusion Sequence Models for Enhanced Protein Representation and Generation.pdf\n",
      "\n",
      "Content: Nature Methods | Volume 22 | May 2025 | 1028–1039\n",
      " 1035\n",
      "Article https://doi.org/10.1038/s41592-025-02636-z\n",
      "test data performance. Thus, the performance of the test data suddenly \n",
      "increased, denoted as...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/xTrimoPGLM- unified 100-billion-parameter pretrained transformer for deciphering the language of proteins.pdf\n",
      "\n",
      "Content: ESM-2, given that residue-based PLMs also implicitly contain structure information (Rao et al.,\n",
      "2020)? To answer it, we conduct an additional structure prediction task, namely contact map predic-\n",
      "tion...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf\n",
      "\n",
      "Content: A C OMPARISON WITH PROST T5\n",
      "A very recent preprint in Heinzinger et al. (2023) proposed ProstT5 which pre-trains the protein\n",
      "language model (PLM) on a mixture of data of Foldseek token sequences and r...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf\n",
      "\n",
      "Content: Nature Methods | Volume 22 | May 2025 | 1028–1039 1028\n",
      "nature methods\n",
      "Article\n",
      "https://doi.org/10.1038/s41592-025-02636-z\n",
      "xTrimoPGLM: unified 100-billion-parameter \n",
      "pretrained transformer for decipheri...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/xTrimoPGLM- unified 100-billion-parameter pretrained transformer for deciphering the language of proteins.pdf\n",
      "\n",
      "Content: Interestingly, the diffusion-based pLMs showcased a distinct sequence reconstruction scaling law that became clear when\n",
      "contrasting DSM and DPLM compared to ESM2 (Figure 4). Even at 90% masking, the d...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Diffusion Sequence Models for Enhanced Protein Representation and Generation.pdf\n",
      "\n",
      "Content: 44. Xingyi Cheng, Bo Chen, Pan Li, Jing Gong, Jie Tang, and Le Song. Training compute-optimal protein language models. bioRxiv, 2024. doi: 10.1101/2024.06.06.597716.\n",
      "Pages: 2024.06.06.597716 Section: ...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Diffusion Sequence Models for Enhanced Protein Representation and Generation.pdf\n",
      "\n",
      "Content: 4.1.3 R ESULTS\n",
      "Table 1 showcases the zero-shot results on the ProteinGym and ClinVar datasets. Based on the\n",
      "results, we draw several noteworthy conclusions:\n",
      "• SaProt outperforms all residue sequence-b...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf\n",
      "\n",
      "Content: Model ClinVar ProteinGym\n",
      "w/o MSA retrieval w/ MSA retrieval\n",
      "AUC Spearman’s ρ Spearman’s ρ\n",
      "ESM-2 (15B) 0.843 0.436 0.479\n",
      "SaProt 0.909 0.478 0.489\n",
      "Table 6: Zero-shot comparison with ESM-2 15B version. C...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf\n",
      "\n",
      "Content: while CASP15 consists of 56 publicly available proteins (released in  \n",
      "May 2022). Initially, we observed the perplexity (the lower the better)  \n",
      "of these two datasets in PLMs, which served as the back...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/xTrimoPGLM- unified 100-billion-parameter pretrained transformer for deciphering the language of proteins.pdf\n",
      "\n",
      "Content: suffix supports both de novo and conditional protein design through sequence \n",
      "generation. b, xTrimoPGLM shows lower perplexity than other leading PLMs like \n",
      "ESM2 and ProGen2-xlarge in evaluations on t...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/xTrimoPGLM- unified 100-billion-parameter pretrained transformer for deciphering the language of proteins.pdf\n",
      "\n",
      "Content: Model ClinV\n",
      "ar ProteinGym(w/o MSA retrieval) DeepLoc\n",
      "Subcellular\n",
      "Binary\n",
      "A\n",
      "UC Spearman’s ρ Acc% Acc%\n",
      "Ev\n",
      "oformer-inspired ESM(35M) 0.589\n",
      "0.178 66.11 89.25\n",
      "MIF 0.638\n",
      "0.256 61.90 85.90\n",
      "SaProt(35M) 0.754\n",
      "0...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf\n",
      "\n",
      "Content: quence, serving as the input for existing residue-based PLMs. Through unsupervised training on\n",
      "massive protein SA-token sequences, we obtain a Structure-aware Protein language model named\n",
      "SaProt. To a...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf\n",
      "\n",
      "Content: summary, while scaling single-sequence models enhances the perfor-\n",
      "mance, it still struggles with OOD data, which is effectively addressed \n",
      "by MSA-based augmentation.\n",
      "Evaluation of protein sequence ge...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/xTrimoPGLM- unified 100-billion-parameter pretrained transformer for deciphering the language of proteins.pdf\n",
      "\n",
      "Content: allows for seamless integration with advances in large-scale foundation AI models, such as\n",
      "BERT (Devlin et al., 2018), BART (Lewis et al., 2019), GPT (Brown et al., 2020), etc.\n",
      "• By utilizing the SA-t...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf\n",
      "\n",
      "Content: advancement in the field of protein language modeling.\n",
      "Evaluation of protein structure prediction\n",
      "Protein sequences encode rich information about their 3D structures \n",
      "and function through evolutionary...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/xTrimoPGLM- unified 100-billion-parameter pretrained transformer for deciphering the language of proteins.pdf\n",
      "\n",
      "Content: For DSM pretaining, we used OMG prot50, a protein-only dataset containing over 207\n",
      "million sequences after clustering the Open MetaGenomic dataset (OMG) at 50% sequence\n",
      "identity (36). We randomly remo...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/Diffusion Sequence Models for Enhanced Protein Representation and Generation.pdf\n",
      "\n",
      "Content: for predicting subsequent words, the prefix-encoding portion \n",
      "remains bidirectional.\n",
      "The framework of xTrimoPGLM, along with its application in down-\n",
      "stream tasks, is depicted in Fig. 1a. Motivated by...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/xTrimoPGLM- unified 100-billion-parameter pretrained transformer for deciphering the language of proteins.pdf\n",
      "\n",
      "Content: We compare SaProt with two types of baselines: sequence-based models and structure-based mod-\n",
      "els. For sequence-based models, we include ESM-1b (Rives et al., 2019), ESM-1v (Meier et al.,\n",
      "2021) (the r...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/SaProt- Protein Language Modeling with Structure-aware Vocabulary.pdf\n",
      "\n",
      "Content: Proteins play vital roles in the sustenance, growth and defense mecha-\n",
      "nisms of living organisms. They provide structural support for many \n",
      "essential biological processes such as synthesizing enzymes,...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/pdfs/xTrimoPGLM- unified 100-billion-parameter pretrained transformer for deciphering the language of proteins.pdf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for doc in ans2['source_documents']:\n",
    "    print(f\"Content: {doc.page_content[:200]}...\")\n",
    "    print(f\"Metadata title: {doc.metadata['source']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "be99a580",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "🤖 **Summary of the main groups of TNT‑degrading microbes (bacteria)**  \n",
      "\n",
      "| **Category** | **Typical genera / species mentioned in the source material** | **Characteristic metabolic features** |\n",
      "|--------------|---------------------------------------------------------------|---------------------------------------|\n",
      "| **Aerobic bacteria** | • *Pseudomonas* spp. (e.g., *P. fluorescens*, *P. putida*, strain IIBx)  <br>• *Bacillus* spp.<br>• *Staphylococcus* spp.<br>• *Mycobacterium* spp.<br>• *Rhodococcus* spp.<br>• *Serratia marcescens*<br>• *Stenotrophomonas* isolates | – Operate under oxic conditions. <br>– Primary route is **nitro‑group reduction** (NO₂ → NO → NHOH → NH₂) producing hydroxylamino‑ and amino‑dinitrotoluenes (ADNTs, DANTs). <br>– Some strains can form **Meisenheimer complexes** that eject nitrite, yielding dinitrotoluene (DNT) which can then be attacked by dioxygenases. <br>– Generally stop at partially reduced intermediates (ADNT, azoxy dimers); complete mineralisation of the aromatic ring is rare in a single aerobic isolate. <br>– Often use TNT only as a **nitrogen source**; carbon from the toluene moiety may be further degraded if DNT is formed. |\n",
      "| **Anaerobic bacteria** | • *Veillonella alkalescens*<br>• *Escherichia coli* (facultative, under strict anoxia)<br>• *Clostridium* spp. (e.g., *C. bifermentans*, *C. acetobutylicum*, *C. pasterianum*, *C. sordellii*)<br>• *Desulfovibrio* spp.<br>• *Methanococcus* spp. | – Operate in the **absence of O₂** (strict anaerobes or facultatives under anoxic conditions). <br>– Carry out a **full six‑electron reduction** of each nitro group, ultimately converting TNT → 2,4,6‑triaminotoluene (**TAT**). <br>– Reduction proceeds through nitroso and hydroxylamino intermediates; the process is often coupled to **fermentation or anaerobic respiration**, sometimes using TNT as a terminal electron acceptor. <br>– TAT is relatively stable under anoxia but can polymerise/oxidise when exposed to O₂; further degradation of the aromatic ring is rarely observed, so the pathway mainly detoxifies TNT by removing nitro groups. |\n",
      "\n",
      "### Main Differences between Aerobic and Anaerobic TNT‑degrading bacteria  \n",
      "\n",
      "| Aspect | Aerobic bacteria | Anaerobic bacteria |\n",
      "|--------|-------------------|--------------------|\n",
      "| **Electron acceptor** | O₂; reactions are generally reductive but occur in the presence of oxygen. | No O₂; reduction uses H₂, NAD(P)H, or other internal electron donors; some use TNT itself as an electron sink/terminal electron acceptor. |\n",
      "| **Extent of nitro‑group removal** | Usually **partial**: 1–2 nitro groups are reduced; ADNTs/DANTs accumulate. Full three‑nitro reduction is rare. | Typically **complete**: all three nitro groups are reduced, yielding TAT. |\n",
      "| **Aromatic‑ring fate** | Limited ring attack; Meisenheimer complex formation can lead to denitration (producing DNT) which may be further oxidised by dioxygenases, but full mineralisation is uncommon. | Ring remains largely intact after nitro reduction; TAT does not undergo efficient oxidative cleavage under anoxia. |\n",
      "| **Typical metabolic outcome** | Production of toxic or recalcitrant amino‑nitrotoluene intermediates and azoxy polymers; often used only for nitrogen acquisition. | Detoxification by converting the explosive nitro groups to amines; energy gain is modest (often coupled to fermentative metabolism). |\n",
      "| **Representative genera** | *Pseudomonas*, *Bacillus*, *Rhodococcus*, *Mycobacterium*, *Serratia* etc. | *Clostridium*, *Veillonella*, *Desulfovibrio*, *Methanococcus*, facultative *E. coli*. |\n",
      "| **Environmental application** | Often employed in **aerobic polishing stages** after an anaerobic reduction step to degrade the resulting DNT/ADNT intermediates. | Used in **anaerobic pre‑treatment** (e.g., bioreactors, sediment zones) to eliminate nitro groups and lower explosive toxicity before an aerobic phase. |\n",
      "\n",
      "---\n",
      "\n",
      "## Source Documents & Metadata Supporting This Answer  \n",
      "\n",
      "| # | Document / Section (as presented in the provided context) | Type of information used |\n",
      "|---|-----------------------------------------------------------|--------------------------|\n",
      "| 1 | **Scope of Document** – “This document provides a comprehensive analysis of microbial TNT degradation.” | General framing that the text is a review of microbial TNT degraders. |\n",
      "| 2 | **TNT‑Degrading Microbial Species** (section) | Lists aerobic bacteria (*Pseudomonas*, *Bacillus*, etc.) and anaerobic organisms (*Veillonella*, *Clostridium*, *Desulfovibrio*), describing their transformations. |\n",
      "| 3 | **Multi‑Omics Approaches and Systems Biology Insights** – subsections on *Clostridium spp.* and other aerobes | Provides details on anaerobic complete reduction to TAT versus aerobic partial reductions (ADNT, DANT). |\n",
      "| 4 | **Genomic Insights: Key Genes and Enzymes** – “Nitroreductases (Type I)” & “Azoreductases” | Describes enzymatic mechanisms typical of aerobic nitro‑group reduction. |\n",
      "| 5 | **Anaerobic Degradation Routes** (paragraphs) | Explicit description of full six‑electron reductions, production of TAT, and use of TNT as a terminal electron acceptor. |\n",
      "| 6 | **References – Ramos J.L. et al. (2001)** “Biological degradation of 2,4,6‑trinitrotoluene (TNT).” | Primary literature source that historically documents both aerobic and anaerobic degraders. |\n",
      "| 7 | **Cabrera M.A. et al. (2022 & 2024)** – comparative genomics of Antarctic *Pseudomonas* isolates | Supports the identification of aerobic *Pseudomonas* species with high xenobiotic‑degrading gene content. |\n",
      "| 8 | **Liao H.Y. et al. (2018)** – transcriptomic & proteomic analysis of *Citrobacter* sp. | Provides evidence for nitroreductase‐driven aerobic TNT transformation. |\n",
      "| 9 | **Ziganshin A.M. et al. (2007)** – hydride‑mediated reduction by yeasts (included to contrast bacterial pathways) | Highlights that anaerobic bacteria achieve full nitro reduction, while yeasts use a different mechanism; reinforces the bacterial distinction. |\n",
      "|10| **EAWAG‑BBD (1999)** “2,4,6‑Trinitrotoluene Degradation Pathway (Aerobic and Anaerobic).” | External database summarising both aerobic and anaerobic routes, corroborating the differences outlined above. |\n",
      "\n",
      "*All statements in the answer are directly drawn from these sections of the supplied context; no external knowledge was introduced.*\n",
      "\n",
      "🤖 **Aerobic bacteria – principal TNT‑degrading genes / enzymes**\n",
      "\n",
      "| Gene / enzyme (function) | What it does in the TNT pathway | Source(s) that mention it |\n",
      "|---------------------------|--------------------------------|----------------------------|\n",
      "| **Type I nitroreductases** (e.g. *nfsA*, *nfsB*) | Two‑electron FMN/FAD‑dependent reduction of a nitro group to nitroso → hydroxylamino → amine; the first entry step for most aerobic degraders. | “Genomic Insights: Key Genes and Enzymes” – description of Type I nitroreductases; also cited in the *Metabolic Pathways of TNT Degradation* section. |\n",
      "| **nemA** (N‑ethylmaleimide reductase) | Reduces TNT to 4‑ADNT and 2‑ADNT; identified as the most up‑regulated gene in a Citrobacter sp. TNT degrader. | “Metabolic Pathways of TNT Degradation” – paragraph on Nitroreductases (Type I). |\n",
      "| **Azoreductase** (azoreductase‑encoding gene) | Highly induced (>1 000‑fold) during TNT exposure; can reduce nitro‑aromatics or hydroxylamine intermediates, acting as a key TNT‑transforming enzyme in *Pseudomonas* sp. TNT3. | “Genomic Insights: Key Genes and Enzymes” (Azoreductases paragraph) and the transcriptomic study of *Pseudomonas* sp. TNT3 under “Aerobic Bacteria”. |\n",
      "| **Old Yellow Enzyme (OYE) family reductases** (oxygen‑insensitive nitroreductases) | Catalyze hydride transfer that generates Meisenheimer σ‑complexes, leading to nitrite release and denitration of TNT. | “Genomic Insights: Key Genes and Enzymes” – mention of OYEs; also in the “Metabolic Pathways” discussion of hydride‑mediated reduction. |\n",
      "| **PETN reductase** (pentaerythritol tetranitrate reductase) | NADPH‑dependent enzyme that attacks TNT’s aromatic ring, forming mono‑ and di‑hydride Meisenheimer complexes; releases nitrite. | “Aerobic Bacteria” – description of *Enterobacter cloacae* PB2. |\n",
      "| **Flavoprotein reductases** (general flavoproteins in *Pseudomonas fluorescens*) | Reduce TNT to hydroxylamino‑ and amino‑dinitrotoluenes; also can mediate ring hydrogenation to Meisenheimer complexes. | “Aerobic Bacteria” – paragraph on *Pseudomonas* spp. |\n",
      "| **Efflux pumps & stress‑response enzymes** (e.g., superoxide dismutase, peroxidases) | Not catalytic for TNT itself but essential for tolerance and allowing the catabolic enzymes to function. | “Multi‑Omics Approaches and Systems Biology Insights” – list of accessory/regulatory genes. |\n",
      "\n",
      "**Anaerobic bacteria – principal TNT‑degrading genes / enzymes**\n",
      "\n",
      "| Gene / enzyme (function) | What it does in the TNT pathway | Source(s) that mention it |\n",
      "|---------------------------|--------------------------------|----------------------------|\n",
      "| **Oxygen‑insensitive nitroreductases** (same families as aerobic, but operating under anoxic conditions) | Carry out stepwise six‑electron reductions of each –NO₂ to –NH₂, producing triaminotoluene (TAT). | “Anaerobic Bacteria” sections describing Veillonella, *Clostridium* spp., Desulfovibrio spp.; also the general nitroreductase discussion in “Genomic Insights”. |\n",
      "| **Nitrite reductase (NasB)** & associated high‑affinity nitrite uptake systems | Assimilate nitrite released from TNT denitration and reduce it to ammonia for biomass synthesis; links TNT reduction to growth on nitrogen. | “Anaerobic Bacteria” – paragraph on *Pseudomonas putida* JLR11’s nitrite uptake/NasB system; also in “Nitrite Reductases and Assimilatory Enzymes”. |\n",
      "| **Hydrogenases / H₂‑oxidizing enzymes** (implicit in Clostridial/Veillonella metabolism) | Provide the strong reducing power needed for the six‑electron reductions of nitro groups under strict anaerobic conditions. | “Anaerobic Bacteria” – description of *Clostridium* spp. using H₂ as electron donor to fully reduce TNT. |\n",
      "| **Bamberger‑rearrangement enzymes / aminophenol‑forming activities** (not named, but implied) | Convert hydroxylamino‑dinitrotoluene intermediates into aminophenols after anaerobic reduction. | “Anaerobic Bacteria” – discussion of Clostridial pathways that generate aminophenols via Bamberger rearrangement. |\n",
      "| **General stress‑response and DNA‑repair proteins** (e.g., nitric‑oxide dioxygenase) | Protect cells from nitrosative damage generated during anaerobic TNT reduction; essential for survival but not directly catabolic. | “Multi‑Omics Approaches” – list of nitrosative‑stress genes up‑regulated in TNT‑challenged microbes. |\n",
      "\n",
      "**Documents & metadata that support the answer**\n",
      "\n",
      "| Information extracted | Document / section title (as given) | Reference/metadata |\n",
      "|----------------------|-------------------------------------|--------------------|\n",
      "| Description of Type I nitroreductases, *nfsA/nfsB* and their role in TNT reduction | “Genomic Insights: Key Genes and Enzymes” – Nitroreductases paragraph; also under “Metabolic Pathways of TNT Degradation”. | No formal citation number; part of the main body. |\n",
      "| *nemA* (N‑ethylmaleimide reductase) activity on TNT | Same “Metabolic Pathways of TNT Degradation” section. | – |\n",
      "| Azoreductase gene highly induced in *Pseudomonas* sp. TNT3 | “Genomic Insights: Key Genes and Enzymes” (Azoreductases paragraph) and the transcriptomic study of *Pseudomonas* sp. TNT3 under “Aerobic Bacteria”. | – |\n",
      "| OYE family enzymes mediating hydride addition/Meisenheimer complexes | “Genomic Insights: Key Genes and Enzymes”; also in the pathway discussion (“Metabolic Pathways”). | – |\n",
      "| PETN reductase (pentaerythritol tetranitrate reductase) from *Enterobacter cloacae* PB2 | “Aerobic Bacteria” – paragraph on Enterobacteriaceae. | – |\n",
      "| Flavoprotein enzymes in *Pseudomonas fluorescens* | “Aerobic Bacteria” – description of Pseudomonas spp. | – |\n",
      "| Accessory genes (efflux pumps, oxidative‑stress enzymes) | “Multi‑Omics Approaches and Systems Biology Insights”. | – |\n",
      "| Anaerobic complete reduction to TAT by Veillonella, *Clostridium*, Desulfovibrio | “Anaerobic Bacteria” sections (multiple paragraphs). | – |\n",
      "| Nitrite reductase NasB & uptake system in *Pseudomonas putida* JLR11 | “Anaerobic Bacteria” – paragraph on nitrite assimilation; also referenced under “Nitrite Reductases and Assimilatory Enzymes”. | – |\n",
      "| Hydrogen‑dependent reduction in Clostridia | “Anaerobic Bacteria” – description of H₂ as electron donor. | – |\n",
      "| Bamberger rearrangement / aminophenol formation in anaerobes | “Anaerobic Bacteria” – discussion of Clostridial pathways. | – |\n",
      "| General stress‑response genes (nitrosative, oxidative) in TNT‑exposed microbes | “Multi‑Omics Approaches and Systems Biology Insights”. | – |\n",
      "\n",
      "All statements above are drawn **exclusively** from the supplied source material; no external knowledge has been added.\n"
     ]
    }
   ],
   "source": [
    "# First turn\n",
    "ans1 = qa_chain.invoke({\"question\": '''Summarize the main group of TNT degrading microbial species, separate into Aerobic vs. Anaerobic bacteria, what are the main differences? \n",
    "                        Use the given context (source documents) to answer the question.\n",
    "                        Give the name of the source documents and metadata that supports your answer.'''})\n",
    "print(\"\\n🤖\", ans1[\"answer\"])\n",
    "\n",
    "# Second turn – the model sees the previous answer in its context now\n",
    "ans2 = qa_chain.invoke({\"question\": '''What are the main genes and enzymes involved in TNT degradation of the Aerobic vs. Anaerobic bacteria that you have summarized? \n",
    "                        Use the given context (source documents) to answer the question.\n",
    "                        Give the name of the source documents and metadata that supports your answer. '''})\n",
    "print(\"\\n🤖\", ans2[\"answer\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "b121ec54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Content: Scope of Document: This document provides a comprehensive analysis of microbial TNT degradation. We first catalog the known TNT-degrading species across bacteria, fungi, and yeasts, highlighting their...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Genomic Insights: Key Genes and Enzymes\n",
      "\n",
      "\n",
      "\n",
      "Microbial TNT degradation relies on a suite of specialized enzymes that initiate TNT transformation despite its inert aromatic structure. Genomic and biochem...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Comparative Genomic Insights: Genome sequencing of TNT-degrading strains has shed light on their evolutionary adaptations. For example, three Antarctic Pseudomonas isolates (TNT3, TNT11, TNT19) were s...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Azoreductases: An unexpected finding from genomic studies is the involvement of azoreductase enzymes in TNT degradation. Azoreductases normally reduce azo bonds (–N=N–) in dyes, but some can also re...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: In summary, the genetic toolkit for TNT biodegradation comprises mostly reductive enzymes that either remove nitro groups or destabilize the aromatic ring, along with supporting systems to manage stre...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Genomic and Metabolic Insights into Microbial Biodegradation of TNT\n",
      "\n",
      "\n",
      "\n",
      "Abstract...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Nitroreductases (Type I): A common entry point for TNT metabolism is the stepwise reduction of nitro (-NO₂) groups to nitroso (-NO), hydroxylamino (-NHOH), and amino (-NH₂) groups. These reactions a...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: In summary, TNT degradation pathways are complex and can follow multiple routes depending on the organism and conditions. Aerobic processes excel at initial transformations and co-metabolic turnover o...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Accessory and Regulatory Genes: Alongside catabolic enzymes, microbes possess regulatory and support genes that enable TNT degradation. These include: (1) oxidative stress response genes – TNT redox...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Aerobic Bacteria: Under oxic conditions, many soil bacteria can initiate TNT degradation by partial reduction of its nitro groups or removal of nitrite. Notable examples include:\n",
      "\n",
      "• Pseudomonas spp.: ...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Transcriptomic Profiling: Genome-wide RNA expression analyses reveal which genes are turned on or off in the presence of TNT. A recent study examined the transcriptome of Pseudomonas sp. TNT3 (an An...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: In summary, omics studies have expanded our understanding from individual enzymes to the global response of microorganisms to TNT. They reveal that TNT degradation is an orchestrated event involving s...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Other Anaerobes: Veillonella alkalescens was one of the first organisms noted to reduce TNT to TAT under strictly anaerobic conditions. Even Escherichia coli (facultative anaerobe) can perform compl...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: 7. Hannink, N.K., et al. (2001). Phytodetoxification of TNT by transgenic plants expressing a bacterial nitroreductase. Nature Biotechnology, 19(12): 1168–1172 .\n",
      "\n",
      "8. Duque, E., et al. (1993). Identifi...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Abstract\n",
      "\n",
      "2,4,6-Trinitrotoluene (TNT) is a recalcitrant explosive pollutant whose breakdown in the environment is facilitated by diverse microbes. This document summarizes key findings on TNT biodegra...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Systems Biology and Modeling: With large omics datasets, scientists have begun constructing regulatory and metabolic network models for TNT-degrading organisms. For instance, integrating the transcr...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Functional Genomics: Another omics-related approach is random mutagenesis or knockout studies combined with sequencing to find genes essential for TNT metabolism. Transposon mutagenesis, for instanc...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: . These azoxy compounds are intensely colored, highly recalcitrant, and represent dead-end immobilization products in many TNT-fed cultures . In summary, aerobic nitro-reduction rapidly eliminates TNT...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Microbial Bioremediation: The use of microorganisms to detoxify and remove TNT from the environment offers a promising alternative to costly physicochemical cleanup methods . Certain bacteria and fung...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Complete Reduction (Anaerobic Pathway): In oxygen-free environments, TNT can be fully reduced to TAT (2,4,6-triaminotoluene) by certain bacteria. This occurs via sequential addition of six electrons (...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Clostridium spp.: These obligate anaerobes are among the most effective TNT degraders in anoxic conditions. Clostridium bifermentans strains transform TNT through successive reductions to diaminonit...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: to its electron transport chain, certain metabolic pathways, and motility functions during TNT exposure. This likely reflects a shift to a slower-growing, stress-resistant state: the bacterium conserv...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Enterobacteriaceae: Enterobacter cloacae PB2 is capable of denitrating TNT. It reduces TNT via formation of a Meisenheimer complex and releases nitrite . An NADPH-dependent enzyme called pentaerythr...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Other Aerobes: Bacillus and Staphylococcus species can reduce TNT, typically to monoamino dinitrotoluene (e.g., 2-amino-4,6-dinitrotoluene) with nitrite release. Mycobacterium and Rhodococcus (Gram-...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: 2. Cabrera, M.Á., Márquez, S.L., & Pérez-Donoso, J.M. (2022). Comparative genomic analysis of Antarctic Pseudomonas isolates with 2,4,6-trinitrotoluene transformation capabilities reveals their unique...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Metabolomics and Pathway Confirmation: While less reported than transcript/proteome data, metabolomic analyses (identifying intermediate compounds) and isotope tracing have been used to confirm TNT ...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Dioxygenases and Ring-Cleavage Enzymes: Direct oxidative ring cleavage of TNT is very challenging due to its electron-poor ring, and no dedicated TNT dioxygenase has been identified. However, once T...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Sulfate-Reducers: Desulfovibrio species can anaerobically respire TNT. For example, Desulfovibrio sp. strain B uses TNT as a nitrogen source and may produce toluene as a downstream product . Other D...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: • Nitrite Reductases and Assimilatory Enzymes: Though not directly modifying TNT, enzymes like assimilatory nitrite reductase (NasB) and ammonium-assimilating enzymes (e.g., glutamine synthetase) are ...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n",
      "Content: Anaerobic Degradation Routes: In oxygen-free environments, TNT undergoes more complete reduction. The hallmark of anaerobic TNT metabolism is the production of triaminotoluene (TAT). This results from...\n",
      "Metadata title: /Users/danid/Downloads/RAG_personal_knowledge_base/new_docs_to_add/Genomic and Metabolic Insights into Microbial Biodegradation of TNT.docx\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for doc in ans2['source_documents']:\n",
    "    print(f\"Content: {doc.page_content[:200]}...\")\n",
    "    print(f\"Metadata title: {doc.metadata['source']}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
